Index: orbsvcs/Notify_Service/Notify_Service.cpp
===================================================================
--- orbsvcs/Notify_Service/Notify_Service.cpp	(revision 81069)
+++ orbsvcs/Notify_Service/Notify_Service.cpp	(working copy)
@@ -44,6 +44,11 @@
                                 command_line.get_ASCII_argv(),
                                 "");
 
+  if (CORBA::is_nil (this->orb_.in ()))
+    ACE_ERROR_RETURN ((LM_ERROR,
+                       " (%P|%t) Unable to create an ORB.\n"),
+                      -1);
+  
   CORBA::Object_var object =
     this->orb_->resolve_initial_references("RootPOA");
 
@@ -123,6 +128,14 @@
         ACE_Sched_Params::priority_min (ACE_Utils::truncate_cast<ACE_Sched_Params::Policy> (this->orb_->orb_core ()->orb_params ()->sched_policy ()),
                                         ACE_Utils::truncate_cast<int> (this->orb_->orb_core ()->orb_params ()->scope_policy ()));
 
+      // If the sched_policy is zero, this indicates that the user
+      // specified NONE as a parameter to -ORBSchedPolicy.  In this case,
+      // we must set the priority to the ACE default in order for the
+      // sched_policy flags to actually remain "NONE".  See
+      // ACE_OS::thr_create() for details.
+      if (this->orb_->orb_core ()->orb_params ()->sched_policy () == 0)
+        priority = ACE_DEFAULT_THREAD_PRIORITY;
+
       if (worker_.activate (flags,
                             this->nthreads_, 0, priority) != 0)
         ACE_ERROR_RETURN ((LM_ERROR,
Index: tao/ORB_Core.cpp
===================================================================
--- tao/ORB_Core.cpp	(revision 81069)
+++ tao/ORB_Core.cpp	(working copy)
@@ -1045,7 +1045,78 @@
 
           arg_shifter.consume_arg ();
         }
+      else if (0 != (current_arg = arg_shifter.get_the_parameter
+                                 (ACE_TEXT("-ORBSchedPolicy"))))
+        {
+          const ACE_TCHAR *name = current_arg;
+          if (ACE_OS::strcasecmp (name,
+                                  ACE_TEXT("SCHED_OTHER")) == 0)
+            {
+              this->orb_params_.ace_sched_policy (ACE_SCHED_OTHER);
+              this->orb_params_.sched_policy (THR_SCHED_DEFAULT);
+            }
+          else if (ACE_OS::strcasecmp (name,
+                                       ACE_TEXT("SCHED_FIFO")) == 0)
+            {
+              this->orb_params_.ace_sched_policy (ACE_SCHED_FIFO);
+              this->orb_params_.sched_policy (THR_SCHED_FIFO);
+            }
+          else if (ACE_OS::strcasecmp (name,
+                                       ACE_TEXT("SCHED_RR")) == 0)
+            {
+              this->orb_params_.ace_sched_policy (ACE_SCHED_RR);
+              this->orb_params_.sched_policy (THR_SCHED_RR);
+            }                             
+          else if (ACE_OS::strcasecmp (name,
+                                       ACE_TEXT("NONE")) == 0)
+            {
+              this->orb_params_.ace_sched_policy (0);
+              this->orb_params_.sched_policy (0);
+            }
+          else
+            {
+              ACE_ERROR ((LM_ERROR,
+                          ACE_TEXT ("Invalid argument '%s' ")
+                          ACE_TEXT ("for -ORBSchedPolicy\n"),
+                          current_arg));
+              throw ::CORBA::INTERNAL (
+                CORBA::SystemException::_tao_minor_code (
+                  TAO_ORB_CORE_INIT_LOCATION_CODE,
+                  0),
+                CORBA::COMPLETED_NO);
+            }
 
+          arg_shifter.consume_arg ();
+        }
+      else if (0 != (current_arg = arg_shifter.get_the_parameter
+                                 (ACE_TEXT("-ORBScopePolicy"))))
+        {
+          const ACE_TCHAR *name = current_arg;
+          if (ACE_OS::strcasecmp (name,
+                                  ACE_TEXT("SYSTEM")) == 0)
+            this->orb_params_.scope_policy (THR_SCOPE_SYSTEM);
+          else if (ACE_OS::strcasecmp (name,
+                                       ACE_TEXT("PROCESS")) == 0)
+            this->orb_params_.scope_policy (THR_SCOPE_PROCESS);
+          else if (ACE_OS::strcasecmp (name,
+                                       ACE_TEXT("NONE")) == 0)
+            this->orb_params_.scope_policy (0);
+          else
+            {
+              ACE_ERROR ((LM_ERROR,
+                          ACE_TEXT ("Invalid argument '%s' ")
+                          ACE_TEXT ("for -ORBScopePolicy\n"),
+                          current_arg));
+              throw ::CORBA::INTERNAL (
+                CORBA::SystemException::_tao_minor_code (
+                  TAO_ORB_CORE_INIT_LOCATION_CODE,
+                  0),
+                CORBA::COMPLETED_NO);
+            }
+
+          arg_shifter.consume_arg ();
+        }                            
+
       ////////////////////////////////////////////////////////////////
       // catch any unknown -ORB args                                //
       ////////////////////////////////////////////////////////////////
