Index: ORB_Core.cpp
===================================================================
--- ORB_Core.cpp	(revision 85444)
+++ ORB_Core.cpp	(working copy)
@@ -294,6 +294,7 @@
   ACE_NEW (this->request_dispatcher_,
            TAO_Request_Dispatcher);
 
+  this->set_sync_scope_hook (TAO_ORB_Core::default_sync_scope_hook);
 }
 
 TAO_ORB_Core::~TAO_ORB_Core (void)
@@ -2925,6 +2926,16 @@
 }
 
 void
+TAO_ORB_Core::default_sync_scope_hook (TAO_ORB_Core *,
+                                       TAO_Stub *,
+                                       bool &has_synchronization,
+                                       Messaging::SyncScope &scope)
+{
+  has_synchronization = true;
+  scope = Messaging::SYNC_WITH_TRANSPORT;
+}
+
+void
 TAO_ORB_Core::call_sync_scope_hook (TAO_Stub *stub,
                                     bool &has_synchronization,
                                     Messaging::SyncScope &scope)
Index: ORB_Core.h
===================================================================
--- ORB_Core.h	(revision 85444)
+++ ORB_Core.h	(working copy)
@@ -537,6 +537,12 @@
 
   void set_sync_scope_hook (Sync_Scope_Hook hook);
 
+  /// Default Sync_Scope_Hook.
+  static void default_sync_scope_hook (TAO_ORB_Core *,
+                                       TAO_Stub *,
+                                       bool &,
+                                       Messaging::SyncScope &);
+
   /// Handle to the factory for protocols_hooks_..
   TAO_Protocols_Hooks *protocols_hooks_;
 
