===================================================================
RCS file: /project/www-ace/htdocs/cvsweb-1.80/cvs-repository/ACE_wrappers/ace/config-g++-common.h,v
retrieving revision 4.38
retrieving revision 4.39
diff -u -r4.38 -r4.39
--- ACE_wrappers/ace/config-g++-common.h        2000/03/23 21:27:02     4.38
+++ ACE_wrappers/ace/config-g++-common.h        2000/09/28 01:32:23     4.39
@@ -1,5 +1,5 @@
 /* -*- C++ -*- */
-// $Id: config-g++-common.h,v 4.38 2000/03/23 21:27:02 nanbor Exp $
+// $Id: config-g++-common.h,v 4.39 2000/09/28 01:32:23 othman Exp $
 
 // This configuration file is designed to be included by another,
 // specific configuration file.  It provides config information common
@@ -23,8 +23,20 @@
 # define ACE_HAS_STD_TEMPLATE_SPECIALIZATION
 # define ACE_HAS_STANDARD_CPP_LIBRARY 1
 # define ACE_USES_OLD_IOSTREAMS
-// For some reason EGCS doesn't define this in its stdlib.
-# define ACE_LACKS_AUTO_PTR
+
+
+# if __GNUC__ <= 2 && __GNUC_MINOR__ < 95
+    // G++ 2.95.x and better properly support the auto_ptr class.
+#   define ACE_LACKS_AUTO_PTR
+
+    // Some versions of egcs, e.g., egcs-2.90.27 980315 (egcs-1.0.2
+    // release), have bugs with static data members in template
+    // classes.
+#   define ACE_LACKS_STATIC_DATA_MEMBER_TEMPLATES
+
+#   define ACE_HAS_GNUC_BROKEN_TEMPLATE_INLINE_FUNCTIONS
+# endif  /* G++ 2.91 or less */
+
 
 # if __GNUC__ == 2 && __GNUC_MINOR__ >= 91
 #   define ACE_HAS_USING_KEYWORD
@@ -40,10 +52,6 @@
 #   define ACE_HAS_TEMPLATE_SPECIALIZATION
 # endif /* __GNUC__ != 2.9  &&  __GNUC__ != 2.91*/
 
-  // Some versions of egcs, e.g., egcs-2.90.27 980315 (egcs-1.0.2 release),
-  // have bugs with static data members in template classes.
-# define ACE_LACKS_STATIC_DATA_MEMBER_TEMPLATES
-
   // __EXCEPTIONS is defined with -fexceptions, the egcs default.  It
   // is not defined with -fno-exceptions, the ACE default for g++.
   // ACE_HAS_EXCEPTIONS is defined in
@@ -60,7 +68,6 @@
 #else  /* ! egcs */
   // Plain old g++.
 # define ACE_LACKS_PLACEMENT_OPERATOR_DELETE
-# define ACE_LACKS_STATIC_DATA_MEMBER_TEMPLATES
 # define ACE_HAS_GNUG_PRE_2_8
 # define ACE_HAS_TEMPLATE_SPECIALIZATION
 # define ACE_LACKS_MIN_MAX_TEMPLATES
@@ -93,7 +100,7 @@
 #else  /* ! ACE_HAS_GNU_REPO */
 # define ACE_HAS_EXPLICIT_TEMPLATE_INSTANTIATION
 #endif /* ! ACE_HAS_GNU_REPO */
-#define ACE_HAS_GNUC_BROKEN_TEMPLATE_INLINE_FUNCTIONS
+
 #define ACE_TEMPLATES_REQUIRE_SOURCE
 
 #include "ace/post.h"
