Index: Transport.h
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/Transport.h,v
retrieving revision 1.12
diff -u -r1.12 Transport.h
--- Transport.h	2001/04/05 17:28:13	1.12
+++ Transport.h	2001/04/13 19:31:39
@@ -52,9 +52,9 @@
   int refcount (void) const;
 
 protected:
-  TAO_Synch_Refcountable (int refcount);
+  TAO_Synch_Refcountable (ACE_Lock *lock, int refcount);
 
-  TAO_SYNCH_MUTEX mutex_;
+  ACE_Lock *refcount_lock_;
 };
 
 
Index: Transport.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/Transport.cpp,v
retrieving revision 1.15
diff -u -r1.15 Transport.cpp
--- Transport.cpp	2001/04/12 22:00:30	1.15
+++ Transport.cpp	2001/04/13 19:31:41
@@ -20,26 +20,28 @@
 
 ACE_RCSID(tao, Transport, "$Id: Transport.cpp,v 1.15 2001/04/12 22:00:30 bala Exp $")
 
-TAO_Synch_Refcountable::TAO_Synch_Refcountable (int refcount)
+TAO_Synch_Refcountable::TAO_Synch_Refcountable (ACE_Lock *lock, int refcount)
   : ACE_Refcountable (refcount)
+  , refcount_lock_ (lock)
 {
 }
 
 TAO_Synch_Refcountable::~TAO_Synch_Refcountable (void)
 {
+  delete this->refcount_lock_;
 }
 
 int
 TAO_Synch_Refcountable::increment (void)
 {
-  ACE_GUARD_RETURN (ACE_SYNCH_MUTEX, ace_mon, this->mutex_, 0);
+  ACE_GUARD_RETURN (ACE_Lock, ace_mon, *this->refcount_lock_, 0);
   return ACE_Refcountable::increment ();
 }
 
 int
 TAO_Synch_Refcountable::decrement (void)
 {
-  ACE_GUARD_RETURN (ACE_SYNCH_MUTEX, ace_mon, this->mutex_, 0);
+  ACE_GUARD_RETURN (ACE_Lock, ace_mon, *this->refcount_lock_, 0);
   return ACE_Refcountable::decrement ();
 }
 
@@ -52,7 +54,7 @@
 // Constructor.
 TAO_Transport::TAO_Transport (CORBA::ULong tag,
                               TAO_ORB_Core *orb_core)
-  : TAO_Synch_Refcountable (1)
+  : TAO_Synch_Refcountable (orb_core->resource_factory ()->create_cached_connection_lock (), 1)
   , tag_ (tag)
   , orb_core_ (orb_core)
   , cache_map_entry_ (0)
