Index: ChangeLog
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ChangeLog,v
retrieving revision 4.10053
diff -u -r4.10053 ChangeLog
--- ChangeLog	22 Mar 2006 22:43:57 -0000	4.10053
+++ ChangeLog	22 Mar 2006 22:48:02 -0000
@@ -1,3 +1,13 @@
+Wed Mar 22 14:44:51 UTC 2006  Ossama Othman  <ossama_othman at symantec dot com>
+
+        * ace/SSL/SSL_Context.cpp:
+        * ace/SSL/SSL_Context.h:
+
+          Addressed access to locks that have already been destroyed by
+          reintroducing some of the original static variable semantics.
+          Some, but not all, of the exception-safety fixes that introduced
+          the order of destruction problem were retained.
+
 Wed Mar 22 18:42:36 UTC 2006  Ossama Othman  <ossama_othman at symantec dot com>
 
         From Russell Mora <russell_mora at symantec dot com>
Index: ace/SSL/SSL_Context.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/SSL/SSL_Context.cpp,v
retrieving revision 1.55
diff -u -r1.55 SSL_Context.cpp
--- ace/SSL/SSL_Context.cpp	4 Dec 2005 14:52:28 -0000	1.55
+++ ace/SSL/SSL_Context.cpp	22 Mar 2006 22:48:02 -0000
@@ -17,8 +17,6 @@
 
 #ifdef ACE_HAS_THREADS
 # include "ace/Thread_Mutex.h"
-# include "ace/Synch_Traits.h"
-# include "ace/Auto_Ptr.h"
 # include "ace/OS_NS_Thread.h"
 #endif  /* ACE_HAS_THREADS */
 
@@ -28,7 +26,7 @@
 
 ACE_RCSID (ACE_SSL,
            SSL_Context,
-           "$Id: SSL_Context.cpp,v 1.55 2005/12/04 14:52:28 ossama Exp $")
+           "$Id: SSL_Context.cpp,v 1.8 2006/03/22 22:35:22 othman Exp $")
 
 
 namespace
@@ -43,7 +41,7 @@
 #ifdef ACE_HAS_THREADS
   /// Array of mutexes used internally by OpenSSL when the SSL
   /// application is multithreaded.
-  ACE_Auto_Basic_Array_Ptr<ACE_SYNCH_MUTEX> ssl_locks;
+  ACE_SSL_Context::lock_type * ssl_locks = 0;
 
   // @@ This should also be managed by a singleton.
 #endif
@@ -112,6 +110,10 @@
 
 ACE_BEGIN_VERSIONED_NAMESPACE_DECL
 
+#ifdef ACE_HAS_THREADS
+ACE_SSL_Context::lock_type * ACE_SSL_Context::locks_ = 0;
+#endif  /* ACE_HAS_THREADS */
+
 ACE_SSL_Context::ACE_SSL_Context (void)
   : context_ (0),
     mode_ (-1),
@@ -152,7 +154,8 @@
 #ifdef ACE_HAS_THREADS
       int const num_locks = ::CRYPTO_num_locks ();
 
-      ssl_locks.reset (new ACE_SYNCH_MUTEX[num_locks]);
+      this->locks_ = new lock_type[num_locks];
+      ssl_locks    = this->locks_;
 
 # if !defined (WIN32)
       // This call isn't necessary on some platforms.  See the CRYPTO
@@ -215,6 +218,9 @@
       // cleaned up.
 #ifdef ACE_HAS_THREADS
       ::CRYPTO_set_locking_callback (0);
+
+      delete [] ACE_SSL_Context::locks_;
+      ACE_SSL_Context::locks_ = 0;
 #endif  /* ACE_HAS_THREADS */
     }
 }
Index: ace/SSL/SSL_Context.h
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/SSL/SSL_Context.h,v
retrieving revision 1.25
diff -u -r1.25 SSL_Context.h
--- ace/SSL/SSL_Context.h	4 Dec 2005 01:57:03 -0000	1.25
+++ ace/SSL/SSL_Context.h	22 Mar 2006 22:48:02 -0000
@@ -4,7 +4,7 @@
 /**
  *  @file    SSL_Context.h
  *
- *  $Id: SSL_Context.h,v 1.25 2005/12/04 01:57:03 ossama Exp $
+ *  $Id: SSL_Context.h,v 1.3 2006/03/22 22:30:10 othman Exp $
  *
  *  @author Carlos O'Ryan <coryan@ece.uci.edu>
  *  @author Ossama Othman <ossama@dre.vanderbilt.edu>
@@ -25,6 +25,10 @@
 
 #include "ace/SString.h"
 
+#ifdef ACE_HAS_THREADS
+# include "ace/Synch_Traits.h"
+#endif  /* ACE_HAS_THREADS */
+
 #include <openssl/ssl.h>
 
 
@@ -72,6 +76,10 @@
 {
 public:
 
+#ifdef ACE_HAS_THREADS
+  typedef ACE_SYNCH_MUTEX lock_type;
+#endif  /* ACE_HAS_THREADS */
+
   enum {
     INVALID_METHOD = -1,
     SSLv2_client = 1,
@@ -237,18 +245,18 @@
   //@{
   /// Seed the underlying random number generator.  This value should
   /// have at least 128 bits of entropy.
-  int random_seed (const char * seed);
+  static int random_seed (const char * seed);
 
   /// Set the Entropy Gathering Daemon (EGD) UNIX domain socket file to
   /// read random seed values from.
-  int egd_file (const char * socket_file);
+  static int egd_file (const char * socket_file);
 
   /**
    * Set the file that contains the random seed value state, and the
    * amount of bytes to read.  "-1" bytes causes the entire file to be
    * read.
    */
-  int seed_file (const char * seed_file, long bytes = -1);
+  static int seed_file (const char * seed_file, long bytes = -1);
   //@}
 
   /// Print SSL error corresponding to the given error code.
@@ -312,6 +320,12 @@
   /// count of successful CA load attempts
   int have_ca_;
 
+#ifdef ACE_HAS_THREADS
+  /// Array of mutexes used internally by OpenSSL when the SSL
+  /// application is multithreaded.
+  static lock_type * locks_;
+#endif  /* ACE_HAS_THREADS */
+
 };
 
 ACE_END_VERSIONED_NAMESPACE_DECL
