--- C:/DOCUME~1/tv2745/LOCALS~1/Temp/VarOut_T.in-revBASE.svn001.tmp.inl	Thu Nov 19 23:12:07 2009
+++ C:/Middleware/ACE_Devo/TAO/tao/VarOut_T.inl	Thu Nov 19 22:54:35 2009
@@ -48,13 +48,6 @@
 
 template<typename T>
 ACE_INLINE
-TAO_Var_Base_T<T>::operator T & ()
-{
-  return *this->ptr_;
-}
-
-template<typename T>
-ACE_INLINE
 TAO_Var_Base_T<T>::operator T & () const
 {
   return *this->ptr_;
@@ -121,6 +114,19 @@
   return *this;
 }
 
+template<typename T>
+ACE_INLINE
+TAO_Fixed_Var_T<T>::operator T & ()
+{
+  // Use plain new, using the ACE_NEW macros will mean we dereference a
+  // nil pointer and crash which is more bad then plain new which could
+  // lead to a bad_alloc exception
+  if (this->ptr_ == 0)
+    this->ptr_ = new T;
+
+  return *this->ptr_;
+}
+
 // Mapping for fixed size.
 template<typename T>
 ACE_INLINE
@@ -181,6 +187,13 @@
   return this->ptr_;
 }
 
+template<typename T>
+ACE_INLINE
+TAO_Var_Var_T<T>::operator T & ()
+{
+  return *this->ptr_;
+}
+
 // Mapping for variable size.
 template<typename T>
 ACE_INLINE
