*** x:/ACE_wrappers/ace/Log_Record.cpp	Tue Feb  8 17:51:54 2005
--- Log_Record.cpp	Mon Apr  4 12:26:04 2005
***************
*** 20,30 ****
  
  ACE_ALLOC_HOOK_DEFINE(ACE_Log_Record)
  
  const ACE_TCHAR *ACE_Log_Record::priority_names_[] =
  {
-   ACE_LIB_TEXT ("<none>"),
    ACE_LIB_TEXT ("LM_SHUTDOWN"),
    ACE_LIB_TEXT ("LM_TRACE"),
    ACE_LIB_TEXT ("LM_DEBUG"),
    ACE_LIB_TEXT ("LM_INFO"),
    ACE_LIB_TEXT ("LM_NOTICE"),
--- 20,29 ----
