Index: Leader_Follower_Flushing_Strategy.cpp
===================================================================
--- Leader_Follower_Flushing_Strategy.cpp	(revision 86643)
+++ Leader_Follower_Flushing_Strategy.cpp	(working copy)
@@ -47,7 +47,8 @@
     {
       TAO_ORB_Core * const orb_core = transport->orb_core ();
 
-      while (!transport->queue_is_empty ())
+      while (!transport->queue_is_empty () &&
+             orb_core->reactor ()->work_pending ())
         {
           // In case max_wait_time==0 we cannot simply run the orb because
           // in multi-threaded applications it can easily happen that
@@ -55,6 +56,8 @@
           // transport we're coping with here and this thread will block.
           // Instead we do run for a small amount of time and then recheck
           // the queue.
+          // Also we have to check the work_pending status for the reason
+          // stated in bugzilla 3697.
           if (max_wait_time == 0)
             {
               ACE_Errno_Guard eguard (errno);
