Index: config-win32-common.h
===================================================================
--- config-win32-common.h	(revision 82282)
+++ config-win32-common.h	(working copy)
@@ -497,6 +497,9 @@
 #  else
 #    pragma comment(lib, "ws2_32.lib")
 #    pragma comment(lib, "mswsock.lib")
+#    if defined (ACE_HAS_IPV6)
+#      pragma comment(lib, "Iphlpapi.lib")
+#    endif
 #  endif /* ACE_HAS_WINCE */
 # endif /* _MSC_VER */
 
