Index: PICurrent.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/PI/PICurrent.cpp,v
retrieving revision 1.4
diff -u -r1.4 PICurrent.cpp
--- PICurrent.cpp	10 Mar 2006 07:19:12 -0000	1.4
+++ PICurrent.cpp	10 May 2006 17:50:43 -0000
@@ -18,6 +18,8 @@
 #include "tao/TAO_Server_Request.h"
 #include "tao/SystemException.h"
 
+#include "ace/Auto_Ptr.h"
+
 TAO_BEGIN_VERSIONED_NAMESPACE_DECL
 
 TAO::PICurrent::PICurrent (TAO_ORB_Core &orb_core)
@@ -63,8 +65,19 @@
   PICurrent_Impl *impl = this->tsc ();
 
   if (impl == 0)
-    ACE_THROW (CORBA::BAD_INV_ORDER (CORBA::OMGVMCID | 14,
-                                     CORBA::COMPLETED_NO));
+    {
+      // Populate TSS with the underlying PICurrent implementation.
+      ACE_NEW_THROW_EX (impl,
+                        TAO::PICurrent_Impl,
+                        CORBA::NO_MEMORY ());
+      ACE_CHECK;
+
+      auto_ptr<TAO::PICurrent_Impl> safe_impl (impl);
+
+      this->orb_core_.set_tss_resource (this->tss_slot_, impl);
+
+      (void) safe_impl.release ();
+    }
 
   impl->set_slot (identifier, data ACE_ENV_ARG_PARAMETER);
   ACE_CHECK;
@@ -104,18 +117,11 @@
 
   if (this->tsc () == 0 && tss_slot_ == 0)
     {
-      TAO::PICurrent_Impl *impl = 0;
-      ACE_NEW_RETURN (impl,
-                      TAO::PICurrent_Impl,
-                      0);
-
       const int result = this->orb_core_.add_tss_cleanup_func (0,
                                                                tss_slot_);
 
       if (result != 0)
         return result;
-
-      this->orb_core_.set_tss_resource (tss_slot_, impl);
     }
 
   return 0;
