Index: IIOP_Connection_Handler.cpp
===================================================================
--- IIOP_Connection_Handler.cpp	(revision 83511)
+++ IIOP_Connection_Handler.cpp	(working copy)
@@ -480,7 +480,7 @@
     IIOP::ListenPointList &listen_list)
 {
   // Get the size of the list
-  CORBA::ULong const len = listen_list.length ();
+  CORBA::ULong len = listen_list.length ();
 
   if (TAO_debug_level > 0 && len == 0)
     {
@@ -490,6 +490,24 @@
                   ACE_TEXT("Received list of size 0, check client config.\n")));
     }
 
+  // @@ We can only handle a single endpoint for now because the
+  // transport is recached for each endpoint, loosing older
+  // information. This means that when more than one listen_point is
+  // sent, the cache will end up only being associated with the last
+  // address.  This poses a problem at invocation time because the
+  // default endpoint selector steps through the profiles/endpoints
+  // sequentially and will try and possibly succeed in connecting to
+  // one of the earlier endpoints.  My assumption is that when this
+  // method is called, it is before the first attempt to connect to
+  // any object in the target process, thus the first listen point in
+  // the list will correspond with the first profile and thus that
+  // should match the cache.
+  //
+  // Probably what needs to be done to make this model work well is to
+  // allow the transport cache to have multiple keys reference the
+  // same transport so that rather than recaching, we simply add a
+  // key.
+  len = 1;
   for (CORBA::ULong i = 0; i < len; ++i)
     {
       IIOP::ListenPoint listen_point = listen_list[i];
Index: IIOP_Transport.cpp
===================================================================
--- IIOP_Transport.cpp	(revision 83511)
+++ IIOP_Transport.cpp	(working copy)
@@ -376,25 +376,14 @@
   // Get the array of endpoints serviced by TAO_IIOP_Acceptor
   const ACE_INET_Addr *endpoint_addr =
     iiop_acceptor->endpoints ();
+  size_t count = iiop_acceptor->endpoint_count ();
 
-  // Get the endpoint count
-  size_t const count =
-    iiop_acceptor->endpoint_count ();
+  ACE_DEBUG ((LM_DEBUG, "\nget_listen_point, count = %d\n\n", count));
 
-  // Get the local address of the connection
-  ACE_INET_Addr local_addr;
-
-  if (this->connection_handler_->peer ().get_local_addr (local_addr) == -1)
+  for (size_t index = 0; index < count; index++)
     {
-      ACE_ERROR_RETURN ((LM_ERROR,
-                         ACE_TEXT ("TAO (%P|%t) - IIOP_Transport::get_listen_point, ")
-                         ACE_TEXT ("could not resolve local host address\n")),
-                        -1);
-    }
-
-  // Note: Looks like there is no point in sending the list of
-  // endpoints on interfaces on which this connection has not
-  // been established. If this is wrong, please correct me.
+      // Get the local address of the connection
+      ACE_INET_Addr local_addr = endpoint_addr[index];
   CORBA::String_var local_interface;
 
   // Get the hostname for the local address
@@ -407,6 +396,7 @@
                          ACE_TEXT ("could not resolve local host name\n")),
                         -1);
     }
+
 #if defined (ACE_HAS_IPV6)
   // If this is an IPv6 decimal linklocal address containing a scopeid than
   // remove the scopeid from the information being sent.
@@ -419,16 +409,6 @@
     }
 #endif /* ACE_HAS_IPV6 */
 
-  for (size_t index = 0;
-       index < count;
-       ++index)
-    {
-      // Make sure port numbers are equal so the following comparison
-      // only concerns the IP(v4/v6) address.
-      local_addr.set_port_number (endpoint_addr[index].get_port_number ());
-
-      if (local_addr == endpoint_addr[index])
-        {
           // Get the count of the number of elements
           const CORBA::ULong len = listen_point_list.length ();
 
@@ -439,7 +419,7 @@
           // same interface
           IIOP::ListenPoint & point = listen_point_list[len];
           point.host = CORBA::string_dup (local_interface.in ());
-          point.port = endpoint_addr[index].get_port_number ();
+      point.port = local_addr.get_port_number ();
 
           if (TAO_debug_level >= 5)
           {
@@ -449,10 +429,7 @@
                         point.host.in (),
                         point.port));
           }
-
-        }
     }
-
   return 1;
 }
 //@@ TAO_TRANSPORT_SPL_COPY_HOOK_END
