Index: Stub.cpp
===================================================================
--- Stub.cpp	(revision 81318)
+++ Stub.cpp	(working copy)
@@ -90,6 +90,9 @@
   if (this->forward_profiles_)
     reset_profiles ();
 
+  // reset_profiles doesn't delete forward_profiles_perm_.
+  delete this->forward_profiles_perm_;
+
   if (this->profile_in_use_ != 0)
     {
       // decrease reference count on profile
Index: Invocation_Base.h
===================================================================
--- Invocation_Base.h	(revision 81318)
+++ Invocation_Base.h	(working copy)
@@ -88,12 +88,6 @@
     /// Return the forwarded object location by loosing ownership.
     CORBA::Object_ptr steal_forwarded_reference (void);
 
-    /// Did the invocation got forwarded to a new location?
-    bool is_forwarded (void) const;
-
-    /// Mutator to set the reply status of the invocation.
-    void reply_received (Invocation_Status s);
-
     /// Return the effective target of the invocation.
     /**
      * Please see the PortableInterceptor specification in the CORBA
@@ -107,6 +101,10 @@
     /// Does this invocation return a response?
     CORBA::Boolean response_expected (void) const;
 
+    /// Accessor and mutator of reply_status of the invocation.
+    GIOP::ReplyStatusType reply_status (void) const;
+    void reply_status (GIOP::ReplyStatusType s);
+
     /// The operaton details of the invocation
     TAO_Operation_Details &operation_details (void);
     //@}
@@ -138,12 +136,12 @@
     /// Forwarded object reference.
     CORBA::Object_var forwarded_to_;
 
-    /// If we get nil forward reference we need more reliable indication of forwarding.
-    bool is_forwarded_;
-
     /// Is response expected?
     bool response_expected_;
 
+    /// A GIOP reply status of the invocation.
+    GIOP::ReplyStatusType reply_status_;
+
   private:
 
     Invocation_Base (const Invocation_Base&);
@@ -189,8 +187,10 @@
     /// Change the exception status.
     void exception (CORBA::Exception *exception);
 
-    /// Invoke status
+    /// Invocation status.
     TAO::Invocation_Status invoke_status (void) const;
+    /// Mutator to set the invocation status.
+    void invoke_status (Invocation_Status s);
 
     PortableInterceptor::ReplyStatus pi_reply_status (void) const;
 
Index: LocateRequest_Invocation_Adapter.cpp
===================================================================
--- LocateRequest_Invocation_Adapter.cpp	(revision 81318)
+++ LocateRequest_Invocation_Adapter.cpp	(working copy)
@@ -80,16 +80,15 @@
             TAO::LocateRequest_Invocation synch (this->target_, resolver, op);
 
             s = synch.invoke (max_wait_time);
-            if (s == TAO_INVOKE_RESTART && synch.is_forwarded ())
+            if (s == TAO_INVOKE_RESTART &&
+                (synch.reply_status () == GIOP::LOCATION_FORWARD ||
+                 synch.reply_status () == GIOP::LOCATION_FORWARD_PERM))
               {
+                CORBA::Boolean const is_permanent_forward =
+                  (synch.reply_status () == GIOP::LOCATION_FORWARD_PERM);
+
                 effective_target = synch.steal_forwarded_reference ();
 
-#if TAO_HAS_INTERCEPTORS == 1
-                CORBA::Boolean const is_permanent_forward =
-                  (synch.pi_reply_status() == GIOP::LOCATION_FORWARD_PERM);
-#else
-                CORBA::Boolean const is_permanent_forward = false;
-#endif
                 this->object_forwarded (effective_target,
                                         resolver.stub (),
                                         is_permanent_forward);
Index: Invocation_Base.cpp
===================================================================
--- Invocation_Base.cpp	(revision 81318)
+++ Invocation_Base.cpp	(working copy)
@@ -41,8 +41,8 @@
                                     bool TAO_INTERCEPTOR (request_is_remote))
     : details_ (details)
     , forwarded_to_ (0)
-    , is_forwarded_ (false)
     , response_expected_ (response_expected)
+    , reply_status_ (GIOP::NO_EXCEPTION)
     , otarget_ (ot)
     , target_ (t)
     , orb_core_ (stub->orb_core ())
@@ -59,17 +59,9 @@
 
   Invocation_Base::~Invocation_Base (void)
   {
-#if TAO_HAS_INTERCEPTORS == 1
-    adapter_ = 0;
-#endif /* TAO_HAS_INTERCEPTORS == 1 */
+    TAO_INTERCEPTOR (adapter_ = 0);
   }
 
-  void
-  Invocation_Base::reply_received (Invocation_Status TAO_INTERCEPTOR (s))
-  {
-    TAO_INTERCEPTOR (invoke_status_ = s);
-  }
-
   TAO_Service_Context &
   Invocation_Base::request_service_context (void)
   {
@@ -105,7 +97,7 @@
             throw;
           }
 
-        if (this->is_forwarded_)
+        if (this->reply_status_ == GIOP::LOCATION_FORWARD)
           return TAO_INVOKE_RESTART;
       }
 
@@ -133,11 +125,7 @@
             throw;
           }
 
-        PortableInterceptor::ReplyStatus const status =
-          this->adapter_->pi_reply_status (*this);
-
-        if (status == PortableInterceptor::LOCATION_FORWARD ||
-            status == PortableInterceptor::TRANSPORT_RETRY)
+        if (this->reply_status_ == GIOP::LOCATION_FORWARD)
           return TAO_INVOKE_RESTART;
       }
 
@@ -164,7 +152,7 @@
             throw;
           }
 
-        if (this->is_forwarded_)
+        if (this->reply_status_ == GIOP::LOCATION_FORWARD)
           return TAO_INVOKE_RESTART;
       }
 
@@ -183,7 +171,7 @@
       {
         this->adapter_->receive_exception (*this);
 
-        if (this->is_forwarded_)
+        if (this->reply_status_ == GIOP::LOCATION_FORWARD)
           {
             status = PortableInterceptor::LOCATION_FORWARD;
           }
@@ -223,7 +211,7 @@
       this->invoke_status_ = TAO::TAO_INVOKE_USER_EXCEPTION;
 
     this->forwarded_to_ = CORBA::Object::_nil ();
-    this->is_forwarded_ = false;
+    this->reply_status_ = GIOP::NO_EXCEPTION;
     this->caught_exception_ = exception;
   }
 
Index: Synch_Invocation.inl
===================================================================
--- Synch_Invocation.inl	(revision 81318)
+++ Synch_Invocation.inl	(working copy)
@@ -17,7 +17,9 @@
   ACE_INLINE
   Reply_Guard::~Reply_Guard (void)
   {
-    this->invocation_->reply_received (this->status_);
+#if TAO_HAS_INTERCEPTORS == 1
+    this->invocation_->invoke_status (this->status_);
+#endif  /* TAO_HAS_INTERCEPTORS == 1 */
   }
 
   ACE_INLINE void
Index: Collocated_Invocation.cpp
===================================================================
--- Collocated_Invocation.cpp	(revision 81318)
+++ Collocated_Invocation.cpp	(working copy)
@@ -76,36 +76,46 @@
                                   request,
                                   this->forwarded_to_.out ());
 
-            this->is_forwarded_ = request.is_forwarded ();
+            if (request.is_forwarded ())
+              {
+                this->reply_status_ = GIOP::LOCATION_FORWARD;
+              }
           }
         else
           {
+            bool is_forwarded = false;
+
             cpb->dispatch (this->effective_target (),
                            this->forwarded_to_.out (),
-                           this->is_forwarded_,
+                           is_forwarded,
                            this->details_.args (),
                            this->details_.args_num (),
                            this->details_.opname (),
                            this->details_.opname_len (),
                            strat);
+
+            if (is_forwarded)
+              {
+                this->reply_status_ = GIOP::LOCATION_FORWARD;
+              }
           }
 
         // Invocation completed succesfully
         s = TAO_INVOKE_SUCCESS;
 
 #if TAO_HAS_INTERCEPTORS == 1
-        if (this->is_forwarded_ ||
+        if (this->reply_status_ == GIOP::LOCATION_FORWARD ||
             this->response_expected_ == false)
           {
-            if (this->is_forwarded_)
-              this->reply_received (TAO_INVOKE_RESTART);
+            if (this->reply_status_ == GIOP::LOCATION_FORWARD)
+              this->invoke_status (TAO_INVOKE_RESTART);
 
             s = this->receive_other_interception ();
           }
         // NOTE: Any other condition that needs handling?
         else if (this->response_expected ())
           {
-            this->reply_received (TAO_INVOKE_SUCCESS);
+            this->invoke_status (TAO_INVOKE_SUCCESS);
 
             s = this->receive_reply_interception ();
           }
@@ -175,7 +185,7 @@
       }
 #endif  /* TAO_HAS_INTERCEPTORS == 1 */
 
-    if (this->is_forwarded_)
+    if (this->reply_status_ == GIOP::LOCATION_FORWARD)
       s =  TAO_INVOKE_RESTART;
 
     return s;
Index: Synch_Invocation.h
===================================================================
--- Synch_Invocation.h	(revision 81318)
+++ Synch_Invocation.h	(working copy)
@@ -169,8 +169,8 @@
 
     Reply_Guard (Invocation_Base *s, Invocation_Status is);
 
-    /// The destructor calls Invocation_Base::reply_received with the
-    /// right reply status, which is useful for PI's.
+    /// The destructor calls Invocation_Base::invoke_status with the
+    /// right invoke status, which is useful for PI's.
     ~Reply_Guard (void);
 
     /// Mutator to set the invocation status.
Index: DynamicInterface/DII_Invocation_Adapter.cpp
===================================================================
--- DynamicInterface/DII_Invocation_Adapter.cpp	(revision 81318)
+++ DynamicInterface/DII_Invocation_Adapter.cpp	(working copy)
@@ -136,16 +136,15 @@
 
     Invocation_Status status = synch.remote_invocation (max_wait_time);
 
-    if (status == TAO_INVOKE_RESTART && synch.is_forwarded ())
+    if (status == TAO_INVOKE_RESTART &&                                                                                                                    
+        (synch.reply_status () == GIOP::LOCATION_FORWARD ||                                                                                              
+         synch.reply_status () == GIOP::LOCATION_FORWARD_PERM))
       {
+        CORBA::Boolean const permanent_forward =
+          (synch.reply_status () == GIOP::LOCATION_FORWARD_PERM);
+
         effective_target = synch.steal_forwarded_reference ();
 
-#if TAO_HAS_INTERCEPTORS == 1
-        CORBA::Boolean const permanent_forward =
-          (synch.pi_reply_status() == GIOP::LOCATION_FORWARD_PERM);
-#else
-        CORBA::Boolean const permanent_forward = false;
-#endif
         this->object_forwarded (effective_target,
                                 r.stub (),
                                 permanent_forward);
Index: Messaging/Asynch_Invocation_Adapter.cpp
===================================================================
--- Messaging/Asynch_Invocation_Adapter.cpp	(revision 81318)
+++ Messaging/Asynch_Invocation_Adapter.cpp	(working copy)
@@ -192,17 +192,14 @@
     Invocation_Status const s = asynch.remote_invocation (max_wait_time);
 
     if (s == TAO_INVOKE_RESTART &&
-        asynch.is_forwarded ())
+        (asynch.reply_status () == GIOP::LOCATION_FORWARD ||
+         asynch.reply_status () == GIOP::LOCATION_FORWARD_PERM))
       {
-        effective_target = asynch.steal_forwarded_reference ();
-
-#if TAO_HAS_INTERCEPTORS == 1
         CORBA::Boolean const permanent_forward =
-          (asynch.pi_reply_status() == GIOP::LOCATION_FORWARD_PERM);
-#else
-        CORBA::Boolean const permanent_forward = false;
-#endif
+          (asynch.reply_status () == GIOP::LOCATION_FORWARD_PERM);
 
+        effective_target = asynch.steal_forwarded_reference ();
+
         this->object_forwarded (effective_target, r.stub (), permanent_forward);
       }
 
Index: Invocation_Base.inl
===================================================================
--- Invocation_Base.inl	(revision 81318)
+++ Invocation_Base.inl	(working copy)
@@ -22,7 +22,7 @@
   Invocation_Base::forwarded_reference (CORBA::Object_ptr o)
   {
     this->forwarded_to_ = CORBA::Object::_duplicate (o);
-    this->is_forwarded_ = true;
+    this->reply_status_ = GIOP::LOCATION_FORWARD;
   }
 
   ACE_INLINE CORBA::Object_ptr
@@ -34,22 +34,28 @@
   ACE_INLINE CORBA::Object_ptr
   Invocation_Base::steal_forwarded_reference (void)
   {
-    this->is_forwarded_ = false;
+    this->reply_status_ = GIOP::NO_EXCEPTION;
     return this->forwarded_to_._retn ();
   }
 
-  ACE_INLINE bool
-  Invocation_Base::is_forwarded (void) const
-  {
-    return this->is_forwarded_;
-  }
-
   ACE_INLINE CORBA::Boolean
   Invocation_Base::response_expected (void) const
   {
     return this->response_expected_;
   }
 
+  ACE_INLINE GIOP::ReplyStatusType
+  Invocation_Base::reply_status (void) const
+  {
+    return this->reply_status_;
+  }
+
+  ACE_INLINE void
+  Invocation_Base::reply_status (GIOP::ReplyStatusType s)
+  {
+    this->reply_status_ = s;
+  }
+
   ACE_INLINE CORBA::Object_ptr
   Invocation_Base::target (void) const
   {
@@ -87,6 +93,12 @@
     return this->invoke_status_;
   }
 
+  ACE_INLINE void
+  Invocation_Base::invoke_status (Invocation_Status s)
+  {
+    this->invoke_status_ = s;
+  }
+
   ACE_INLINE bool
   Invocation_Base::is_remote_request (void) const
   {
Index: PI/ClientRequestInterceptor_Adapter_Impl.cpp
===================================================================
--- PI/ClientRequestInterceptor_Adapter_Impl.cpp	(revision 81318)
+++ PI/ClientRequestInterceptor_Adapter_Impl.cpp	(working copy)
@@ -276,7 +276,8 @@
         reply_status = PortableInterceptor::SUCCESSFUL;
         break;
       case TAO::TAO_INVOKE_RESTART:
-        if (invocation_base.is_forwarded ())
+        if (invocation_base.reply_status () == GIOP::LOCATION_FORWARD ||
+            invocation_base.reply_status () == GIOP::LOCATION_FORWARD_PERM)
           reply_status = PortableInterceptor::LOCATION_FORWARD;
         else
           reply_status = PortableInterceptor::TRANSPORT_RETRY;
Index: Invocation_Adapter.cpp
===================================================================
--- Invocation_Adapter.cpp	(revision 81318)
+++ Invocation_Adapter.cpp	(working copy)
@@ -166,17 +166,15 @@
 
     status = coll_inv.invoke (this->cpb_, strat);
 
-    if (status == TAO_INVOKE_RESTART && coll_inv.is_forwarded ())
+    if (status == TAO_INVOKE_RESTART &&
+        (coll_inv.reply_status () == GIOP::LOCATION_FORWARD ||
+         coll_inv.reply_status () == GIOP::LOCATION_FORWARD_PERM))
       {
-        effective_target = coll_inv.steal_forwarded_reference ();
-
-#if TAO_HAS_INTERCEPTORS == 1
         CORBA::Boolean const is_permanent_forward =
-          (coll_inv.pi_reply_status() == GIOP::LOCATION_FORWARD_PERM);
-#else
-        CORBA::Boolean const is_permanent_forward = false;
-#endif
+          (coll_inv.reply_status () == GIOP::LOCATION_FORWARD_PERM);
 
+        effective_target = coll_inv.steal_forwarded_reference ();
+
         this->object_forwarded (effective_target, stub, is_permanent_forward);
       }
 
@@ -302,17 +300,14 @@
     Invocation_Status const status = synch.remote_twoway (max_wait_time);
 
     if (status == TAO_INVOKE_RESTART &&
-        synch.is_forwarded ())
+        (synch.reply_status () == GIOP::LOCATION_FORWARD ||
+         synch.reply_status () == GIOP::LOCATION_FORWARD_PERM))
       {
-        effective_target = synch.steal_forwarded_reference ();
-
-#if TAO_HAS_INTERCEPTORS == 1
         CORBA::Boolean const is_permanent_forward =
-          (synch.pi_reply_status() == GIOP::LOCATION_FORWARD_PERM);
-#else
-        CORBA::Boolean const is_permanent_forward = false;
-#endif
+          (synch.reply_status () == GIOP::LOCATION_FORWARD_PERM);
 
+        effective_target = synch.steal_forwarded_reference ();
+
         this->object_forwarded (effective_target,
                                 r.stub (),
                                 is_permanent_forward);
@@ -331,16 +326,15 @@
 
     Invocation_Status const s = synch.remote_oneway (max_wait_time);
 
-    if (s == TAO_INVOKE_RESTART && synch.is_forwarded ())
+    if (s == TAO_INVOKE_RESTART &&
+        (synch.reply_status () == GIOP::LOCATION_FORWARD ||
+         synch.reply_status () == GIOP::LOCATION_FORWARD_PERM))
       {
+        CORBA::Boolean const is_permanent_forward =
+          (synch.reply_status () == GIOP::LOCATION_FORWARD_PERM);
+
         effective_target = synch.steal_forwarded_reference ();
 
-#if TAO_HAS_INTERCEPTORS == 1
-        CORBA::Boolean const is_permanent_forward =
-          (synch.pi_reply_status() == GIOP::LOCATION_FORWARD_PERM);
-#else
-        CORBA::Boolean const is_permanent_forward = false;
-#endif
         this->object_forwarded (effective_target,
                                 r.stub (),
                                 is_permanent_forward);
Index: Synch_Invocation.cpp
===================================================================
--- Synch_Invocation.cpp	(revision 81318)
+++ Synch_Invocation.cpp	(working copy)
@@ -381,6 +381,9 @@
 
                    throw ::CORBA::INTERNAL (0, CORBA::COMPLETED_NO);
                 }
+
+              // This is the only place where we set in permanent forward state.
+              this->reply_status (GIOP::LOCATION_FORWARD_PERM);
             }
 
           return s;
