Index: ace/Select_Reactor_T.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/Select_Reactor_T.cpp,v
retrieving revision 4.95
diff -u -r4.95 Select_Reactor_T.cpp
--- ace/Select_Reactor_T.cpp	9 May 2006 07:25:08 -0000	4.95
+++ ace/Select_Reactor_T.cpp	25 May 2006 22:29:50 -0000
@@ -1280,6 +1280,8 @@
 
   do
     {
+      int initial_count = active_handle_count;
+
       // Note that we keep track of changes to our state.  If any of
       // the dispatch_*() methods below return -1 it means that the
       // <wait_set_> state has changed as the result of an
@@ -1354,7 +1356,7 @@
 
       // if state changed, we need to re-eval active_handle_count,
       // so we will not end with an endless loop
-      if (this->state_changed_)
+      if (this->state_changed_ || initial_count == active_handle_count)
       {
           active_handle_count = this->dispatch_set_.rd_mask_.num_set ()
               + this->dispatch_set_.wr_mask_.num_set ()
