Index: SOCK_Dgram.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/SOCK_Dgram.cpp,v
retrieving revision 4.47
diff -u -u -r4.47 SOCK_Dgram.cpp
--- SOCK_Dgram.cpp	2001/12/03 21:30:10	4.47
+++ SOCK_Dgram.cpp	2002/04/01 20:23:37
@@ -101,11 +101,13 @@
   ACE_TRACE ("ACE_SOCK_Dgram::shared_open");
   int error = 0;
 
-  if (local == ACE_Addr::sap_any
-      && protocol_family == PF_INET)
+  if (local == ACE_Addr::sap_any)
     {
-      if (ACE::bind_port (this->get_handle ()) == -1)
-        error = 1;
+      if (protocol_family == PF_INET)
+        {
+          if (ACE::bind_port (this->get_handle ()) == -1)
+            error = 1;
+        }
     }
   else if (ACE_OS::bind (this->get_handle (),
                          ACE_reinterpret_cast (sockaddr *,
@@ -480,7 +482,7 @@
     htonl (interface_addr.get_ip_address ());
 #else
   ifreq if_address;
-  
+
 #if defined (ACE_PSOS)
   // Look up the interface by number, not name.
   if_address.ifr_ifno = ACE_OS::atoi (option_value);
