Index: ace/Service_Config.cpp
===================================================================
--- ace/Service_Config.cpp	(revision 79366)
+++ ace/Service_Config.cpp	(working copy)
@@ -47,8 +47,10 @@
 {
   if (ACE::debug ())
     ACE_DEBUG ((LM_DEBUG,
-                ACE_TEXT ("ACE (%P|%t) Service_Config_Guard:<ctor>")
-                ACE_TEXT (" - repo=%@ superceded by repo=%@\n"),
+                ACE_TEXT ("ACE (%P|%t) Service_Config_Guard:<ctor=%@>")
+                ACE_TEXT (" - config=%@ repo=%@ superceded by repo=%@\n"),
+                this,
+                this->saved_,
                 this->saved_->repo_,
                 psg->repo_));
 
@@ -63,8 +65,9 @@
 
   if (ACE::debug ())
     ACE_DEBUG ((LM_DEBUG,
-                ACE_TEXT ("ACE (%P|%t) Service_Config_Guard:<dtor>")
+                ACE_TEXT ("ACE (%P|%t) Service_Config_Guard:<dtor=%@>")
                 ACE_TEXT (" - new repo=%@\n"),
+                this,
                 this->saved_->repo_));
 }
 
@@ -182,7 +185,11 @@
   // The base class open_i increments this and we are
   // forwarding to it, so we don't have to increment here.
   if (this->is_opened_ != 0)
-    return 0;
+    return ACE_Service_Gestalt::open_i (program_name,
+                                        logger_key,
+                                        ignore_static_svcs,
+                                        ignore_default_svc_conf_file,
+                                        ignore_debug_flag);
 
   // Check for things we need to do on a per-process basis and which
   // may not be safe, or wise to do an a per instance basis
Index: ace/Service_Types.cpp
===================================================================
--- ace/Service_Types.cpp	(revision 79366)
+++ ace/Service_Types.cpp	(working copy)
@@ -60,7 +60,7 @@
   ACE_TRACE ("ACE_Service_Type_Impl::fini");
   if (ACE::debug ())
     ACE_DEBUG ((LM_DEBUG,
-                ACE_TEXT ("destroying %s, flags = %d\n"),
+                ACE_TEXT ("(%P|%t) ST::fini - destroying %s, flags = %d\n"),
                 this->name_,
                 this->flags_));
 
Index: TAO/tao/ORB.cpp
===================================================================
--- TAO/tao/ORB.cpp	(revision 79365)
+++ TAO/tao/ORB.cpp	(working copy)
@@ -1086,8 +1086,59 @@
 
 // ****************************************************************
 
+namespace TAO
+{
+  bool
+  parse_orb_opt (ACE_Argv_Type_Converter& command_line,
+                 const ACE_TCHAR* orb_opt,
+                 ACE_CString& opt_arg)
+  {
+
+    if (opt_arg.length () > 0)
+      return false;
+
+    ACE_Arg_Shifter arg_shifter (command_line.get_argc (),
+                                 command_line.get_TCHAR_argv ());
+
+    size_t opt_len = ACE_OS::strlen (orb_opt);
+
+    bool found = false;
+    while (arg_shifter.is_anything_left ())
+      {
+        const ACE_TCHAR *current_arg = arg_shifter.get_current ();
+
+        if (ACE_OS::strcasecmp (current_arg,
+                                orb_opt) == 0)
+          {
+            found = true;
+            arg_shifter.consume_arg ();
+            if (arg_shifter.is_parameter_next ())
+              {
+                opt_arg =
+                  ACE_TEXT_ALWAYS_CHAR (arg_shifter.get_current ());
+                arg_shifter.consume_arg ();
+              }
+          }
+        else if (ACE_OS::strncasecmp (current_arg, orb_opt,
+                                      opt_len) == 0)
+          {
+            arg_shifter.consume_arg ();
+            // The rest of the argument is the ORB id...
+            // but we should skip an optional space...
+            if (current_arg[opt_len] == ' ')
+              opt_arg =
+                ACE_TEXT_ALWAYS_CHAR (current_arg + opt_len + 1);
+            else
+              opt_arg = ACE_TEXT_ALWAYS_CHAR (current_arg + opt_len);
+          }
+        else
+          arg_shifter.ignore_arg ();
+      }
+    return found;
+  }
+}
+
 // ORB initialization, per OMG document 98-12-01.
-
 CORBA::ORB_ptr
 CORBA::ORB_init (int &argc, char *argv[], const char *orbid)
 {
@@ -1141,74 +1192,109 @@
         CORBA::COMPLETED_NO);
     }
 
-  if (orbid_string.length () == 0)
-    {
-      ACE_Arg_Shifter arg_shifter (command_line.get_argc (),
-                                   command_line.get_TCHAR_argv ());
+  TAO::parse_orb_opt (command_line, ACE_TEXT("-ORBid"), orbid_string);
 
-      while (arg_shifter.is_anything_left ())
-        {
-          const ACE_TCHAR *current_arg = arg_shifter.get_current ();
-
-          static const ACE_TCHAR orbid_opt[] = ACE_TEXT ("-ORBid");
-          size_t orbid_len = ACE_OS::strlen (orbid_opt);
-          if (ACE_OS::strcasecmp (current_arg,
-                                  orbid_opt) == 0)
-            {
-              arg_shifter.consume_arg ();
-              if (arg_shifter.is_parameter_next ())
-                {
-                  orbid_string =
-                    ACE_TEXT_ALWAYS_CHAR (arg_shifter.get_current ());
-                  arg_shifter.consume_arg ();
-                }
-            }
-          else if (ACE_OS::strncasecmp (current_arg, orbid_opt,
-                                        orbid_len) == 0)
-            {
-              arg_shifter.consume_arg ();
-              // The rest of the argument is the ORB id...
-              // but we should skip an optional space...
-              if (current_arg[orbid_len] == ' ')
-                orbid_string =
-                  ACE_TEXT_ALWAYS_CHAR (current_arg + orbid_len + 1);
-              else
-                orbid_string = ACE_TEXT_ALWAYS_CHAR (current_arg + orbid_len);
-            }
-          else
-            arg_shifter.ignore_arg ();
-        }
-    }
-
   // Get ORB Core
   TAO_ORB_Core_Auto_Ptr oc (
     TAO::ORB_Table::instance ()->find (orbid_string.c_str ()));
 
-  // The ORB was already initialized.  Just return that one.
+    // The ORB was already initialized.  Just return that one.
   if (oc.get () != 0)
     {
       return CORBA::ORB::_duplicate (oc->orb ());
     }
-  else
+
+
+  // Determine the service object registry this ORB will use. The choises
+  // are: (a) the legacy (global); (b) its own, local, or (c) share somebody
+  // else's configuration
+
+  // By default use the process (application?) global configuration context
+  ACE_Service_Gestalt* gestalt = ACE_Service_Config::current ();
+
+  // Use this string variable to hold the config identity
+  ACE_CString orbconfig_string;
+  ACE_Auto_Ptr<ACE_Service_Gestalt> guard_gestalt(0);
+
+  if (TAO::parse_orb_opt (command_line,
+                         ACE_TEXT("-ORBGestalt"),
+                         orbconfig_string))
     {
-      // An ORB corresponding to the desired ORBid doesn't exist so create
-      // a new one.
-      TAO_ORB_Core * tmp = 0;
-      ACE_NEW_THROW_EX (tmp,
-                        TAO_ORB_Core (orbid_string.c_str ()),
-                        CORBA::NO_MEMORY (
-                          CORBA::SystemException::_tao_minor_code (0,
-                                                                   ENOMEM),
-                          CORBA::COMPLETED_NO));
+      // Need a local repo? Make one which typically should not be as
+      // big as the default repository
+      if  (ACE_OS::strcasecmp (orbconfig_string.c_str (),
+                                  ACE_TEXT("LOCAL")) == 0)
+        {
+          ACE_NEW_THROW_EX (gestalt,
+                            ACE_Service_Gestalt
+                            (ACE_Service_Gestalt::MAX_SERVICES / 4, true),
+                            CORBA::NO_MEMORY
+                            (CORBA::SystemException::_tao_minor_code (0,
+                                                                      ENOMEM),
+                             CORBA::COMPLETED_NO));
+          guard_gestalt.reset(gestalt);
+        }
+      else if (ACE_OS::strncmp (orbconfig_string.c_str (),
+                                ACE_TEXT ("ORB:"), 4) == 0)
+        {
+          // @TODO: At some point, we need to implement a lookup of an
+          // existing configuration context based on the orbid
+          // following the "ORB:" It may be the case that contexts may
+          // be initialized separate from the ORB of the same ID, in
+          // which case we would need some sort of table of contexts
+          // at which point we would either find this one or make a
+          // new one. And making a new one would require assigning it
+          // to the guard_gestalt just like the local.
+          if (TAO_debug_level > 0)
+            {
+              ACE_ERROR ((LM_ERROR,
+                          ACE_TEXT ("ERROR: Sharing ORB configuration ")
+                          ACE_TEXT ("contexts is not yet supported\n")));
+            }
 
-      // The ORB table increases the reference count on the ORB Core
-      // so do not release it here.  Allow the TAO_ORB_Core_Auto_Ptr
-      // to decrease the reference count on the ORB Core when it goes
-      // out of scope.
-      oc.reset (tmp);
+          throw ::CORBA::BAD_PARAM
+            (CORBA::SystemException::_tao_minor_code
+             ( TAO_ORB_CORE_INIT_LOCATION_CODE,
+               ENOTSUP),
+             CORBA::COMPLETED_NO);
+        }
+      else
+        {
+          if (TAO_debug_level > 0)
+            {
+              ACE_ERROR ((LM_ERROR,
+                          ACE_TEXT ("ERROR: -ORBGestalt unknown value <%s>\n"),
+                          orbconfig_string.c_str()));
+            }
+
+          throw ::CORBA::BAD_PARAM
+            (CORBA::SystemException::_tao_minor_code
+             ( TAO_ORB_CORE_INIT_LOCATION_CODE,
+               EINVAL),
+             CORBA::COMPLETED_NO);
+        }
     }
 
+  // An ORB corresponding to the desired ORBid doesn't exist so create
+  // a new one.
+  TAO_ORB_Core * tmp = 0;
+  ACE_NEW_THROW_EX (tmp,
+                    TAO_ORB_Core (orbid_string.c_str (), gestalt),
+                    CORBA::NO_MEMORY (
+                      CORBA::SystemException::_tao_minor_code (0,
+                                                               ENOMEM),
+                      CORBA::COMPLETED_NO));
 
+  // The ORB table increases the reference count on the ORB Core
+  // so do not release it here.  Allow the TAO_ORB_Core_Auto_Ptr
+  // to decrease the reference count on the ORB Core when it goes
+  // out of scope.
+  oc.reset (tmp);
+
+  // The ORB now owns its configuration and the auto pointer is not
+  // necessary anymore.
+  guard_gestalt.release ();
+
   // Having the ORB's default static services be shared among all ORBs
   // is tempting from the point of view of reducing the dynamic
   // footprint. However, if the ORB in a DLL and the rest of that
@@ -1218,18 +1304,10 @@
   // which will have disastrous consequences at the process
   // shutdown. Hence, the ACE_Service_Config_Guard ensures that for
   // the current thread, any references to the global
-  // ACE_Service_Config will be forwarded to the ORB's.
+  // ACE_Service_Config will be re-routed to this ORB's service gestalt.
 
-  // Making sure the initialization process in the current thread uses
-  // the correct service repository (ours), instead of the global one.
   ACE_Service_Config_Guard scg (oc->configuration ());
 
-  /*
-   * Currently I choose to make the ORB an owner of its configuration,
-   * which in general is not quite correct because it is very common ORBs to
-   * need to share the same configuration.
-   */
-
   // Initialize the Service Configurator.  This must occur before the
   // ORBInitializer::pre_init() method is invoked on each registered
   // ORB initializer.
Index: TAO/tao/ORB_Core.cpp
===================================================================
--- TAO/tao/ORB_Core.cpp	(revision 79365)
+++ TAO/tao/ORB_Core.cpp	(working copy)
@@ -180,7 +180,8 @@
 
 // ****************************************************************
 
-TAO_ORB_Core::TAO_ORB_Core (const char *orbid)
+TAO_ORB_Core::TAO_ORB_Core (const char *orbid,
+                            ACE_Service_Gestalt* gestalt)
   : protocols_hooks_ (0),
     network_priority_protocols_hooks_ (0),
 #if TAO_USE_LOCAL_MEMORY_POOL == 1
@@ -258,6 +259,7 @@
     bidir_giop_policy_ (0),
     flushing_strategy_ (0),
     codeset_manager_ (0),
+    config_ (gestalt),
     sync_scope_hook_ (0),
     timeout_hook_ (0)
 {
@@ -285,19 +287,6 @@
   ACE_NEW (this->request_dispatcher_,
            TAO_Request_Dispatcher);
 
-  // @TODO: Can this be dynamic container instead?
-  if (ACE_OS::strnlen (this->orbid_, 1) == 0)
-    {
-      ACE_NEW (this->config_,
-               ACE_Service_Gestalt (ACE_Service_Gestalt::MAX_SERVICES,
-                                    false)); // (Re)use the process-global repository
-    }
-  else
-    {
-      ACE_NEW (this->config_,
-               ACE_Service_Gestalt (ACE_Service_Gestalt::MAX_SERVICES / 4,
-                                    true)); // Use our own service repository
-    }
 }
 
 TAO_ORB_Core::~TAO_ORB_Core (void)
@@ -339,7 +328,9 @@
   // This will destroy the service repository for this core
   (void) TAO::ORB::close_services (this->config_);
 
-  delete this->config_;
+  if (this->config_ != ACE_Service_Config::global())
+    delete this->config_;
+
   this->config_ = 0;
 }
 
Index: TAO/tao/ORB_Core.h
===================================================================
--- TAO/tao/ORB_Core.h	(revision 79365)
+++ TAO/tao/ORB_Core.h	(working copy)
@@ -164,7 +164,7 @@
 public:
 
   /// Constructor.
-  TAO_ORB_Core (const char* id);
+  TAO_ORB_Core (const char* id, ACE_Service_Gestalt* g);
 
   /// Accessor for the ORB parameters.
   TAO_ORB_Parameters *orb_params (void);
