Index: ace/OS_Thread_Adapter.cpp
===================================================================
--- ace/OS_Thread_Adapter.cpp	(revision 97735)
+++ ace/OS_Thread_Adapter.cpp	(working copy)
@@ -136,30 +136,26 @@
 #if defined (ACE_WIN32) || defined (ACE_HAS_TSS_EMULATION)
       // Call TSS destructors.
       ACE_OS::cleanup_tss (0 /* not main thread */);
-
-# if defined (ACE_WIN32)
-      // Exit the thread.  Allow CWinThread-destructor to be invoked
-      // from AfxEndThread.  _endthreadex will be called from
-      // AfxEndThread so don't exit the thread now if we are running
-      // an MFC thread.
-#   if defined (ACE_HAS_MFC) && (ACE_HAS_MFC != 0)
-      // Not spawned by ACE_Thread_Manager, use the old buggy
-      // version.  You should seriously consider using
-      // ACE_Thread_Manager to spawn threads.  The following code
-      // is know to cause some problem.
-      CWinThread *pThread = ::AfxGetThread ();
-
-      if (!pThread || pThread->m_nThreadID != ACE_OS::thr_self ())
-        ACE_ENDTHREADEX (status);
-      else
-        ::AfxEndThread (status);
-#   else
-      ACE_ENDTHREADEX (status);
-#   endif /* ACE_HAS_MFC && ACE_HAS_MFS != 0*/
-# endif /* ACE_WIN32 */
 #endif /* ACE_WIN32 || ACE_HAS_TSS_EMULATION */
     }
+#if defined (ACE_WIN32)
+  // Exit the thread.  Allow CWinThread-destructor to be invoked
+  // from AfxEndThread.  _endthreadex will be called from
+  // AfxEndThread so don't exit the thread now if we are running
+  // an MFC thread.
+# if defined (ACE_HAS_MFC) && (ACE_HAS_MFC != 0)
+  // Not spawned by ACE_Thread_Manager, use the old buggy
+  // version.  You should seriously consider using
+  // ACE_Thread_Manager to spawn threads.  The following code
+  // is know to cause some problem.
+  CWinThread *pThread = ::AfxGetThread ();
 
+  if (pThread && pThread->m_nThreadID == ACE_OS::thr_self ())
+      ::AfxEndThread (status);
+# endif /* ACE_HAS_MFC && ACE_HAS_MFC != 0*/
+  ACE_ENDTHREADEX (status);
+#endif /* ACE_WIN32 */
+
   return status;
 }
 
Index: ace/Thread_Adapter.cpp
===================================================================
--- ace/Thread_Adapter.cpp	(revision 97735)
+++ ace/Thread_Adapter.cpp	(working copy)
@@ -104,6 +104,7 @@
   void *arg = this->arg_;
 
 #if defined (ACE_WIN32) && defined (ACE_HAS_MFC) && (ACE_HAS_MFC != 0)
+  int using_afx = -1;
   ACE_OS_Thread_Descriptor *thr_desc = this->thr_desc_;
 #endif /* ACE_WIN32 && ACE_HAS_MFC && (ACE_HAS_MFC != 0) */
 
@@ -193,7 +194,6 @@
 
 #if defined (ACE_WIN32) || defined (ACE_HAS_TSS_EMULATION)
 # if defined (ACE_WIN32) && defined (ACE_HAS_MFC) && (ACE_HAS_MFC != 0)
-      int using_afx = -1;
       if (thr_desc)
         using_afx = ACE_BIT_ENABLED (thr_desc->flags (), THR_USE_AFX);
 # endif /* ACE_WIN32 && ACE_HAS_MFC && (ACE_HAS_MFC != 0) */
@@ -200,40 +200,38 @@
           // Call TSS destructors.
       ACE_OS::cleanup_tss (0 /* not main thread */);
 
-# if defined (ACE_WIN32)
-      // Exit the thread.  Allow CWinThread-destructor to be invoked
-      // from AfxEndThread.  _endthreadex will be called from
-      // AfxEndThread so don't exit the thread now if we are running
-      // an MFC thread.
-#   if defined (ACE_HAS_MFC) && (ACE_HAS_MFC != 0)
-      if (using_afx != -1)
-        {
-          if (using_afx)
-            ::AfxEndThread ((DWORD) status);
-          else
-            ACE_ENDTHREADEX (status);
-        }
-      else
-        {
-          // Not spawned by ACE_Thread_Manager, use the old buggy
-          // version.  You should seriously consider using
-          // ACE_Thread_Manager to spawn threads.  The following code
-          // is know to cause some problem.
-          CWinThread *pThread = ::AfxGetThread ();
-
-          if (!pThread || pThread->m_nThreadID != ACE_OS::thr_self ())
-            ACE_ENDTHREADEX (status);
-          else
-            ::AfxEndThread ((DWORD)status);
-        }
-#   else
-
-      ACE_ENDTHREADEX (status);
-#   endif /* ACE_HAS_MFC && ACE_HAS_MFS != 0*/
-# endif /* ACE_WIN32 */
 #endif /* ACE_WIN32 || ACE_HAS_TSS_EMULATION */
     }
+#if defined (ACE_WIN32)
+  // Exit the thread.  Allow CWinThread-destructor to be invoked
+  // from AfxEndThread.  _endthreadex will be called from
+  // AfxEndThread so don't exit the thread now if we are running
+  // an MFC thread.
+# if defined (ACE_HAS_MFC) && (ACE_HAS_MFC != 0)
+  if (using_afx != -1)
+  {
+      if (using_afx)
+      {
+          ::AfxEndThread ((DWORD) status);
+      }
+  }
+  else
+  {
+      // Not spawned by ACE_Thread_Manager, use the old buggy
+      // version.  You should seriously consider using
+      // ACE_Thread_Manager to spawn threads.  The following code
+      // is know to cause some problem.
+      CWinThread *pThread = ::AfxGetThread ();
 
+      if (pThread && pThread->m_nThreadID == ACE_OS::thr_self ())
+      {
+          ::AfxEndThread ((DWORD)status);
+      }
+  }
+# endif /* ACE_HAS_MFC && ACE_HAS_MFC != 0*/
+  ACE_ENDTHREADEX (status);
+#endif /* ACE_WIN32 */
+
   return status;
 }
 
