Index: Synch_Invocation.cpp
===================================================================
--- Synch_Invocation.cpp	(revision 92621)
+++ Synch_Invocation.cpp	(working copy)
@@ -564,6 +564,9 @@
                         && (ACE_OS_String::strcmp (type_id.in (),
                                 "IDL:omg.org/CORBA/INV_OBJREF:1.0") == 0))))))
       {
+        // If we are here then possibly we'll need a restart.
+        mon.set_status (TAO_INVOKE_RESTART);
+
         if (do_forward)
           this->stub ()->forwarded_on_exception (true);
 
@@ -597,6 +600,7 @@
           }
 
         // Fall through and raise an exception.
+        mon.set_status (TAO_INVOKE_FAILURE);
       }
 
     CORBA::SystemException *ex = TAO::create_system_exception (type_id.in ());
