Index: TAO/tao/Transport.h
===================================================================
--- TAO/tao/Transport.h	(revision 85158)
+++ TAO/tao/Transport.h	(working copy)
@@ -30,6 +30,7 @@
 #include "tao/Message_Semantics.h"
 #include "ace/Time_Value.h"
 #include "ace/Basic_Stats.h"
+#include "ace/Copy_Disabled.h"
 
 struct iovec;
 
@@ -243,7 +244,7 @@
  * https://svn.dre.vanderbilt.edu/viewvc/Middleware/trunk/TAO/docs/pluggable_protocols/index.html?revision=HEAD
  *
  */
-class TAO_Export TAO_Transport
+class TAO_Export TAO_Transport : private ACE_Copy_Disabled
 {
 public:
 
@@ -928,9 +929,20 @@
   /// partial_message_ data member.
   void allocate_partial_message_block (void);
 
-  // Disallow copying and assignment.
-  TAO_Transport (const TAO_Transport&);
-  void operator= (const TAO_Transport&);
+  /**
+   * @brief Re-factor computation of I/O timeouts based on operation
+   * timeouts.
+   * Depending on the wait strategy, we need to timeout I/O operations or
+   * not.  For example, if we are using a non-blocking strategy, we want
+   * to pass 0 to all I/O operations, and rely on the ACE_NONBLOCK
+   * settings on the underlying sockets.  However, for blocking strategies
+   * we want to pass the operation timeouts, to respect the application
+   * level policies.
+   *
+   * This function was introduced as part of the fixes for bug 3647.
+   */
+  ACE_Time_Value const *io_timeout(
+      ACE_Time_Value const * operation_timeout) const;
 
   /*
    * Specialization hook to add concrete private methods from
Index: TAO/tao/Transport.cpp
===================================================================
--- TAO/tao/Transport.cpp	(revision 85158)
+++ TAO/tao/Transport.cpp	(working copy)
@@ -410,7 +410,8 @@
   //    implementation to this base class method, and leave any TCP
   //    specific configuration out of this base class method.
   //      -Ossama
-  return this->send (iov, iovcnt, bytes_transferred, timeout);
+  return this->send (iov, iovcnt, bytes_transferred,
+                     this->io_timeout (timeout));
 }
 #endif  /* TAO_HAS_SENDFILE==1 */
 
@@ -927,7 +928,8 @@
                              byte_count);
   else
 #endif  /* TAO_HAS_SENDFILE==1 */
-    retval = this->send (iov, iovcnt, byte_count, max_wait_time);
+    retval = this->send (iov, iovcnt, byte_count,
+                         this->io_timeout (max_wait_time));
 
   if (TAO_debug_level == 5)
     {
@@ -1852,7 +1854,7 @@
   // Read the message into the existing message block on heap
   ssize_t const n = this->recv (q_data->msg_block ()->wr_ptr(),
                                 recv_size,
-                                max_wait_time);
+                                this->io_timeout (max_wait_time));
 
   if (n <= 0)
     {
@@ -2068,7 +2070,7 @@
   // the stack.
   ssize_t const n = this->recv (message_block.wr_ptr (),
                                 recv_size,
-                                max_wait_time);
+                                this->io_timeout (max_wait_time));
 
   // If there is an error return to the reactor..
   if (n <= 0)
@@ -2755,6 +2757,17 @@
 {
 }
 
+ACE_Time_Value const *
+TAO_Transport::io_timeout(
+    ACE_Time_Value const * operation_timeout) const
+{
+  if (this->wait_strategy()->can_process_upcalls())
+  {
+    return 0;
+  }
+  return operation_timeout;
+}
+
 /*
  * Hook to add concrete implementations from the derived class onto
  * TAO's transport.
