--- ACE_wrappers.orig/ace/MMAP_memory_Pool.cpp	2005-07-18 03:25:51.000000000 -0700
+++ ACE_wrappers/ace/MMAP_Memory_Pool.cpp	2005-08-26 16:54:27.065907200 -0700
@@ -348,6 +348,11 @@
                            ACE_LIB_TEXT ("MMAP_Memory_Pool::init_acquire, EEXIST")),
                           0);
 
+#if (ACE_HAS_POSITION_INDEPENDENT_POINTERS == 1)
+        ACE_BASED_POINTER_REPOSITORY::instance ()->bind (this->mmap_.addr(),
+                                                         this->mmap_.size());
+#endif /* ACE_HAS_POSITION_INDEPENDENT_POINTERS == 1 */
+
       return this->mmap_.addr ();
     }
   else
