*** ACE_wrappers/performance-tests/UDP/udp_test.cpp	2012-05-16 03:23:04.000000000 +0900
--- udp_test.cpp	2012-09-02 02:09:00.173381400 +0900
***************
*** 697,704 ****
        if (ACE_OS::ace_isdigit(argv[getopt.opt_ind ()][0]))
          {
            if (remote_addr.set (dstport,
!                                (ACE_UINT32) ACE_OS::inet_addr
!                                  (ACE_TEXT_ALWAYS_CHAR(argv[getopt.opt_ind ()]))) == -1)
              ACE_ERROR_RETURN ((LM_ERROR,
                                 "invalid IP address: %s\n",
                                 argv[getopt.opt_ind ()]),
--- 697,704 ----
        if (ACE_OS::ace_isdigit(argv[getopt.opt_ind ()][0]))
          {
            if (remote_addr.set (dstport,
!                                htonl((ACE_UINT32) ACE_OS::inet_addr
!                                  (ACE_TEXT_ALWAYS_CHAR(argv[getopt.opt_ind ()])))) == -1)
              ACE_ERROR_RETURN ((LM_ERROR,
                                 "invalid IP address: %s\n",
                                 argv[getopt.opt_ind ()]),
