diff -r -u TAO/orbsvcs/orbsvcs/IFRService.orig/IFR_Service_Utils.cpp TAO/orbsvcs/orbsvcs/IFRService/IFR_Service_Utils.cpp
--- TAO/orbsvcs/orbsvcs/IFRService.orig/IFR_Service_Utils.cpp	2008-07-22 09:54:58.050473700 -0400
+++ TAO/orbsvcs/orbsvcs/IFRService/IFR_Service_Utils.cpp	2008-07-22 10:01:25.555433700 -0400
@@ -346,7 +346,7 @@
     }
     
   const char *ifr_ior = this->ifr_ior_.in ();
-  size_t ior_size = ACE_OS::strlen (ifr_ior) + 1; 
+  size_t ior_size = ACE_OS::strlen (ifr_ior);
   ssize_t result = ACE_OS::write (fd, ifr_ior, ior_size);
   
   if (result != static_cast<ssize_t> (ior_size))
