Index: Block_Flushing_Strategy.cpp
===================================================================
RCS file: /usr/local/cvsroot/repository1/ACETAO5.6.1/TAO/tao/Block_Flushing_Strategy.cpp,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 Block_Flushing_Strategy.cpp
--- Block_Flushing_Strategy.cpp	4 Oct 2007 18:44:16 -0000	1.1.1.1
+++ Block_Flushing_Strategy.cpp	3 Dec 2007 21:30:39 -0000
@@ -25,7 +25,7 @@
                                             TAO_Queued_Message *msg,
                                             ACE_Time_Value *)
 {
-  while (!msg->all_data_sent ())
+  while (!transport->queue_is_empty())
     {
       if (transport->handle_output () == -1)
         return -1;
