Index: tao/corbafwd.h
===================================================================
RCS file: /home/src/ACE_wrappers/TAO/tao/corbafwd.h,v
retrieving revision 1.1.1.3
diff -u -r1.1.1.3 corbafwd.h
--- tao/corbafwd.h	13 Dec 2002 14:35:14 -0000	1.1.1.3
+++ tao/corbafwd.h	29 Aug 2003 15:50:54 -0000
@@ -1654,6 +1654,7 @@
 const CORBA::ULong TAO_POA_INACTIVE                           = (0x12U << 7);
 const CORBA::ULong TAO_CONNECTOR_REGISTRY_INIT_LOCATION_CODE  = (0x13U << 7);
 const CORBA::ULong TAO_AMH_REPLY_LOCATION_CODE                = (0x14U << 7);
+const CORBA::ULong TAO_CONNECTOR_TOO_MANY_PENDING             = (0x15U << 7);
 // *Don't* use TAO_<location>_MINOR_CODE greater than 0x1FU!
 //@}
 
Index: tao/Transport_Connector.cpp
===================================================================
RCS file: /home/src/ACE_wrappers/TAO/tao/Transport_Connector.cpp,v
retrieving revision 1.1.1.2
diff -u -r1.1.1.2 Transport_Connector.cpp
--- tao/Transport_Connector.cpp	13 Dec 2002 14:35:14 -0000	1.1.1.2
+++ tao/Transport_Connector.cpp	29 Aug 2003 15:50:54 -0000
@@ -9,6 +9,9 @@
 #include "debug.h"
 #include "Connect_Strategy.h"
 #include "Client_Strategy_Factory.h"
+#include "Endpoint.h"
+
+#include "ace/Atomic_Op.h"
 
 
 #if !defined (__ACE_INLINE__)
@@ -253,6 +256,48 @@
       // Successfull
       return 0;
     }
+
+  typedef ACE_Atomic_Op<ACE_Thread_Mutex,long> Atomic_Counter;
+  static Atomic_Counter concurrent_counter;
+  struct Counter_Guard
+  {
+    Counter_Guard(Atomic_Counter & counter)
+      : counter_(counter)
+      , value_(++counter_)
+    {
+    }
+    ~Counter_Guard()
+    {
+      --counter_;
+    }
+    long value() const
+    {
+      return value_;
+    }
+  private:
+    Atomic_Counter & counter_;
+    long value_;
+  };
+
+  Counter_Guard counter_guard(concurrent_counter);
+  const long max_pending = 128;
+  if(counter_guard.value() >= max_pending)
+  {
+    ACE_ERROR(
+        (LM_WARNING,
+            "(%P|%t) Too many (%d / %d) simultaneous connection "
+            "attempts in the ORB, raising TRANSIENT\n",
+            counter_guard.value(), max_pending));
+
+    ACE_THROW_RETURN(
+        CORBA::TRANSIENT(
+            CORBA_SystemException::_tao_minor_code (
+                TAO_CONNECTOR_TOO_MANY_PENDING,
+                errno),
+            CORBA::COMPLETED_NO),
+        -1);
+  }
+  
 
   // @@TODO: This is not the right place for this!
   // Purge connections (if necessary)
Index: tao/Exception.cpp
===================================================================
RCS file: /home/src/ACE_wrappers/TAO/tao/Exception.cpp,v
retrieving revision 1.1.1.4
diff -u -r1.1.1.4 Exception.cpp
--- tao/Exception.cpp	10 Aug 2003 20:23:39 -0000	1.1.1.4
+++ tao/Exception.cpp	29 Aug 2003 15:50:54 -0000
@@ -474,6 +474,9 @@
         case TAO_AMH_REPLY_LOCATION_CODE:
           location = "Failure when trying to send AMH reply";
           break;
+        case TAO_CONNECTOR_TOO_MANY_PENDING:
+          location = "Too many outstanding connections";
+          break;
         default:
           location = "unknown location";
         }
