Index: ace/Service_Types.cpp
===================================================================
--- ace/Service_Types.cpp	(revision 74354)
+++ ace/Service_Types.cpp	(working copy)
@@ -89,6 +89,7 @@
                                                   u_int f,
                                                   ACE_Service_Object_Exterminator gobbler)
   : ACE_Service_Type_Impl (so, s_name, f, gobbler)
+  , initialized_ (-1)
 {
   ACE_TRACE ("ACE_Service_Object_Type::ACE_Service_Object_Type");
 }
@@ -105,8 +106,10 @@
 
   if (so == 0)
     return -1;
-  else
-    return so->init (argc, argv);
+
+  this->initialized_ = so->init (argc, argv);
+
+  return this->initialized_;
 }
 
 int
@@ -119,18 +122,12 @@
   ACE_Service_Object * const so =
     static_cast<ACE_Service_Object *> (obj);
 
-  if (so)
-    {
+  // Call fini() if an only if, the object was successfuly
+  // initialized, i.e. init() returned 0. This is necessary to
+  // maintain the ctor/dtor-like semantics for init/fini.
+  if (so && initialized_ == 0)
       so->fini ();
 
-      // @TODO: Why is this disabled?
-#if 0
-      if (ACE_BIT_ENABLED (this->flags_,
-                           ACE_Service_Type::DELETE_OBJ))
-        delete so;
-#endif /* 1 */
-    }
-
   return ACE_Service_Type_Impl::fini ();
 }
 
Index: ace/Service_Types.h
===================================================================
--- ace/Service_Types.h	(revision 74354)
+++ ace/Service_Types.h	(working copy)
@@ -106,6 +106,10 @@
   virtual int init (int argc, ACE_TCHAR *argv[]) const;
   virtual int fini (void) const;
   virtual int info (ACE_TCHAR **str, size_t len) const;
+
+private:
+  /// Holds the initialization status (result of object->init())
+  mutable int initialized_;
 };
 
 /**
