Index: ACE_wrappers/ace/Service_Repository.h
===================================================================
--- ACE_wrappers/ace/Service_Repository.h	(revision 73853)
+++ ACE_wrappers/ace/Service_Repository.h	(working copy)
@@ -27,6 +27,7 @@
 ACE_BEGIN_VERSIONED_NAMESPACE_DECL
 
 class ACE_Service_Type;
+class Gestalt_Current_DLL_Guard;
 
 #define ACE_Component_Repository ACE_Service_Repository
 /**
@@ -48,6 +49,7 @@
 {
 public:
   friend class ACE_Service_Repository_Iterator;
+  friend class Gestalt_Current_DLL_Guard;
 
   enum
   {
Index: ACE_wrappers/ace/Parse_Node.cpp
===================================================================
--- ACE_wrappers/ace/Parse_Node.cpp	(revision 73853)
+++ ACE_wrappers/ace/Parse_Node.cpp	(working copy)
@@ -405,13 +405,6 @@
 {
   ACE_TRACE ("ACE_Location_Node::open_dll");
 
-#ifndef ACE_NLOGGING
-  if (ACE::debug ())
-    ACE_DEBUG ((LM_DEBUG,
-                ACE_LIB_TEXT ("(%P|%t) LN::open_dll - path=%s\n"),
-                this->pathname ()));
-#endif /* ACE_NLOGGING */
-
   if (-1 == this->dll_.open (this->pathname ()))
     {
       ++yyerrno;
@@ -756,6 +749,77 @@
   delete[] const_cast<ACE_TCHAR *> (this->function_name_);
 }
 
+ACE_ALLOC_HOOK_DEFINE (ACE_Service_Type_Factory)
+
+ACE_Service_Type_Factory::ACE_Service_Type_Factory (ACE_TCHAR const *name,
+                                                    int type,
+                                                    ACE_Location_Node *location,
+                                                    int active)
+  : name_ (name)
+  , type_ (type)
+  , location_ (location)
+  , is_active_ (active)
+{
+}
+
+
+ACE_Service_Type_Factory::~ACE_Service_Type_Factory (void)
+{
+}
+
+
+ACE_Service_Type *
+ACE_Service_Type_Factory::make_service_type (ACE_Service_Gestalt *cfg) const
+{
+  ACE_TRACE ("ACE_Service_Type_Factory::make_service_type");
+
+  u_int flags = ACE_Service_Type::DELETE_THIS
+    | (this->location_->dispose () == 0 ? 0 : ACE_Service_Type::DELETE_OBJ);
+
+  ACE_Service_Object_Exterminator gobbler = 0;
+
+  int yyerrno = 0;
+  void *sym = this->location_->symbol (cfg, yyerrno, &gobbler);
+
+  if (sym != 0)
+    {
+      ACE_Service_Type_Impl *stp
+        = ACE_Service_Config::create_service_type_impl (this->name (),
+                                                        this->type_,
+                                                        sym,
+                                                        flags,
+                                                        gobbler);
+      if (stp == 0)
+        ++yyerrno;
+
+      ACE_Service_Type *tmp = 0;
+      ACE_NEW_RETURN (tmp,
+                      ACE_Service_Type (this->name (),
+                                        stp,
+                                        this->location_->dll (),
+                                        this->is_active_),
+                      0);
+      return tmp;
+    }
+  else
+    {
+#ifndef ACE_NLOGGING
+      ACE_ERROR ((LM_ERROR,
+                  ACE_LIB_TEXT ("(%P|%t) Unable to find service \'%s\'\n"),
+                  this->name ()));
+#endif
+      ++yyerrno;
+      return 0;
+    }
+}
+
+ACE_TCHAR const*
+ACE_Service_Type_Factory::name (void) const
+{
+  return name_.c_str ();
+}
+
+
 ACE_END_VERSIONED_NAMESPACE_DECL
 
 #endif /* ACE_USES_CLASSIC_SVC_CONF == 1 */
Index: ACE_wrappers/ace/Parse_Node.h
===================================================================
--- ACE_wrappers/ace/Parse_Node.h	(revision 73853)
+++ ACE_wrappers/ace/Parse_Node.h	(working copy)
@@ -476,6 +476,44 @@
   ACE_UNIMPLEMENTED_FUNC (ACE_Static_Function_Node& operator= (const ACE_Static_Function_Node&))
 };
 
+// A helper class used to safely register dynamic services, which may contains
+// subordinate static services. It is used to capture the necessary data during
+// the parsing, but perform the actuall instantiation later.
+class ACE_Service_Type_Factory
+{
+public:
+  ACE_Service_Type_Factory (ACE_TCHAR const *name,
+                            int type,
+                            ACE_Location_Node *location,
+                            int active);
+
+  ~ACE_Service_Type_Factory (void);
+
+  ACE_Service_Type *make_service_type (ACE_Service_Gestalt *pcfg) const;
+
+  ACE_TCHAR const* name (void) const;
+
+  /// Declare the dynamic allocation hooks.
+  ACE_ALLOC_HOOK_DECLARE;
+
+private:
+
+  /**
+   * Not implemented to enforce no copying
+   */
+  ACE_UNIMPLEMENTED_FUNC
+    (ACE_Service_Type_Factory(const ACE_Service_Type_Factory&))
+
+  ACE_UNIMPLEMENTED_FUNC
+    (ACE_Service_Type_Factory& operator=(const ACE_Service_Type_Factory&))
+
+private:
+  ACE_TString name_;
+  int type_;
+  ACE_Auto_Ptr<ACE_Location_Node> location_;
+  int is_active_;
+};
+
 ACE_END_VERSIONED_NAMESPACE_DECL
 
 #endif /* ACE_USES_CLASSIC_SVC_CONF == 1 */
Index: ACE_wrappers/ace/Service_Object.h
===================================================================
--- ACE_wrappers/ace/Service_Object.h	(revision 73853)
+++ ACE_wrappers/ace/Service_Object.h	(working copy)
@@ -147,6 +147,10 @@
 
   /// 1 if <fini> on <type_> has already been called, otherwise 0.
   int fini_already_called_;
+
+private:
+  friend class Gestalt_Current_DLL_Guard;
+
 };
 
 /**
Index: ACE_wrappers/ace/Service_Gestalt.h
===================================================================
--- ACE_wrappers/ace/Service_Gestalt.h	(revision 73853)
+++ ACE_wrappers/ace/Service_Gestalt.h	(working copy)
@@ -29,13 +29,20 @@
 #include "ace/Singleton.h"
 #include "ace/OS_NS_signal.h"
 #include "ace/Synch_Traits.h"
+#include "ace/DLL.h"
 
 ACE_BEGIN_VERSIONED_NAMESPACE_DECL
 
+#if (ACE_USES_CLASSIC_SVC_CONF == 1)
 class ACE_Service_Type_Factory;
+class ACE_Location_Node;
+#else
+class ACE_XML_Svc_Conf;
+class ACE_DLL;
+#endif /* ACE_USES_CLASSIC_SVC_CONF == 1 */
+
 class ACE_Static_Svc_Descriptor;
 class ACE_Svc_Conf_Param;
-
 class ACE_Service_Gestalt;
 
 /**
@@ -43,6 +50,21 @@
  *
  * @brief Supplies common server operations for dynamic and static
  * configuration of services.
+ *
+ * The Gestalt embodies the concept of configuration context. On one
+ * hand, it is a flat namespace, where names correspond to a Service
+ * Object instance. A Gestalt owns the Service Repository instance,
+ * which in turn owns the Service Object instances.
+ *
+ * Another aspect of a Gestalt is its responsibility for
+ * record-keeping and accounting for the meta-data, necessary for
+ * locating, removing or instantiating a service.
+ *
+ * A repository underlies an instance of a gestalt and its lifetime
+ * may or may not be bounded by the lifetime of the gestalt, that owns
+ * it. This feature is important for the derived classes and the
+ * Service Config in particular.
+
  */
 class ACE_Export ACE_Service_Gestalt
 {
@@ -194,10 +216,10 @@
 
   /**
    * Handle the command-line options intended for the
-   * <ACE_Service_Config>.  Note that <argv[0]> is assumed to be the
+   * <ACE_Service_Gestalt>.  Note that <argv[0]> is assumed to be the
    * program name.
+   *
    * The arguments that are valid in a call to this method are
-   * - '-b' Option to indicate that we should be a daemon
    * - '-d' Turn on debugging mode
    * - '-f' Option to read in the list of svc.conf file names
    * - '-k' Option to read a wide string where in the logger output can
@@ -227,6 +249,8 @@
   int insert (ACE_Static_Svc_Descriptor *stsd);
 
   // = Utility methods.
+
+#if (ACE_USES_CLASSIC_SVC_CONF == 1)
   /// Dynamically link the shared object file and retrieve a pointer to
   /// the designated shared object in this file. Also account for the
   /// possiblity to have static services registered when loading the DLL, by
@@ -235,18 +259,20 @@
   /// problems.
   int initialize (const ACE_Service_Type_Factory *,
                   const ACE_TCHAR *parameters);
+#endif /* (ACE_USES_CLASSIC_SVC_CONF == 1) */
 
   // Dynamically link the shared object file and retrieve a pointer to
   // the designated shared object in this file.
   // @obsolete
-  // @note This is error-prone in the presense of dynamic
-  // services with their own static services. This method will allow those
-  // static services to register *before* the dynamic service that owns them.
-  // Upon finalization of the static services the process may crash, because
-  // the dynamic service's DLL may have been already released, together with
-  // the memory in which the static services reside.
-  // It may not crash, for instance, when the first static service to register
-  // is the same as the dynamic service being loaded. You should be so lucky!
+  // @note This is error-prone in the presense of dynamic services,
+  // which in turn initialize their own static services. This method
+  // will allow those static services to register *before* the dynamic
+  // service that owns them.  Upon finalization of the static services
+  // the process will typically crash, because the dynamic service's
+  // DLL may have been already released, together with the memory in
+  // which the static services reside.  It may not crash, for
+  // instance, when the first static service to register is the same
+  // as the dynamic service being loaded. You should be so lucky!
   int initialize (const ACE_Service_Type *,
                   const ACE_TCHAR *parameters);
 
@@ -335,7 +361,7 @@
 #else
   /// Helper function to dynamically link in the XML Service Configurator
   /// parser.
-  ACE_XML_Svc_Conf *get_xml_svc_conf (ACE_DLL &d);
+  ACE_XML_Svc_Conf* get_xml_svc_conf (ACE_DLL &d);
 #endif /* ACE_USES_CLASSIC_SVC_CONF == 1 */
 
   // Dynamically link the shared object file and retrieve a pointer to
@@ -411,47 +437,46 @@
   /// the static_svcs_ list.
   ACE_PROCESSED_STATIC_SVCS* processed_static_svcs_;
 
+private:
+  /// As dynamic service objects are loaded, static service objects
+  /// that come with the same DLL code, may also be registered. The
+  /// static services hoever, have no information about the context in
+  /// which their registration takes place and their relation with the
+  /// DLL is lost. A situation can easily arise where the dynamic
+  /// service object is finalized and its DLL - unmapped, but some of
+  /// the DLL's static services may still be around. Their
+  /// finalization would not be possible, if the DLL containg that
+  /// code is gone. To prevent cases like this, we keep a handler for
+  /// the last DLL dlopen()-ed. On the current thread, there will be
+  /// just one and its handle will be kept here to be used by every
+  /// static service that registers itself. In the space between
+  /// dlopen()-ing a DLL and retrieving the symbol to call init() on,
+  /// any static service that registers will be associated with the
+  /// "current" DLL. During finalization, this will keep the reference
+  /// count positive, until the last static or dynamic service object
+  /// in a DLL is finalized, which will finally unload the DLL.
+  //  ACE_DLL current_dll_;
+  
+  friend class Gestalt_Current_DLL_Guard;
+  
 }; /* class ACE_Service_Gestalt */
 
-
-class ACE_Location_Node;
-
-// A helper class used to safely register dynamic services, which may contains
-// subordinate static services. It is used to capture the necessary data during
-// the parsing, but perform the actuall instantiation later.
-class ACE_Service_Type_Factory
+class Gestalt_Current_DLL_Guard 
 {
 public:
-  ACE_Service_Type_Factory (ACE_TCHAR const *name,
-                            int type,
-                            ACE_Location_Node *location,
-                            int active);
-
-  ~ACE_Service_Type_Factory (void);
-
-  ACE_Service_Type *make_service_type (ACE_Service_Gestalt *pcfg) const;
-
-  ACE_TCHAR const* name (void) const;
-
-  /// Declare the dynamic allocation hooks.
-  ACE_ALLOC_HOOK_DECLARE;
-
+  Gestalt_Current_DLL_Guard (ACE_Service_Gestalt* gestalt);
+  ~Gestalt_Current_DLL_Guard (void);
+  int accept (const ACE_DLL& adll);
+protected:
+  ACE_UNIMPLEMENTED_FUNC (Gestalt_Current_DLL_Guard 
+                            (const Gestalt_Current_DLL_Guard&))
+  ACE_UNIMPLEMENTED_FUNC (Gestalt_Current_DLL_Guard& operator= 
+                            (const Gestalt_Current_DLL_Guard&))
 private:
-
-  /**
-   * Not implemented to enforce no copying
-   */
-  ACE_UNIMPLEMENTED_FUNC
-    (ACE_Service_Type_Factory(const ACE_Service_Type_Factory&))
-
-  ACE_UNIMPLEMENTED_FUNC
-    (ACE_Service_Type_Factory& operator=(const ACE_Service_Type_Factory&))
-
-private:
-  ACE_TString name_;
-  int type_;
-  ACE_Auto_Ptr<ACE_Location_Node> location_;
-  int is_active_;
+  ACE_Service_Gestalt* gestalt_;
+  ACE_Service_Repository* repo_;
+  ACE_Service_Gestalt::ACE_PROCESSED_STATIC_SVCS* processed_static_svcs_;
+  bool accept_;
 };
 
 
Index: ACE_wrappers/ace/Service_Gestalt.cpp
===================================================================
--- ACE_wrappers/ace/Service_Gestalt.cpp	(revision 73853)
+++ ACE_wrappers/ace/Service_Gestalt.cpp	(working copy)
@@ -171,6 +171,8 @@
 
 
 
+
+
 // ----------------------------------------
 
 ACE_Service_Gestalt::Processed_Static_Svc::
@@ -180,23 +182,90 @@
 {
   ACE_NEW_NORETURN (name_, ACE_TCHAR[ACE_OS::strlen(assd->name_)+1]);
   ACE_OS::strcpy(name_,assd->name_);
-  if (ACE::debug ())
-    ACE_DEBUG ((LM_DEBUG,
-                ACE_LIB_TEXT ("(%P|%t) PSS::ctor -  name = %s\n"),
-                name_));
 }
 
 ACE_Service_Gestalt::Processed_Static_Svc::
 ~Processed_Static_Svc (void)
 {
-  if (ACE::debug ())
-    ACE_DEBUG ((LM_DEBUG,
-                ACE_LIB_TEXT ("(%P|%t) PSS::dtor -  name = %s\n"),
-                name_));
-
   delete [] name_;
 }
 
+
+
+Gestalt_Current_DLL_Guard::Gestalt_Current_DLL_Guard (ACE_Service_Gestalt* gestalt)
+  : gestalt_ (gestalt)
+  , repo_ (gestalt->repo_)
+  , processed_static_svcs_ (gestalt->processed_static_svcs_)
+{
+  ACE_NEW_NORETURN (this->gestalt_->repo_,
+                    ACE_Service_Repository (128));
+   
+  if (this->processed_static_svcs_ == 0)
+    ACE_NEW_NORETURN (this->processed_static_svcs_,
+                      ACE_Service_Gestalt::ACE_PROCESSED_STATIC_SVCS);
+ 
+  this->gestalt_->processed_static_svcs_ = 0;
+}
+ 
+ 
+int
+Gestalt_Current_DLL_Guard::accept (const ACE_DLL& adll)
+{
+  ACE_Service_Repository* target = this->repo_;
+  ACE_Service_Repository* source = this->gestalt_->repo_;
+   
+  if (target->total_size_ <= source->current_size_ + target->current_size_)
+    return -1; // no memory
+   
+  for (size_t i = 0; i < source->current_size_; i++)
+    {
+      ACE_Service_Type* ast = const_cast<ACE_Service_Type*> (source->service_vector_[i]);
+      if (ast->dll_.dll_name_ == 0)
+        {
+          ast->dll_ = adll;
+ 
+          ACE_DEBUG ((LM_DEBUG, 
+                      ACE_TEXT ("ACE (%P|%t) GCDG::accept, service=%s, dll=%s\n"),
+                      ast->name_, adll.dll_name_));
+        }
+ 
+      target->insert (ast);
+      source->service_vector_[i] = 0;
+    }
+  source->current_size_ = 0;
+   
+  if (this->gestalt_->processed_static_svcs_ != 0 && !this->gestalt_->processed_static_svcs_->is_empty())
+    {
+      ACE_Service_Gestalt::Processed_Static_Svc **pss = 0;
+      for (ACE_Service_Gestalt::ACE_PROCESSED_STATIC_SVCS_ITERATOR iter (*this->gestalt_->processed_static_svcs_);
+           iter.next (pss) != 0;
+           iter.advance ())
+        {
+          ACE_Service_Gestalt::Processed_Static_Svc *tmp = 0;
+          ACE_NEW_RETURN (tmp, 
+                          ACE_Service_Gestalt::Processed_Static_Svc((*pss)->assd_), 
+                          -1);
+          this->processed_static_svcs_->insert(tmp);
+ 
+          delete *pss;
+        }
+    }
+ 
+  return 0;
+}
+ 
+Gestalt_Current_DLL_Guard::~Gestalt_Current_DLL_Guard (void)
+{
+  delete this->gestalt_->repo_;
+  this->gestalt_->repo_ =
+    this->repo_;
+ 
+  delete this->gestalt_->processed_static_svcs_;
+  this->gestalt_->processed_static_svcs_ = 
+    this->processed_static_svcs_;
+}
+ 
+
 // ----------------------------------------
 
 ACE_Service_Gestalt::~ACE_Service_Gestalt (void)
@@ -386,79 +455,15 @@
 
 
 
-ACE_ALLOC_HOOK_DEFINE (ACE_Service_Type_Factory)
+/// (Statically) initialize a service object.  Locates the service
+/// object's registration record, identified by svc_name and
+/// initializes it, using the supplied command line parameters. Own
+/// repository is searched first and if the service is located it is
+/// initializaed and activated. The "processed" services descriptors
+/// from the global repository are looked up otherwice, and if one is
+/// found it is duplicated locally. The initialization and activatin
+/// then proceeds as described above.
 
-ACE_Service_Type_Factory::ACE_Service_Type_Factory (ACE_TCHAR const *name,
-                                                    int type,
-                                                    ACE_Location_Node *location,
-                                                    int active)
-  : name_ (name)
-  , type_ (type)
-  , location_ (location)
-  , is_active_ (active)
-{
-}
-
-
-ACE_Service_Type_Factory::~ACE_Service_Type_Factory (void)
-{
-}
-
-
-ACE_Service_Type *
-ACE_Service_Type_Factory::make_service_type (ACE_Service_Gestalt *cfg) const
-{
-  ACE_TRACE ("ACE_Service_Type_Factory::make_service_type");
-
-  u_int flags = ACE_Service_Type::DELETE_THIS
-    | (this->location_->dispose () == 0 ? 0 : ACE_Service_Type::DELETE_OBJ);
-
-  ACE_Service_Object_Exterminator gobbler = 0;
-
-  int yyerrno = 0;
-  void *sym = this->location_->symbol (cfg, yyerrno, &gobbler);
-
-  if (sym != 0)
-    {
-      ACE_Service_Type_Impl *stp
-        = ACE_Service_Config::create_service_type_impl (this->name (),
-                                                        this->type_,
-                                                        sym,
-                                                        flags,
-                                                        gobbler);
-      if (stp == 0)
-        ++yyerrno;
-
-      ACE_Service_Type *tmp = 0;
-      ACE_NEW_RETURN (tmp,
-                      ACE_Service_Type (this->name (),
-                                        stp,
-                                        this->location_->dll (),
-                                        this->is_active_),
-                      0);
-      return tmp;
-    }
-  else
-    {
-#ifndef ACE_NLOGGING
-      ACE_ERROR ((LM_ERROR,
-                  ACE_LIB_TEXT ("(%P|%t) Unable to find service \'%s\'\n"),
-                  this->name ()));
-#endif
-      ++yyerrno;
-      return 0;
-    }
-}
-
-ACE_TCHAR const*
-ACE_Service_Type_Factory::name (void) const
-{
-  return name_.c_str ();
-}
-
-
-///
-
 int
 ACE_Service_Gestalt::initialize (const ACE_TCHAR *svc_name,
                                  const ACE_TCHAR *parameters)
@@ -480,7 +485,6 @@
 
   const ACE_Service_Type *srp = 0;
   for (int i = 0; this->repo_->find (svc_name, &srp) == -1 && i < 2; i++)
-    //  if (this->repo_->find (svc_name, &srp) == -1)
     {
       const ACE_Static_Svc_Descriptor *assd =
         ACE_Service_Config::global()->find_processed_static_svc(svc_name);
@@ -523,12 +527,13 @@
 }
 
 
+#if (ACE_USES_CLASSIC_SVC_CONF == 1)
 int
 ACE_Service_Gestalt::initialize (const ACE_Service_Type_Factory *stf,
                                  const ACE_TCHAR *parameters)
 {
   ACE_TRACE ("ACE_Service_Gestalt::initialize");
-
+  
 #ifndef ACE_NLOGGING
   if (ACE::debug () > 1)
     ACE_DEBUG ((LM_DEBUG,
@@ -537,11 +542,11 @@
                 this->repo_,
                 stf->name ()));
 #endif
-
+  
   ACE_Service_Type *srp = 0;
   int const retv = this->repo_->find (stf->name (),
                                       (const ACE_Service_Type **) &srp);
-
+  
   // If there is an active service already, it must first be removed,
   // before it could be re-installed.
   if (retv >= 0)
@@ -554,9 +559,9 @@
                            stf->name ()),
                           0);
 #endif
-        return 0;
+      return 0;
     }
-
+  
   // There is an inactive service by that name, so it may have been
   // either inactivated, or just a forward declaration for a service,
   // that is in the process of being loaded. If the latter, then we
@@ -574,6 +579,7 @@
                        ACE_LIB_TEXT ("already performing one.\n"),
                        stf->name ()),
                       -1);
+  
 
   // Reserve a spot for the dynamic service by inserting an incomplete
   // service declaration, i.e. one that can not produce a service
@@ -584,21 +590,25 @@
   // registered *after* the dynamic service that loads them, so that
   // their finalization is complete *before* finalizing the dynamic
   // service.
-  ACE_Service_Type_Forward_Declaration_Guard dummy (this->repo_,
-                                                    stf->name ());
+  //  ACE_Service_Type_Forward_Declaration_Guard dummy (this->repo_,
+  //                                                  stf->name ());
 
-  // make_service_type() is doing the dynamic loading and also runs
-  // any static initializers
-  ACE_Auto_Ptr<ACE_Service_Type> tmp (stf->make_service_type (this));
-  if (tmp.get () != 0 &&
-      this->initialize_i (tmp.get (), parameters) == 0)
+    Gestalt_Current_DLL_Guard current_dll_guard (this);
+    
+    // make_service_type() is doing the dynamic loading and also runs
+    // any static initializers
+    ACE_Auto_Ptr<ACE_Service_Type> tmp (stf->make_service_type (this));
+    if (tmp.get () != 0 
+        && this->initialize_i (tmp.get (), parameters) == 0)
     {
+      current_dll_guard.accept (tmp->dll ());
+
       // All good the ACE_Service_Type instance is now owned by the repository
       // and we should make sure it is not destroyed upon exit from this method.
       (void)tmp.release ();
       return 0;
     }
-
+  
   // Something went wrong ...
 #ifndef ACE_NLOGGING
   if (ACE::debug ())
@@ -607,9 +617,10 @@
                        stf->name()),
                       -1);
 #endif
-
+  
   return -1;
 }
+#endif /* (ACE_USES_CLASSIC_SVC_CONF == 1) */
 
 
 // Dynamically link the shared object file and retrieve a pointer to
@@ -631,7 +642,7 @@
 
   if (ACE::debug ())
     ACE_DEBUG ((LM_DEBUG,
-                ACE_LIB_TEXT ("(%P|%t) SG::initialize - looking up dynamic ")
+                ACE_LIB_TEXT ("(%P|%t) SG::initialize OBSOLETE - looking up dynamic ")
                 ACE_LIB_TEXT (" service \'%s\' to initialize\n"),
                 sr->name ()));
 
@@ -739,12 +750,12 @@
 
 int
 ACE_Service_Gestalt::process_directive_i (const ACE_Static_Svc_Descriptor &ssd,
-                                        int force_replace)
+                                          int force_replace)
 {
 #ifndef ACE_NLOGGING
   if (ACE::debug () > 2)
     ACE_DEBUG ((LM_DEBUG,
-                ACE_LIB_TEXT ("(%P|%t) SG::process_directive, ")
+                ACE_LIB_TEXT ("ACE (%P|%t) SG::process_directive, ")
                 ACE_LIB_TEXT ("repo=%@, replace=%d - %s\n"),
                 this->repo_,
                 force_replace,
@@ -789,6 +800,8 @@
   return this->repo_->insert (service_type);
 }
 
+
+
 #if (ACE_USES_CLASSIC_SVC_CONF == 1)
 
 int
@@ -834,15 +847,16 @@
     return 0;
 }
 
-#else
+#else /* ACE_USES_CLASSIC_SVC_CONF == 1 */
 
 ACE_XML_Svc_Conf *
 ACE_Service_Gestalt::get_xml_svc_conf (ACE_DLL &xmldll)
 {
   if (xmldll.open (ACE_LIB_TEXT ("ACEXML_XML_Svc_Conf_Parser")) == -1)
     ACE_ERROR_RETURN ((LM_ERROR,
-                       ACE_LIB_TEXT ("Fail to open ACEXML_XML_Svc_Conf_Parser: %p\n"),
-                       "ACE_Service_Config::get_xml_svc_conf"),
+                       ACE_LIB_TEXT ("Fail to open ACEXML_XML_Svc_Conf_Parser: %s\n"),
+                       //                       "ACE_Service_Gestalt::get_xml_svc_conf"),
+                       xmldll.error ()),
                       0);
 
   void *foo;
@@ -860,7 +874,7 @@
 
   return factory ();
 }
-#endif /* ACE_USES_CLASSIC_SVC_CONF == 1 */
+#endif /* ACE_USES_CLASSIC_SVC_CONF != 1 */
 
 int
 ACE_Service_Gestalt::process_file (const ACE_TCHAR file[])
@@ -885,9 +899,6 @@
   // The entry will be automaticaly removed once the thread exits this block.
   ACE_Service_Type_Forward_Declaration_Guard recursion_guard (this->repo_, file);
 
-  /*
-   * @TODO: Test with ACE_USES_CLASSIC_SVC_CONF turned off!
-   */
 #if (ACE_USES_CLASSIC_SVC_CONF == 1)
   int result = 0;
 
@@ -928,7 +939,7 @@
   ACE_DLL dll;
 
   auto_ptr<ACE_XML_Svc_Conf>
-    xml_svc_conf (ACE_Service_Config::get_xml_svc_conf (dll));
+    xml_svc_conf (this->get_xml_svc_conf (dll));
 
   if (xml_svc_conf.get () == 0)
     return -1;
Index: ACE_wrappers/ace/Service_Config.h
===================================================================
--- ACE_wrappers/ace/Service_Config.h	(revision 73853)
+++ ACE_wrappers/ace/Service_Config.h	(working copy)
@@ -141,7 +141,7 @@
  * @class ACE_Service_Config
  *
  * @brief Supplies common server operations for dynamic and static
- * configuration of services.
+ * configuration of service. 
  *
  * The ACE_Service_Config uses the Monostate pattern.  Therefore,
  * you can only have one of these instantiated per-process. It
@@ -535,10 +535,6 @@
   /// and process_directive() both call.  Returns the number of errors
   /// that occurred.
   static int process_directives_i (ACE_Svc_Conf_Param *param);
-#else
-  /// Helper function to dynamically link in the XML Service Configurator
-  /// parser.
-  static ACE_XML_Svc_Conf *get_xml_svc_conf (ACE_DLL &d);
 #endif /* ACE_USES_CLASSIC_SVC_CONF == 1 */
 
   /// Become a daemon.
Index: ACE_wrappers/ace/Service_Config.inl
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/Service_Config.inl,v
retrieving revision 4.6
diff -b -u -r4.6 Service_Config.inl
--- ACE_wrappers/ace/Service_Config.inl      27 May 2006 15:52:26 -0000      4.6
+++ ACE_wrappers/ace/Service_Config.inl      24 Jul 2006 19:15:10 -0000
@@ -96,7 +96,14 @@
                                                      parameters);
 }

+/// Process a file containing a list of service configuration
+/// directives.
+ACE_INLINE int ACE_Service_Config::process_file (const ACE_TCHAR file[])
+{
+  return ACE_Service_Config::current ()->process_file (file);
+}

+///
 ACE_INLINE int
 ACE_Service_Config::process_directive (const ACE_TCHAR directive[])
 {
Index: ACE_wrappers/ace/DLL_Manager.cpp
===================================================================
--- ACE_wrappers/ace/DLL_Manager.cpp	(revision 73853)
+++ ACE_wrappers/ace/DLL_Manager.cpp	(working copy)
@@ -121,16 +121,23 @@
           ACE_TString *name = 0;
           while (name_iter.next (name))
             {
-              if (ACE::debug ())
-                ACE_DEBUG ((LM_DEBUG,
-                            ACE_LIB_TEXT ("(%P|%t) ACE_DLL_Handle::open: ")
-                            ACE_LIB_TEXT ("calling dlopen on ")
-                            ACE_LIB_TEXT ("\"%s\"\n"), name->c_str ()));
-
               // The ACE_SHLIB_HANDLE object is obtained.
               this->handle_ = ACE_OS::dlopen (name->c_str (),
                                               open_mode);
 
+              if (ACE::debug () > 1)
+                {
+                  ACE_DEBUG ((LM_DEBUG,
+                              ACE_LIB_TEXT ("ACE (%P|%t) DLL_Handle::open ")
+                              ACE_LIB_TEXT ("(\"%s\", 0x%x) -> %s: %s\n"), 
+                              name->c_str (), 
+                              open_mode,
+                              ACE_LIB_TEXT ((this->handle_ != ACE_SHLIB_INVALID_HANDLE) 
+                                            ? "succeeded"
+                                            : "failed"),
+                              this->error()->c_str()));
+                }
+
               if (this->handle_ != ACE_SHLIB_INVALID_HANDLE)   // Good one?
                 break;
 
@@ -140,11 +147,11 @@
               // mask it.
               // @TODO: If we've found our DLL _and_ it's
               // broken, should we continue at all?
-              if (ACE::debug () && (errno != 0) && (errno != ENOENT))
+              if ((errno != 0) && (errno != ENOENT) && ACE::debug ())
                 ACE_ERROR ((LM_ERROR,
-                            ACE_LIB_TEXT ("(%P|%t) ACE_DLL_Handle::open: ")
-                            ACE_LIB_TEXT ("Attempt to open \'%s\' failed ")
-                            ACE_LIB_TEXT ("(%d): %s\n"),
+                            ACE_LIB_TEXT ("ACE (%P|%t) DLL_Handle::open ")
+                            ACE_LIB_TEXT ("(\'%s\') failed, errno=")
+                            ACE_LIB_TEXT ("%d: %s\n"),
                             name->c_str (),
                             errno,
                             this->error ()->c_str ()));
@@ -163,11 +170,18 @@
                   ACE_OS::strcat (aix_pathname, ACE_LIB_TEXT ("(shr.o)"));
                   open_mode |= RTLD_MEMBER;
 
-                  if (ACE::debug ())
+                  if (ACE::debug () > 1)
+                    {
                     ACE_DEBUG ((LM_DEBUG,
-                                ACE_LIB_TEXT ("(%P|%t) ACE_DLL_Handle::open: ")
-                                ACE_LIB_TEXT ("calling dlopen on ")
-                                ACE_LIB_TEXT ("\"%s\"\n"), aix_pathname));
+                                  ACE_LIB_TEXT ("ACE (%P|%t) DLL_Handle::open ")
+                                  ACE_LIB_TEXT ("(\"%s\", 0x%x) -> %s: %s\n"), 
+                                  aix_pathname, 
+                                  open_mode,
+                                  ACE_LIB_TEXT ((this->handle_ != ACE_SHLIB_INVALID_HANDLE) 
+                                                ? "succeeded"
+                                                : "failed"),
+                                  this->error()->c_str()));
+                    }
 
                   this->handle_ = ACE_OS::dlopen (aix_pathname, open_mode);
                   if (this->handle_ != ACE_SHLIB_INVALID_HANDLE)
@@ -181,9 +195,9 @@
                   // should we continue at all?
                   if (ACE::debug () && (errno != 0) && (errno != ENOENT))
                     ACE_ERROR ((LM_ERROR,
-                                ACE_LIB_TEXT ("(%P|%t) ACE_DLL_Handle::open: ")
-                                ACE_LIB_TEXT ("Attempt to open \'%s\' failed")
-                                ACE_LIB_TEXT (" (%d): %s\n"),
+                                ACE_LIB_TEXT ("ACE (%P|%t) DLL_Handle::open ")
+                                ACE_LIB_TEXT ("(\'%s\') failed, errno=")
+                                ACE_LIB_TEXT ("%d: %s\n"),
                                 name->c_str (),
                                 errno,
                                 this->error ()->c_str ()));
@@ -198,9 +212,8 @@
             {
               if (ACE::debug ())
                 ACE_ERROR ((LM_ERROR,
-                            ACE_LIB_TEXT ("(%P|%t) DLL_Handle::open: ")
-                            ACE_LIB_TEXT ("Invalid handle while ")
-                            ACE_LIB_TEXT ("opening DLL \"%s\": %s\n"),
+                            ACE_LIB_TEXT ("ACE (%P|%t) DLL_Handle::open (\"%s\"): ")
+                            ACE_LIB_TEXT ("Invalid handle error: %s\n"),
                             this->dll_name_,
                             this->error ()->c_str ()));
 
Index: tests/Service_Config_Test.cpp
===================================================================
--- ACE_wrappers/tests/Service_Config_Test.cpp	(revision 73853)
+++ ACE_wrappers/tests/Service_Config_Test.cpp	(working copy)
@@ -175,12 +175,25 @@
 testLimits (int , ACE_TCHAR *[])
 {
   static const ACE_TCHAR *svc_desc1 =
-    ACE_TEXT ("dynamic Test_Object_1_More Service_Object * ")
+#if ACE_USES_CLASSIC_SVC_CONF == 1  
+  ACE_TEXT ("dynamic Test_Object_1_More Service_Object * ")
     ACE_TEXT ("  Service_Config_DLL:_make_Service_Config_DLL() \"Test_Object_1_More\"");
+#else
+  ACE_TEXT ("<dynamic id=\"Test_Object_1_More_Service_Object\" type=\"Service_Object\">")
+    ACE_TEXT ("  <initializer init=\"_make_Service_Config_DLL\" path=\"Service_Config_DLL\" params=\"Test_Object_1_More\"/>")
+  ACE_TEXT ("</dynamic>");
+#endif
 
+
   static const ACE_TCHAR *svc_desc2 =
+#if ACE_USES_CLASSIC_SVC_CONF == 1  
     ACE_TEXT ("dynamic Test_Object_2_More Service_Object * ")
     ACE_TEXT ("  Service_Config_DLL:_make_Service_Config_DLL() \"Test_Object_2_More\"");
+#else
+  ACE_TEXT ("<dynamic id=\"Test_Object_2_More_Service_Object\" type=\"Service_Object\">")
+    ACE_TEXT ("  <initializer init=\"_make_Service_Config_DLL\" path=\"Service_Config_DLL\" params=\"Test_Object_2_More\"/>")
+  ACE_TEXT ("</dynamic>");
+#endif
 
   // Ensure enough room for just one ...
   ACE_Service_Gestalt one (1, true);
