Index: orbsvcs/orbsvcs/FaultTolerance/FT_Service_Callbacks.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/orbsvcs/orbsvcs/FaultTolerance/FT_Service_Callbacks.cpp,v
retrieving revision 1.28
diff -u -r1.28 FT_Service_Callbacks.cpp
--- orbsvcs/orbsvcs/FaultTolerance/FT_Service_Callbacks.cpp	27 Jan 2005 19:44:46 -0000	1.28
+++ orbsvcs/orbsvcs/FaultTolerance/FT_Service_Callbacks.cpp	2 Jun 2005 16:47:38 -0000
@@ -66,12 +66,10 @@
 
 }
 
-CORBA::Boolean
+TAO_Service_Callbacks::Profile_Equivalence
 TAO_FT_Service_Callbacks::is_profile_equivalent (const TAO_Profile *this_p,
                                                  const TAO_Profile *that_p)
 {
-  // At this point we assume that all the checks for other things
-  // within the profiles have been satisfied
   const TAO_Tagged_Components &this_comp =
     this_p->tagged_components ();
 
@@ -81,54 +79,58 @@
   IOP::TaggedComponent this_tc, that_tc;
   this_tc.tag = that_tc.tag = IOP::TAG_FT_GROUP;
 
-  if ((this_comp.get_component (this_tc) == 1) &&
-      (that_comp.get_component (that_tc) == 1))
+  if (this_comp.get_component (this_tc) == 1)
     {
-
-      TAO_InputCDR this_cdr (reinterpret_cast<const char*> (this_tc.component_data.get_buffer ()),
+      if (that_comp.get_component (that_tc) == 1)
+        {
+          TAO_InputCDR this_cdr (reinterpret_cast<const char*> (this_tc.component_data.get_buffer ()),
                              this_tc.component_data.length ());
 
-      TAO_InputCDR that_cdr (reinterpret_cast<const char*> (that_tc.component_data.get_buffer ()),
+          TAO_InputCDR that_cdr (reinterpret_cast<const char*> (that_tc.component_data.get_buffer ()),
                              that_tc.component_data.length ());
 
-      CORBA::Boolean this_byte_order;
-      CORBA::Boolean that_byte_order;
+          CORBA::Boolean this_byte_order;
+          CORBA::Boolean that_byte_order;
 
-      if (this_cdr >> ACE_InputCDR::to_boolean (this_byte_order) == 0 ||
-          that_cdr >> ACE_InputCDR::to_boolean (that_byte_order) == 0)
-        {
-          // return no equivalent
-          return 0;
-        }
+          if (this_cdr >> ACE_InputCDR::to_boolean (this_byte_order) == 0 ||
+              that_cdr >> ACE_InputCDR::to_boolean (that_byte_order) == 0)
+            {
+              // Have tags but full of garbage - not equivalent
+              return TAO_Service_Callbacks::NOT_EQUIVALENT;
+            }
 
-      this_cdr.reset_byte_order (static_cast<int> (this_byte_order));
-      that_cdr.reset_byte_order (static_cast<int> (that_byte_order));
+          this_cdr.reset_byte_order (static_cast<int> (this_byte_order));
+          that_cdr.reset_byte_order (static_cast<int> (that_byte_order));
 
 
-      FT::TagFTGroupTaggedComponent this_group_component;
-      FT::TagFTGroupTaggedComponent that_group_component;
+          FT::TagFTGroupTaggedComponent this_group_component;
+          FT::TagFTGroupTaggedComponent that_group_component;
+
+          this_cdr >> this_group_component;
+          that_cdr >> that_group_component;
+
+          // check if domain id and group id are the same
+          if ((ACE_OS::strcmp (this_group_component.group_domain_id,
+                               that_group_component.group_domain_id) == 0) &&
+              (this_group_component.object_group_id ==
+               that_group_component.object_group_id))
+             {
+               // Both have matching tags - true
+               return TAO_Service_Callbacks::EQUIVALENT;
+             }
+        }
 
-      this_cdr >> this_group_component;
-      that_cdr >> that_group_component;
-
-      // check if domain id and group id are the same
-      if ((ACE_OS::strcmp (this_group_component.group_domain_id,
-                           that_group_component.group_domain_id) == 0) &&
-          (this_group_component.object_group_id ==
-           that_group_component.object_group_id))
-         {
-           return 1;
-         }
+      return TAO_Service_Callbacks::NOT_EQUIVALENT;
     }
-  // If both the profiles did not have a tag group component then they
-  // are equivalent.
-  else if ((this_comp.get_component (this_tc) == 0) &&
-           (that_comp.get_component (that_tc) == 0))
+  else
     {
-      return 1;
+      if (that_comp.get_component (that_tc) == 1)
+        {
+          return TAO_Service_Callbacks::NOT_EQUIVALENT;
+        }
     }
 
-  return 0;
+  return TAO_Service_Callbacks::DONT_KNOW;
 }
 
 CORBA::ULong
Index: orbsvcs/orbsvcs/FaultTolerance/FT_Service_Callbacks.h
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/orbsvcs/orbsvcs/FaultTolerance/FT_Service_Callbacks.h,v
retrieving revision 1.21
diff -u -r1.21 FT_Service_Callbacks.h
--- orbsvcs/orbsvcs/FaultTolerance/FT_Service_Callbacks.h	5 Jan 2005 19:31:54 -0000	1.21
+++ orbsvcs/orbsvcs/FaultTolerance/FT_Service_Callbacks.h	2 Jun 2005 16:47:38 -0000
@@ -50,7 +50,8 @@
   virtual CORBA::Boolean object_is_nil (CORBA::Object_ptr obj);
 
   /// Check for equivalency of the two profiles
-  virtual CORBA::Boolean is_profile_equivalent (const TAO_Profile *,
+  virtual TAO_Service_Callbacks::Profile_Equivalence is_profile_equivalent (
+                                                const TAO_Profile *,
                                                 const TAO_Profile *);
   /// Calculate the hash
   virtual CORBA::ULong hash_ft (TAO_Profile *p,
Index: tao/ORB_Core.h
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/ORB_Core.h,v
retrieving revision 1.263
diff -u -r1.263 ORB_Core.h
--- tao/ORB_Core.h	26 May 2005 10:43:03 -0000	1.263
+++ tao/ORB_Core.h	2 Jun 2005 16:47:40 -0000
@@ -835,7 +835,8 @@
   /**
    * For details on how this is used please see the FT service
    */
-  CORBA::Boolean is_profile_equivalent (const TAO_Profile *this_p,
+  TAO_Service_Callbacks::Profile_Equivalence is_profile_equivalent (
+                                        const TAO_Profile *this_p,
                                         const TAO_Profile *that_p);
 
   /// Hook for the services to determine the <hash> value of a
Index: tao/ORB_Core.i
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/ORB_Core.i,v
retrieving revision 1.138
diff -u -r1.138 ORB_Core.i
--- tao/ORB_Core.i	26 May 2005 10:43:03 -0000	1.138
+++ tao/ORB_Core.i	2 Jun 2005 16:47:40 -0000
@@ -124,11 +124,12 @@
 }
 
 
-ACE_INLINE CORBA::Boolean
+ACE_INLINE TAO_Service_Callbacks::Profile_Equivalence
 TAO_ORB_Core::is_profile_equivalent (const TAO_Profile *this_p,
                                      const TAO_Profile *that_p)
 {
-  CORBA::Boolean retval = 1;
+  TAO_Service_Callbacks::Profile_Equivalence retval 
+    = TAO_Service_Callbacks::DONT_KNOW;
 
   if (this->ft_service_.service_callback ())
     {
Index: tao/Profile.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/Profile.cpp,v
retrieving revision 1.74
diff -u -r1.74 Profile.cpp
--- tao/Profile.cpp	21 May 2005 18:37:27 -0000	1.74
+++ tao/Profile.cpp	2 Jun 2005 16:47:41 -0000
@@ -700,17 +700,27 @@
 CORBA::Boolean
 TAO_Profile::is_equivalent (const TAO_Profile *other)
 {
-  return
-    other != 0
-    && this->tag () == other->tag ()
-    && this->version_ == other->version ()
-    && this->endpoint_count () == other->endpoint_count ()
-    && this->object_key () == other->object_key ()
-    && this->do_is_equivalent (other)
-    && this->is_equivalent_hook (other);
+  CORBA::Boolean result = 0;
+  if (other)
+    {
+      TAO_Service_Callbacks::Profile_Equivalence callback
+        = this->is_equivalent_hook (other);
+      switch (callback)
+        {
+          case TAO_Service_Callbacks::DONT_KNOW:
+            return this->tag () == other->tag ()
+                && this->version_ == other->version ()
+                && this->endpoint_count () == other->endpoint_count ()
+                && this->object_key () == other->object_key ()
+                && this->do_is_equivalent (other);
+          case TAO_Service_Callbacks::EQUIVALENT:
+            result = 1;
+        }
+    }       
+  return result;
 }
 
-CORBA::Boolean
+TAO_Service_Callbacks::Profile_Equivalence
 TAO_Profile::is_equivalent_hook (const TAO_Profile *other)
 {
   // Allow services to apply their own definition of "equivalence."
@@ -832,13 +842,13 @@
   return (CORBA::Boolean) (op == 0 ? 0 : this->body_ == op->body_);
 }
 
-CORBA::Boolean
+TAO_Service_Callbacks::Profile_Equivalence
 TAO_Unknown_Profile::is_equivalent_hook (const TAO_Profile * /* other */)
 {
   // Override the default implementation since we don't need the
   // additional checks it performs.
 
-  return 1;
+  return TAO_Service_Callbacks::DONT_KNOW;
 }
 
 CORBA::ULong
Index: tao/Profile.h
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/Profile.h,v
retrieving revision 1.57
diff -u -r1.57 Profile.h
--- tao/Profile.h	26 May 2005 09:22:36 -0000	1.57
+++ tao/Profile.h	2 Jun 2005 16:47:41 -0000
@@ -23,6 +23,7 @@
 
 #include "tao/GIOP_Message_Version.h"
 #include "tao/Refcounted_ObjectKey.h"
+#include "tao/Service_Callbacks.h"
 
 class ACE_Lock;
 class TAO_MProfile;
@@ -275,7 +276,8 @@
    * method in that it has a default implementation that may or not be
    * applicable to all TAO_Profile subclasses.
    */
-  virtual CORBA::Boolean is_equivalent_hook (const TAO_Profile * other);
+  virtual TAO_Service_Callbacks::Profile_Equivalence is_equivalent_hook (
+                                            const TAO_Profile * other);
 
   CORBA::ULong hash_service_i (CORBA::ULong m);
 
@@ -389,7 +391,8 @@
 protected:
 
   virtual CORBA::Boolean do_is_equivalent (const TAO_Profile* other_profile);
-  virtual CORBA::Boolean is_equivalent_hook (const TAO_Profile* other_profile);
+  virtual TAO_Service_Callbacks::Profile_Equivalence is_equivalent_hook (
+                                           const TAO_Profile* other_profile);
 
 private:
   virtual void create_profile_body (TAO_OutputCDR &encap) const;
Index: tao/Service_Callbacks.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/Service_Callbacks.cpp,v
retrieving revision 1.13
diff -u -r1.13 Service_Callbacks.cpp
--- tao/Service_Callbacks.cpp	4 Jan 2005 16:48:15 -0000	1.13
+++ tao/Service_Callbacks.cpp	2 Jun 2005 16:47:41 -0000
@@ -42,11 +42,11 @@
 }
 
 
-CORBA::Boolean
+TAO_Service_Callbacks::Profile_Equivalence
 TAO_Service_Callbacks::is_profile_equivalent (const TAO_Profile *,
                                               const TAO_Profile *)
 {
-  return 0;
+  return DONT_KNOW;
 }
 
 CORBA::ULong
Index: tao/Service_Callbacks.h
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/Service_Callbacks.h,v
retrieving revision 1.16
diff -u -r1.16 Service_Callbacks.h
--- tao/Service_Callbacks.h	12 Apr 2005 10:57:01 -0000	1.16
+++ tao/Service_Callbacks.h	2 Jun 2005 16:47:41 -0000
@@ -58,6 +58,14 @@
 {
 
 public:
+  
+  enum Profile_Equivalence
+  {
+    EQUIVALENT,
+    NOT_EQUIVALENT,
+    DONT_KNOW
+  };
+    
 
   /// Dtor
   virtual ~TAO_Service_Callbacks (void);
@@ -80,8 +88,8 @@
 
 
   /// Check for equivalency of the two profiles
-  virtual CORBA::Boolean is_profile_equivalent (const TAO_Profile *,
-                                                const TAO_Profile *);
+  virtual Profile_Equivalence is_profile_equivalent (const TAO_Profile *,
+                                                     const TAO_Profile *);
 
   /// Calculate the hash value..
   virtual CORBA::ULong hash_ft (TAO_Profile *p,
