Index: templates/wb26.mpd
===================================================================
--- templates/wb26.mpd	(revision 1281)
+++ templates/wb26.mpd	(working copy)
@@ -21,7 +21,7 @@
 		<link>
 			<name><%source_file%></name>
 			<type>1</type>
-			<location><%cwd%>/<%source_file%></location>
+			<location><%full_path(source_file)%></location>
 		</link>
 <%endfor%>
 <%endif%>
@@ -30,7 +30,7 @@
                 <link>
                         <name><%header_file%></name>
                         <type>1</type>
-                        <location><%cwd%>/<%header_file%></location>
+                        <location><%full_path(header_file)%></location>
                 </link>
 <%endfor%>
 <%endif%>
@@ -39,7 +39,7 @@
                 <link>
                         <name><%inline_file%></name>
                         <type>1</type>
-                        <location><%cwd%>/<%inline_file%></location>
+                        <location><%full_path(inline_file)%></location>
                 </link>
 <%endfor%>
 <%if(template_files)%>
@@ -47,7 +47,7 @@
                 <link>
                         <name><%template_file%></name>
                         <type>1</type>
-                        <location><%cwd%>/<%template_file%></location>
+                        <location><%full_path(template_file)%></location>
                 </link>
 <%endfor%>
 <%endif%>
Index: templates/wb26wrproject.mpd
===================================================================
--- templates/wb26wrproject.mpd	(revision 1281)
+++ templates/wb26wrproject.mpd	(working copy)
@@ -26,7 +26,7 @@
                 <stringAttribute value="-I$(WIND_BASE)/target/h"/>
                 <stringAttribute value="-I$(WIND_BASE)/target/h/wrn/coreip"/>
 <%foreach(includes)%>
-                <stringAttribute value="-I<%include%>"/>
+                <stringAttribute value="-I<%full_path(include)%>"/>
 <%endfor%>
             </listAttribute>
             <listAttribute key="BLD::Info|Macros">
Index: modules/TemplateParser.pm
===================================================================
--- modules/TemplateParser.pm	(revision 1281)
+++ modules/TemplateParser.pm	(working copy)
@@ -64,6 +64,7 @@
                  'has_extension'   => 5,
                  'keyname_used'    => 0,
                  'scope'           => 0,
+                 'full_path'       => 2,
                 );
 
 my(%target_type_vars) = ('type_is_static'   => 1,
@@ -1449,6 +1450,41 @@
 }
 
 
+sub perform_full_path {
+  my($self) = shift;
+  my($value) = shift;
+
+  #Expand all defined env vars
+  $value =~ s/\$\((\w+)\)/$ENV{$1} || '$(' . $1 . ')'/ge;
+
+  #If we expanded all env vars, get absolute path
+  if ($value =~ /\$\(\w+\)/) {
+    $self->{'error_in_handle'} = "<%full_path%> couldn't expand environment " 
+        . "variables in $value";
+    return $value;
+  }
+
+  my($dir) = $self->mpc_dirname($value);
+  if (!-e $dir) {
+    $self->{'error_in_handle'} = "The directory $dir does not exist -- can't ".
+        "convert to an absolute path in <%full_path%>\n";
+    return $value;
+  }
+  
+  return Cwd::abs_path($dir) . ($self->convert_slashes() ? '\\' : '/')
+    . $self->mpc_basename($value);
+}
+
+
+sub handle_full_path {
+  my($self) = shift;
+  my($name) = shift;
+  my($val)  = $self->get_value_with_default($name);
+
+  $self->append_current($self->perform_full_path($val));
+}
+
+
 sub evaluate_nested_functions {
   my($self) = shift;
   my($name) = shift;
@@ -1759,7 +1795,11 @@
           }
           else {
             my($func) = 'handle_' . $name;
+            undef $self->{'error_in_handle'};
             $self->$func($val);
+            if ($self->{'error_in_handle'}) {
+              $errorString = $self->{'error_in_handle'};
+            }
           }
         }
       }
