Index: Active_Policy_Strategies.inl
===================================================================
--- Active_Policy_Strategies.inl	(revision 80257)
+++ Active_Policy_Strategies.inl	(working copy)
@@ -56,6 +56,24 @@
     {
       return this->servant_retention_strategy_;
     }
+
+    ACE_INLINE
+    Active_Policy_Strategies_Cleanup_Guard::~Active_Policy_Strategies_Cleanup_Guard (void)
+    {
+      if (this->ptr_)
+        {
+          this->ptr_->cleanup ();
+        }
+    }
+
+    ACE_INLINE
+    Active_Policy_Strategies *
+    Active_Policy_Strategies_Cleanup_Guard::_retn (void)
+    {
+      Active_Policy_Strategies *temp = this->ptr_;
+      this->ptr_ = 0;
+      return temp;
+    }
   }
 }
 
Index: Active_Policy_Strategies.h
===================================================================
--- Active_Policy_Strategies.h	(revision 80257)
+++ Active_Policy_Strategies.h	(working copy)
@@ -91,6 +91,22 @@
       IdUniquenessStrategyFactory *id_uniqueness_strategy_factory_;
       IdAssignmentStrategyFactory *id_assignment_strategy_factory_;
     };
+
+    /**
+     * This class quards the cleanup of strategies if something went wrong
+     * in the code that called Active_Policy_Strategies::update().
+     */
+    class Active_Policy_Strategies_Cleanup_Guard
+    {
+    public:
+      Active_Policy_Strategies_Cleanup_Guard (Active_Policy_Strategies *p) : ptr_ (p) {}
+      ~Active_Policy_Strategies_Cleanup_Guard (void);
+
+      Active_Policy_Strategies *_retn (void);
+
+    private:
+      Active_Policy_Strategies *ptr_;
+    };
   }
 }
 
Index: Root_POA.cpp
===================================================================
--- Root_POA.cpp	(revision 80257)
+++ Root_POA.cpp	(working copy)
@@ -231,8 +231,9 @@
     servant_for_key_to_object_ (0)
 {
   // Since we are keeping a reference to a POAManager, we need to
-  // increment the reference count
-  this->poa_manager_._add_ref();
+  // increment the reference count but we do this safely.
+  PortableServer::POAManager_var pm_guard (
+    PortableServer::POAManager::_duplicate(&this->poa_manager_));
 
   // Parse the policies that are used in the critical path in
   // a cache.
@@ -268,6 +269,8 @@
   // Set the active strategies to be used by this POA
   this->active_policy_strategies_.update (this->cached_policies_,
                                           this);
+  TAO::Portable_Server::Active_Policy_Strategies_Cleanup_Guard aps_cleanup_guard (
+    &this->active_policy_strategies_);
 
   // Set the folded name of this POA.
   this->set_folded_name (parent);
@@ -309,6 +312,10 @@
                                           this->system_name_.in ());
       throw;
     }
+
+  // Now when everything is fine we can release the quards.
+  pm_guard._retn ();
+  aps_cleanup_guard._retn ();
 }
 
 TAO_Root_POA::~TAO_Root_POA (void)
