Index: OS_Memory.h
===================================================================
--- OS_Memory.h	(.../vendor/ACE/current/ace/OS_Memory.h)	(revision 463)
+++ OS_Memory.h	(.../trunk/ACE_wrappers/ace/OS_Memory.h)	(revision 463)
@@ -115,9 +115,9 @@
 #    if (HPUX_VERS >= 1100)
 #      if ((__HP_aCC < 32500 && !defined (RWSTD_NO_NAMESPACE)) || \
            defined (ACE_USES_STD_NAMESPACE_FOR_STDCPP_LIB))
-#        define ACE_bad_alloc std::bad_alloc
-#        define ACE_nothrow   std::nothrow
-#        define ACE_nothrow_t std::nothrow_t
+#        define ACE_bad_alloc ::std::bad_alloc
+#        define ACE_nothrow   ::std::nothrow
+#        define ACE_nothrow_t ::std::nothrow_t
 #      else
 #        define ACE_bad_alloc bad_alloc
 #        define ACE_nothrow   nothrow
@@ -125,9 +125,9 @@
 #      endif /* __HP_aCC */
 #    elif ((__HP_aCC <  12500 && !defined (RWSTD_NO_NAMESPACE)) || \
            defined (ACE_USES_STD_NAMESPACE_FOR_STDCPP_LIB))
-#      define ACE_bad_alloc std::bad_alloc
-#      define ACE_nothrow   std::nothrow
-#      define ACE_nothrow_t std::nothrow_t
+#      define ACE_bad_alloc ::std::bad_alloc
+#      define ACE_nothrow   ::std::nothrow
+#      define ACE_nothrow_t ::std::nothrow_t
 #    else
 #      define ACE_bad_alloc bad_alloc
 #      define ACE_nothrow   nothrow
@@ -143,11 +143,11 @@
 #        define ACE_throw_bad_alloc throw ACE_bad_alloc ("no more memory")
 #      else
 #        include /**/ <new>
-#        define ACE_bad_alloc std::bad_alloc
+#        define ACE_bad_alloc ::std::bad_alloc
 #        if defined (ACE_HAS_NEW_NOTHROW)
 #          if defined (ACE_USES_STD_NAMESPACE_FOR_STDCPP_LIB)
-#            define ACE_nothrow   std::nothrow
-#            define ACE_nothrow_t std::nothrow_t
+#            define ACE_nothrow   ::std::nothrow
+#            define ACE_nothrow_t ::std::nothrow_t
 #          else
 #            define ACE_nothrow   nothrow
 #            define ACE_nothrow_t nothrow_t
@@ -158,10 +158,10 @@
 #  elif defined (ACE_USES_STD_NAMESPACE_FOR_STDCPP_LIB)
 #    include /**/ <new>
 #    if !defined (ACE_bad_alloc)
-#      define ACE_bad_alloc std::bad_alloc
+#      define ACE_bad_alloc ::std::bad_alloc
 #    endif
-#    define ACE_nothrow   std::nothrow
-#    define ACE_nothrow_t std::nothrow_t
+#    define ACE_nothrow   ::std::nothrow
+#    define ACE_nothrow_t ::std::nothrow_t
      // MFC changes the behavior of operator new at all MSVC versions from 6 up.
 #    if defined (ACE_HAS_MFC) && (ACE_HAS_MFC == 1)
 #      define ACE_throw_bad_alloc AfxThrowMemoryException ()
