--- OS_NS_Thread.inl	2009-01-09 13:15:39.799125000 -0800
+++ /cygdrive/c/temp/ACE/ACE_wrappers/ace/OS_NS_Thread.inl	2007-07-14 03:48:10.000000000 -0700
@@ -472,8 +472,6 @@
   if (*event == 0)
     ACE_FAIL_RETURN (-1);
 
-  // Make sure to set errno to ERROR_ALREADY_EXISTS if necessary.
-  ACE_OS::set_errno_to_last_error ();
   return 0;
 #else  /* ACE_WIN32 */
   return ACE_OS::event_init (event,
@@ -1708,11 +1706,7 @@
     ACE_FAIL_RETURN (-1);
   /* NOTREACHED */
   else
-    {
-      // Make sure to set errno to ERROR_ALREADY_EXISTS if necessary.
-      ACE_OS::set_errno_to_last_error ();
-      return 0;
-    }
+    return 0;
 #    else /* ACE_USES_WINCE_SEMA_SIMULATION */
   int result = -1;
 
@@ -1796,11 +1790,7 @@
     ACE_FAIL_RETURN (-1);
   /* NOTREACHED */
   else
-    {
-      // Make sure to set errno to ERROR_ALREADY_EXISTS if necessary.
-      ACE_OS::set_errno_to_last_error ();
-      return 0;
-    }
+    return 0;
 #   else /* ACE_USES_WINCE_SEMA_SIMULATION */
   int result = -1;
 
