Index: TAO/tao/IIOP_Acceptor.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/IIOP_Acceptor.cpp,v
retrieving revision 1.109
diff -u -r1.109 IIOP_Acceptor.cpp
--- TAO/tao/IIOP_Acceptor.cpp	23 Dec 2005 10:39:10 -0000	1.109
+++ TAO/tao/IIOP_Acceptor.cpp	10 Mar 2006 00:46:08 -0000
@@ -989,7 +989,7 @@
 #endif /* TAO_NO_COPY_OCTET_SEQUENCES == 1 */
 
   CORBA::Octet major;
-  CORBA::Octet minor;
+  CORBA::Octet minor = CORBA::Octet();;
 
   // Read the version. We just read it here. We don't*do any*
   // processing.
Index: TAO/tao/GIOP_Message_Generator_Parser_10.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/GIOP_Message_Generator_Parser_10.cpp,v
retrieving revision 1.34
diff -u -r1.34 GIOP_Message_Generator_Parser_10.cpp
--- TAO/tao/GIOP_Message_Generator_Parser_10.cpp	2 Nov 2005 07:13:02 -0000	1.34
+++ TAO/tao/GIOP_Message_Generator_Parser_10.cpp	10 Mar 2006 00:45:07 -0000
@@ -341,7 +341,7 @@
 
   request.request_id (req_id);
 
-  CORBA::Octet response_flags;
+  CORBA::Octet response_flags = CORBA::Octet();;
   hdr_status = hdr_status && input.read_octet (response_flags);
   request.response_expected ((response_flags != 0));
 
Index: TAO/tao/GIOP_Message_Generator_Parser_12.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/GIOP_Message_Generator_Parser_12.cpp,v
retrieving revision 1.31
diff -u -r1.31 GIOP_Message_Generator_Parser_12.cpp
--- TAO/tao/GIOP_Message_Generator_Parser_12.cpp	2 Nov 2005 07:13:02 -0000	1.31
+++ TAO/tao/GIOP_Message_Generator_Parser_12.cpp	10 Mar 2006 00:45:32 -0000
@@ -266,7 +266,7 @@
 
   request.request_id (req_id);
 
-  CORBA::Octet response_flags;
+  CORBA::Octet response_flags = CORBA::Octet();;
   hdr_status = hdr_status && input.read_octet (response_flags);
 
   request.response_expected ((response_flags > 0));
Index: TAO/tao/GIOP_Message_Lite.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/GIOP_Message_Lite.cpp,v
retrieving revision 1.91
diff -u -r1.91 GIOP_Message_Lite.cpp
--- TAO/tao/GIOP_Message_Lite.cpp	23 Feb 2006 17:20:16 -0000	1.91
+++ TAO/tao/GIOP_Message_Lite.cpp	10 Mar 2006 00:45:49 -0000
@@ -1347,7 +1347,7 @@
 
   request.request_id (req_id);
 
-   CORBA::Octet response_flags;
+  CORBA::Octet response_flags = CORBA::Octet();;
   hdr_status = hdr_status && input.read_octet (response_flags);
   request.response_expected ((response_flags != 0));
Index: TAO/tao/AnyTypeCode/skip.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/AnyTypeCode/skip.cpp,v
retrieving revision 1.3
diff -u -r1.3 skip.cpp
--- TAO/tao/AnyTypeCode/skip.cpp	3 Nov 2005 17:38:44 -0000	1.3
+++ TAO/tao/AnyTypeCode/skip.cpp	10 Mar 2006 01:59:54 -0000
@@ -368,13 +368,13 @@
   ACE_CHECK_RETURN (TAO::TRAVERSE_STOP);
 
   // Save the discriminator value in a temporary variable...
-  CORBA::Short short_v;
-  CORBA::UShort ushort_v;
-  CORBA::Long long_v;
-  CORBA::ULong ulong_v;
-  CORBA::ULong enum_v;
-  CORBA::Char char_v;
-  CORBA::WChar wchar_v;
+  CORBA::Short short_v = CORBA::Short();
+  CORBA::UShort ushort_v = CORBA::UShort();
+  CORBA::Long long_v = CORBA::Long();
+  CORBA::ULong ulong_v = CORBA::ULong();
+  CORBA::ULong enum_v = CORBA::ULong();
+  CORBA::Char char_v = CORBA::Char();
+  CORBA::WChar wchar_v = CORBA::WChar();
   CORBA::Boolean boolean_v = false;;
 
   switch (kind)
Index: TAO/tao/AnyTypeCode/append.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tao/AnyTypeCode/append.cpp,v
retrieving revision 1.5
diff -u -r1.5 append.cpp
--- TAO/tao/AnyTypeCode/append.cpp	3 Nov 2005 17:38:44 -0000	1.5
+++ TAO/tao/AnyTypeCode/append.cpp	10 Mar 2006 01:58:57 -0000
@@ -415,13 +415,13 @@
   ACE_CHECK_RETURN (TAO::TRAVERSE_STOP);
 
   // Save the discriminator value in a temporary variable...
-  CORBA::Short short_v;
-  CORBA::UShort ushort_v;
-  CORBA::Long long_v;
-  CORBA::ULong ulong_v;
-  CORBA::ULong enum_v;
-  CORBA::Char char_v;
-  CORBA::WChar wchar_v;
+  CORBA::Short short_v = CORBA::Short();
+  CORBA::UShort ushort_v = CORBA::UShort();
+  CORBA::Long long_v = CORBA::Long();
+  CORBA::ULong ulong_v = CORBA::ULong();
+  CORBA::ULong enum_v = CORBA::ULong();
+  CORBA::Char char_v = CORBA::Char();
+  CORBA::WChar wchar_v = CORBA::WChar();
   CORBA::Boolean boolean_v = false;
 
   switch (kind)

