Index: Dev_Poll_Reactor.cpp
===================================================================
--- Dev_Poll_Reactor.cpp	(revision 88800)
+++ Dev_Poll_Reactor.cpp	(working copy)
@@ -1281,13 +1281,13 @@
       // With epoll, events are registered with oneshot, so the handle is
       // effectively suspended; future calls to epoll_wait() will select
       // the next event, so they're not managed here.
-      // The hitch to this is that the notify handler must always be resumed
-      // immediately, before letting go of the guard. Else it's possible to
-      // get into a state where all handles, including the notify pipe, are
-      // suspended and that means the wait thread can't be interrupted.
-      info->suspended = true;
-      if (eh == this->notify_handler_)
-        this->resume_handler_i (handle);
+      // The hitch to this is that the notify handler is always registered
+      // WITHOUT oneshot and is never suspended/resumed. This avoids endless
+      // notify loops caused by the notify handler requiring a resumption
+      // which requires the token, which requires a notify, etc. described
+      // in Bugzilla 3714. So, never suspend the notify handler.
+      if (eh != this->notify_handler_)
+        info->suspended = true;
 #endif /* ACE_HAS_DEV_POLL */
 
       int status = 0;   // gets callback status, below.
@@ -1306,6 +1306,9 @@
         // if callback returns > 0. We come back with either 0 or < 0.
         status = this->upcall (eh, callback, handle);
 
+        if (eh == this->notify_handler_)
+          return status == 0 ? 1 : -1;
+
         // All state in the handler repository may have changed during the
         // upcall while other threads had the token. Thus, reacquire the
         // token and evaluate what's needed. If the upcalled handler is still
@@ -1428,8 +1431,13 @@
      ACE_OS::memset (&epev, 0, sizeof (epev));
      static const int op = EPOLL_CTL_ADD;
 
-     epev.events  = this->reactor_mask_to_poll_event (mask) | EPOLLONESHOT;
      epev.data.fd = handle;
+     epev.events  = this->reactor_mask_to_poll_event (mask);
+     // All but the notify handler get registered with oneshot to facilitate
+     // auto suspend before the upcall. See dispatch_io_event for more
+     // information.
+     if (event_handler != this->notify_handler_)
+       epev.events |= EPOLLONESHOT;
 
      if (::epoll_ctl (this->poll_fd_, op, handle, &epev) == -1)
        {
