diff -crB XtReactor.orig/XtReactor.cpp XtReactor/XtReactor.cpp
*** XtReactor.orig/XtReactor.cpp	Tue Sep 16 11:35:44 2008
--- XtReactor/XtReactor.cpp	Mon Jan 11 13:34:28 2010
***************
*** 203,276 ****
    // Make sure we have a valid context
    ACE_ASSERT (this->context_ != 0);
  
    int result = ACE_Select_Reactor::register_handler_i (handle,
                                                         handler, mask);
    if (result == -1)
      return -1;
  
!   int condition = 0;
! 
! #if !defined ACE_WIN32
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::READ_MASK))
!     ACE_SET_BITS (condition, XtInputReadMask);
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::WRITE_MASK))
!     ACE_SET_BITS (condition, XtInputWriteMask);
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::EXCEPT_MASK))
!     ACE_SET_BITS (condition, XtInputExceptMask);
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::ACCEPT_MASK))
!     ACE_SET_BITS (condition, XtInputReadMask);
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::CONNECT_MASK)){
!       ACE_SET_BITS (condition, XtInputWriteMask); // connected, you may write
!       ACE_SET_BITS (condition, XtInputReadMask);  // connected, you have data/err
!   }
! #else
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::READ_MASK))
!     ACE_SET_BITS (condition, XtInputReadWinsock);
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::WRITE_MASK))
!     ACE_SET_BITS (condition, XtInputWriteWinsock);
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::EXCEPT_MASK))
!     ACE_NOTSUP_RETURN(-1);
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::ACCEPT_MASK))
!     ACE_SET_BITS (condition, XtInputReadWinsock);
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::CONNECT_MASK)){
!       ACE_SET_BITS (condition, XtInputWriteWinsock); // connected, you may write
!       ACE_SET_BITS (condition, XtInputReadWinsock);  // connected, you have data/err
!   }
! #endif /* !ACE_WIN32 */
! 
!   if (condition != 0)
!     {
!       ACE_XtReactorID *XtID = this->ids_;
! 
!       while(XtID)
!         {
!           if (XtID->handle_ == handle)
!             {
!               ::XtRemoveInput (XtID->id_);
! 
!               XtID->id_ = ::XtAppAddInput (this->context_,
!                                            (int) handle,
!                                            (XtPointer) condition,
!                                            InputCallbackProc,
!                                            (XtPointer) this);
!               return 0;
!             }
!           else
!             XtID = XtID->next_;
!         }
! 
!       ACE_NEW_RETURN (XtID,
!                       ACE_XtReactorID,
!                       -1);
!       XtID->next_ = this->ids_;
!       XtID->handle_ = handle;
!       XtID->id_ = ::XtAppAddInput (this->context_,
!                                   (int) handle,
!                                   (XtPointer) condition,
!                                   InputCallbackProc,
!                                   (XtPointer) this);
!       this->ids_ = XtID;
!     }
    return 0;
  }
  
--- 203,220 ----
    // Make sure we have a valid context
    ACE_ASSERT (this->context_ != 0);
  
+ #if defined ACE_WIN32
+   // Let's handle this special case before we do any real work.
+   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::EXCEPT_MASK))
+     ACE_NOTSUP_RETURN(-1);
+ #endif /* ACE_WIN32 */
+ 
    int result = ACE_Select_Reactor::register_handler_i (handle,
                                                         handler, mask);
    if (result == -1)
      return -1;
  
!   synchronize_XtInput (handle);
    return 0;
  }
  
***************
*** 290,350 ****
  {
    ACE_TRACE ("ACE_XtReactor::remove_handler_i");
  
!   // In the registration phase we registered first with
!   // ACE_Select_Reactor and then with X.  Now we are now doing things
!   // in reverse order.
  
!   // First clean up the corresponding X11Input.
!   this->remove_XtInput (handle);
  
!   // Now let the reactor do its work.
!   return ACE_Select_Reactor::remove_handler_i (handle,
                                                 mask);
  }
  
  void
! ACE_XtReactor::remove_XtInput (ACE_HANDLE handle)
  {
!   ACE_TRACE ("ACE_XtReactor::remove_XtInput");
  
!   ACE_XtReactorID *XtID = this->ids_;
  
!   if (XtID)
      {
!       if (XtID->handle_ == handle)
!         {
!           ::XtRemoveInput (XtID->id_);
!           this->ids_ = XtID->next_;
!           delete XtID;
!           return;
!         }
! 
!       ACE_XtReactorID *NextID = XtID->next_;
! 
!       while (NextID)
!         {
!           if (NextID->handle_ == handle)
!             {
!               ::XtRemoveInput(NextID->id_);
!               XtID->next_ = NextID->next_;
!               delete NextID;
!               return;
!             }
!           else
!             {
!               XtID = NextID;
!               NextID = NextID->next_;
!             }
!         }
      }
  }
  
  int
! ACE_XtReactor::remove_handler_i (const ACE_Handle_Set &handles,
!                                  ACE_Reactor_Mask mask)
  {
!   return ACE_Select_Reactor::remove_handler_i (handles,
!                                                mask);
  }
  
  // The following functions ensure that there is an Xt timeout for the
--- 234,370 ----
  {
    ACE_TRACE ("ACE_XtReactor::remove_handler_i");
  
!   int result  = ACE_Select_Reactor::remove_handler_i (handle,
! 						      mask);
!   if (result == -1)
!     return -1;
  
!   synchronize_XtInput (handle);
!   return 0;
! }
  
! int
! ACE_XtReactor::remove_handler_i (const ACE_Handle_Set &handles,
!                                  ACE_Reactor_Mask mask)
! {
!   return ACE_Select_Reactor::remove_handler_i (handles,
                                                 mask);
  }
  
+ int
+ ACE_XtReactor::suspend_i (ACE_HANDLE handle)
+ {
+   ACE_TRACE ("ACE_XtReactor::suspend_i");
+ 
+   int result  = ACE_Select_Reactor::suspend_i (handle);
+ 
+   if (result == -1)
+     return -1;
+ 
+   synchronize_XtInput (handle);
+   return 0;
+ }
+ 
+ int
+ ACE_XtReactor::resume_i (ACE_HANDLE handle)
+ {
+   ACE_TRACE ("ACE_XtReactor::resume_i");
+ 
+   int result  =  ACE_Select_Reactor::resume_i (handle);
+ 
+   if (result == -1)
+     return -1;
+ 
+   synchronize_XtInput (handle);
+   return 0;
+ }
+ 
  void
! ACE_XtReactor::synchronize_XtInput(ACE_HANDLE handle)
  {
!   ACE_TRACE ("ACE_XtReactor::synchronize_XtInput");
  
!   // The idea here is to call this function after the base class has
!   // processed the register/remove/suspend/resume_handler request. The
!   // resulting mask is used to find out which XtInput mask we need.
  
!   // Find existing handler in linked list.
!   ACE_XtReactorID **XtID = &(this->ids_);
! 
!   while (*XtID && (*XtID)->handle_ != handle)
!     XtID = &((*XtID)->next_);
! 
!   // Remove existing input handler.
!   if (*XtID)
!     ::XtRemoveInput ((*XtID)->id_);
! 
!   int condition = compute_Xt_condition (handle);
! 
!   if (condition == 0) // No input handler needed.
      {
!       if (*XtID)
! 	{
! 	  // Remove linked list entry.
! 	  ACE_XtReactorID *toDelete  = *XtID;
! 	  *XtID = (*XtID)->next_;
! 	  delete toDelete;
! 	}
!       return;
      }
+ 
+   if (*XtID == 0)
+     {
+       // Create new node.
+       ACE_XtReactorID *tmp = new ACE_XtReactorID;
+       tmp->next_ = this->ids_;
+       tmp->handle_ = handle;
+       this->ids_ = tmp;
+       XtID = &(this->ids_);
+     }
+ 
+   // Finally, add input handler.
+   (*XtID)->id_ = ::XtAppAddInput (this->context_,
+                                   (int) handle,
+                                   (XtPointer) condition,
+                                   InputCallbackProc,
+                                   (XtPointer) this);
  }
  
  int
! ACE_XtReactor::compute_Xt_condition(ACE_HANDLE handle)
  {
!   ACE_TRACE ("ACE_XtReactor::compute_Xt_condition");
! 
!   // Retrieve current wait mask from base class.
!   // The returned value is either a combination of READ/WRITE/EXCEPT_MASK
!   // or -1.
!   int mask =this->bit_ops(handle,
! 			  0,
! 			  this->wait_set_,
! 			  ACE_Reactor::GET_MASK);
! 
!   if (mask == -1) // No active mask.
!     return 0;
! 
!  int condition = 0;
! 
! #if !defined ACE_WIN32
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::READ_MASK))
!     ACE_SET_BITS (condition, XtInputReadMask);
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::WRITE_MASK))
!     ACE_SET_BITS (condition, XtInputWriteMask);
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::EXCEPT_MASK))
!     ACE_SET_BITS (condition, XtInputExceptMask);
! #else
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::READ_MASK))
!     ACE_SET_BITS (condition, XtInputReadWinsock);
!   if (ACE_BIT_ENABLED (mask, ACE_Event_Handler::WRITE_MASK))
!     ACE_SET_BITS (condition, XtInputWriteWinsock);
!   // EXCEPT_MASK is not supported for WIN32. As this was
!   // already handled in register_handler_i, no check here. 
! #endif /* !ACE_WIN32 */
! 
!   return condition;
  }
  
  // The following functions ensure that there is an Xt timeout for the
diff -crB XtReactor.orig/XtReactor.h XtReactor/XtReactor.h
*** XtReactor.orig/XtReactor.h	Tue Sep 16 11:35:44 2008
--- XtReactor/XtReactor.h	Sat Dec 12 15:28:28 2009
***************
*** 101,108 ****
    virtual int remove_handler_i (const ACE_Handle_Set &handles,
                                  ACE_Reactor_Mask);
  
!   /// Removes an Xt handle.
!   virtual void remove_XtInput (ACE_HANDLE handle);
  
    /// Wait for events to occur.
    virtual int wait_for_multiple_events (ACE_Select_Reactor_Handle_Set &,
--- 101,117 ----
    virtual int remove_handler_i (const ACE_Handle_Set &handles,
                                  ACE_Reactor_Mask);
  
!   /// Suspend the <Event_Handler> associated with @a handle
!   virtual int suspend_i (ACE_HANDLE handle);
! 
!   /// Resume the <Event_Handler> associated with @a handle
!   virtual int resume_i (ACE_HANDLE handle);
! 
!   /// Synchronize XtInputHandler for @a handle
!   virtual void synchronize_XtInput(ACE_HANDLE handle);
! 
!   /// Compute needed Xt condition by looking at base class wait set. 
!   virtual int compute_Xt_condition(ACE_HANDLE handle);
  
    /// Wait for events to occur.
    virtual int wait_for_multiple_events (ACE_Select_Reactor_Handle_Set &,
