Index: Connection_Handler.h
===================================================================
--- Connection_Handler.h	(revision 91422)
+++ Connection_Handler.h	(working copy)
@@ -58,11 +58,8 @@
 public:
 
   /// Constructor
-  TAO_Connection_Handler (void);
+  explicit TAO_Connection_Handler (TAO_ORB_Core *orb_core);
 
-  /// Constructor
-  TAO_Connection_Handler (TAO_ORB_Core *orb_core);
-
   /// Destructor
   virtual ~TAO_Connection_Handler (void);
 
Index: GIOP_Message_Base.cpp
===================================================================
--- GIOP_Message_Base.cpp	(revision 91422)
+++ GIOP_Message_Base.cpp	(working copy)
@@ -688,8 +688,11 @@
     }
 #endif
   if (TAO_debug_level > 9)
-    { //due to alignment data block has an offset which needs to be corrected
-      this->dump_msg ("recv",
+    {
+      char buf[48];
+      ACE_OS::sprintf (buf, "Transport[%d] recv", transport->id ());
+      //due to alignment data block has an offset which needs to be corrected
+      this->dump_msg (buf,
                       reinterpret_cast <u_char *> (db->base () + rd_pos - TAO_GIOP_MESSAGE_HEADER_LEN),
                       db->size ()  + rd_pos - TAO_GIOP_MESSAGE_HEADER_LEN);
     }
@@ -800,7 +803,9 @@
 #endif
   if (TAO_debug_level > 9)
     {
-      this->dump_msg ("recv",
+      char buf[48];
+      ACE_OS::sprintf (buf, "Transport[%d] recv", params.transport_->id ());
+      this->dump_msg (buf,
                       reinterpret_cast <u_char *> (db->base () + rd_pos - TAO_GIOP_MESSAGE_HEADER_LEN),
                       db->size ()  + rd_pos - TAO_GIOP_MESSAGE_HEADER_LEN);
     }
Index: Transport_Connector.cpp
===================================================================
--- Transport_Connector.cpp	(revision 91422)
+++ Transport_Connector.cpp	(working copy)
@@ -734,7 +734,14 @@
       result = tcm.cache_transport (&desc, transport);
       if (result == -1)
       {
-
+        if (TAO_debug_level > 2)
+        {
+          ACE_DEBUG ((LM_DEBUG,
+                      ACE_TEXT("TAO (%P|%t) - Transport_Connector::")
+                      ACE_TEXT("wait_for_connection_completion, ")
+                      ACE_TEXT("transport [%d], Failed to cache transport.\n"),
+                      transport->id ()));
+        }
       }
     }
   else if (transport->connection_handler ()->is_timeout ())
Index: IIOP_Acceptor.cpp
===================================================================
--- IIOP_Acceptor.cpp	(revision 91422)
+++ IIOP_Acceptor.cpp	(working copy)
@@ -282,7 +282,7 @@
       ACE_DEBUG ((LM_DEBUG,
                   ACE_TEXT ("TAO (%P|%t) - ")
                   ACE_TEXT ("IIOP_Acceptor::open, address==%C, options=%C\n"),
-                  address, options));
+                  address, options != 0 ? options : "(null)"));
     }
 
   this->orb_core_ = orb_core;
Index: IIOP_Connection_Handler.cpp
===================================================================
--- IIOP_Connection_Handler.cpp	(revision 91422)
+++ IIOP_Connection_Handler.cpp	(working copy)
@@ -30,10 +30,12 @@
   Reference_Count rc = TAO_IIOP_SVC_HANDLER::add_reference ();
   if (TAO_debug_level > 9)
     {
+      TAO_Transport *tport = this->transport ();
       ACE_DEBUG ((LM_DEBUG,
                   ACE_TEXT("TAO (%P|%t) - IIOP_Connection_Handler[%d]::")
                   ACE_TEXT("add_reference, up to %d\n"),
-                  this->transport (), rc));
+                  tport != 0 ? tport->id () : 0,
+                  rc));
     }
   return rc;
 
@@ -48,7 +50,9 @@
     {
       ACE_DEBUG ((LM_DEBUG,
                   ACE_TEXT("TAO (%P|%t) - IIOP_Connection_Handler[%d]::")
-                  ACE_TEXT("remove_reference, down to %d\n"), tport, rc));
+                  ACE_TEXT("remove_reference, down to %d\n"),
+                  tport != 0 ? tport->id () : 0,
+                  rc));
     }
   return rc;
 }
@@ -81,10 +85,12 @@
 
   if (TAO_debug_level > 9)
     {
+      TAO_Transport *tport = static_cast<TAO_Transport *> (specific_transport);
       ACE_DEBUG ((LM_DEBUG,
                   ACE_TEXT("TAO (%P|%t) - IIOP_Connection_Handler[%d] ctor, ")
                   ACE_TEXT("this=%@\n"),
-                  static_cast<TAO_Transport *> (specific_transport), this));
+                  tport != 0 ? tport->id () : 0,
+                  this));
     }
 
   // store this pointer (indirectly increment ref count)
@@ -93,6 +99,16 @@
 
 TAO_IIOP_Connection_Handler::~TAO_IIOP_Connection_Handler (void)
 {
+  if (TAO_debug_level > 9)
+    {
+      TAO_Transport *tport = this->transport ();
+      ACE_DEBUG ((LM_DEBUG,
+                  ACE_TEXT("TAO (%P|%t) - IIOP_Connection_Handler[%d] dtor, ")
+                  ACE_TEXT("this=%@,transport=%@\n"),
+                  tport != 0 ? tport->id () : 0,
+                  this,
+                  tport));
+    }
   delete this->transport ();
   int const result =
     this->release_os_resources ();
Index: Connection_Handler.cpp
===================================================================
--- Connection_Handler.cpp	(revision 91422)
+++ Connection_Handler.cpp	(working copy)
@@ -384,7 +384,7 @@
     {
       ACE_DEBUG  ((LM_DEBUG,
                    "TAO (%P|%t) - Connection_Handler[%d]::"
-                   "close_connection_eh\n",
+                   "close_connection_eh end\n",
                    id));
     }
 
Index: Transport.cpp
===================================================================
--- Transport.cpp	(revision 91422)
+++ Transport.cpp	(working copy)
@@ -201,6 +201,13 @@
 
 TAO_Transport::~TAO_Transport (void)
 {
+  if (TAO_debug_level > 9)
+    {
+      ACE_DEBUG ((LM_DEBUG, ACE_TEXT ("TAO (%P|%t) - Transport[%d]::~Transport\n"),
+                  this->id_
+                  ));
+    }
+  
   delete this->messaging_object_;
 
   delete this->ws_;
@@ -488,6 +495,14 @@
     entry = this->cache_map_entry_;
     this->cache_map_entry_ = 0;
   }
+  
+  if (TAO_debug_level > 3)
+    {
+      ACE_DEBUG ((LM_DEBUG,
+                  ACE_TEXT ("TAO (%P|%t) - Transport[%d]::purge_entry\n")
+                  ACE_TEXT ("entry is %@\n"),
+                  this->id (), entry));
+    }
   return this->transport_cache_manager ().purge_entry (entry);
 }
 
@@ -958,7 +973,7 @@
     retval = this->send (iov, iovcnt, byte_count,
                          this->io_timeout (dc));
 
-  if (TAO_debug_level == 5)
+  if (TAO_debug_level > 9)
     {
       dump_iov (iov, iovcnt, this->id (),
                 byte_count, ACE_TEXT("drain_queue_helper"));
@@ -2675,6 +2690,11 @@
 void
 TAO_Transport::pre_close (void)
 {
+  if (TAO_debug_level > 9)
+    {
+      ACE_DEBUG ((LM_DEBUG, ACE_TEXT ("TAO (%P|%t) - Transport[%d]::pre_close\n"),
+                  this->id_));
+    }
   // @TODO: something needs to be done with is_connected_. Checking it is
   // guarded by a mutex, but setting it is not. Until the need for mutexed
   // protection is required, the transport cache is holding its own copy
@@ -2694,8 +2714,10 @@
 TAO_Transport::post_open (size_t id)
 {
   if (TAO_debug_level > 9)
-    ACE_DEBUG ((LM_DEBUG, ACE_TEXT ("TAO (%P|%t) - Transport::post_open, ")
-                ACE_TEXT ("tport id changed from %d to %d\n"), this->id_, id));
+    {
+      ACE_DEBUG ((LM_DEBUG, ACE_TEXT ("TAO (%P|%t) - Transport::post_open, ")
+                  ACE_TEXT ("tport id changed from %d to %d\n"), this->id_, id));
+    }
   this->id_ = id;
 
   // When we have data in our outgoing queue schedule ourselves
