Index: RT_ORB_Loader.cpp
===================================================================
--- RT_ORB_Loader.cpp	(revision 87620)
+++ RT_ORB_Loader.cpp	(working copy)
@@ -31,6 +31,19 @@
 {
   ACE_TRACE ("TAO_RT_ORB_Loader::init");
 
+  ACE_Service_Gestalt *gestalt = ACE_Service_Config::current ();
+
+  ACE_Service_Object * const rt_loader =
+    ACE_Dynamic_Service<ACE_Service_Object>::instance (
+      gestalt,
+      "RT_ORB_Loader");
+
+  if (rt_loader != 0 && rt_loader != this)
+    {
+      this->initialized_ = true;
+      return rt_loader->init (argc, argv);
+    }
+
   // Only allow initialization once.
   if (this->initialized_)
     return 0;
