diff -rubw ACE-bleeding/ace/Time_Value.h ACE-current/ace/Time_Value.h
--- ACE-bleeding/ace/Time_Value.h	2005-04-19 12:20:58.000000000 -0400
+++ ACE-current/ace/Time_Value.h	2005-10-14 10:11:20.000000000 -0400
@@ -197,6 +197,14 @@
   void sec (long sec);
 
   /// Get microseconds.
+#if !defined (ACE_LACKS_LONGLONG_T)
+  /**
+   * @return Sum of second field (in microseconds) and microsecond field
+   *         and return them via the @param usec parameter.
+   */
+  void to_usec (ACE_UINT64 &usec) const;
+
+#endif /*ACE_LACKS_LONGLONG_T */
   /**
    * @return The microsecond field/component of this ACE_Time_Value.
    *
diff -rubw ACE-bleeding/ace/Time_Value.inl ACE-current/ace/Time_Value.inl
--- ACE-bleeding/ace/Time_Value.inl	2005-08-15 16:06:58.000000000 -0400
+++ ACE-current/ace/Time_Value.inl	2005-10-14 10:11:24.000000000 -0400
@@ -153,6 +153,17 @@
 
 // Returns number of micro-seconds.
 
+#if !defined (ACE_LACKS_LONGLONG_T)
+ACE_INLINE void
+ACE_Time_Value::to_usec (ACE_UINT64 &usec) const
+{
+  // ACE_OS_TRACE ("ACE_Time_Value::usec");
+  usec = static_cast<ACE_UINT64> (this->tv_.tv_sec);
+  usec *= 1000000;
+  usec += this->tv_.tv_usec;
+}
+#endif /*ACE_LACKS_LONGLONG_T*/
+
 ACE_INLINE long
 ACE_Time_Value::usec (void) const
 {
diff -rubw ACE-bleeding/ace/UUID.cpp ACE-current/ace/UUID.cpp
--- ACE-bleeding/ace/UUID.cpp	2005-08-19 12:23:11.000000000 -0400
+++ ACE-current/ace/UUID.cpp	2005-10-14 10:11:05.000000000 -0400
@@ -448,12 +448,14 @@
   void
   UUID_Generator::get_systemtime (UUID_time & timestamp)
   {
-    const UUID_time timeOffset = 0;
-    //const UUID_time timeOffset = 0x1B21DD213814000;
+    //const UUID_time timeOffset = 0;
+    const UUID_time timeOffset = ACE_UINT64_LITERAL (0x1B21DD213814000);
 
     /// Get the time of day, convert to 100ns ticks then add the offset.
     ACE_Time_Value now = ACE_OS::gettimeofday();
-    UUID_time time = now.sec() * 10000000 + now.usec() * 10;
+    ACE_UINT64 time;
+    now.to_usec (time);
+    time = time * 10;
     timestamp = time + timeOffset;
 }
 
