Index: ace/Parse_Node.cpp
===================================================================
--- ace/Parse_Node.cpp	(revision 78243)
+++ ace/Parse_Node.cpp	(working copy)
@@ -30,15 +30,53 @@
 #endif /* ACE_HAS_DUMP */
 }
 
+#include "ace/ARGV.h"
+
 void
 ACE_Stream_Node::apply (ACE_Service_Gestalt *config, int &yyerrno)
 {
   ACE_TRACE ("ACE_Stream_Node::apply");
 
-  if (config->initialize (this->node_->record (config),
-                          this->node_->parameters ()) == -1)
-    ++yyerrno;
+  const ACE_Service_Type *sst = this->node_->record (config);
+  if (sst == 0)
+      const_cast<ACE_Static_Node *> (this->node_)->apply (config, yyerrno);
+  if (yyerrno != 0) return;
 
+  sst = this->node_->record (config);
+  ACE_Stream_Type *st = 
+      dynamic_cast<ACE_Stream_Type *> (const_cast<ACE_Service_Type_Impl *> (sst->type ()));
+
+  for (const ACE_Static_Node *module = dynamic_cast<const ACE_Static_Node*> (this->mods_);
+       module != 0;
+       module = dynamic_cast<ACE_Static_Node*> (module->link()))
+  {
+    ACE_ARGV args (module->parameters ());
+
+    const ACE_Service_Type *mst = module->record (config);
+    if (mst == 0)
+      const_cast<ACE_Static_Node *> (module)->apply (config, yyerrno);
+
+    if (yyerrno != 0)
+    {
+      ACE_ERROR ((LM_ERROR,
+                  ACE_LIB_TEXT ("dynamic initialization failed for Module %s\n"),
+                  module->name ()));
+      ++yyerrno;
+    }
+
+    ACE_Module_Type const * const mt1 = static_cast <ACE_Module_Type const *> (module->record (config)->type());
+    ACE_Module_Type *mt = const_cast<ACE_Module_Type *>(mt1);
+
+    if (st->push (mt) == -1)
+    {
+      ACE_ERROR ((LM_ERROR,
+                  ACE_LIB_TEXT ("dynamic initialization failed for Stream %s\n"),
+                  this->node_->name ()));
+      ++yyerrno;
+    }
+
+  } 
+
 #ifndef ACE_NLOGGING
   if (ACE::debug ())
     ACE_DEBUG ((LM_DEBUG,
@@ -76,7 +114,14 @@
 ACE_Parse_Node::link (ACE_Parse_Node *n)
 {
   ACE_TRACE ("ACE_Parse_Node::link");
-  this->next_ = n;
+
+  // Find the last list entry (if any) ...
+  ACE_Parse_Node *t = this;
+  while (t->next_ != 0)
+      t = t->next_;
+
+  // ... and insert n there.
+  t->next_ = n;
 }
 
 ACE_Stream_Node::ACE_Stream_Node (const ACE_Static_Node *str_ops,
@@ -718,7 +763,7 @@
   ACE_Static_Svc_Descriptor *ssd = 0;
   if (config->find_static_svc_descriptor (this->function_name_, &ssd) == -1)
     {
-      yyerrno++;
+      ++yyerrno;
       ACE_ERROR_RETURN ((LM_ERROR,
        ACE_LIB_TEXT ("(%P|%t) No static service ")
        ACE_LIB_TEXT ("registered for function %s\n"),
@@ -728,7 +773,7 @@
 
   if (ssd->alloc_ == 0)
     {
-      yyerrno++;
+      ++yyerrno;
 
       if (this->symbol_ == 0)
         {
Index: ace/Service_Config.h
===================================================================
--- ace/Service_Config.h	(revision 78243)
+++ ace/Service_Config.h	(working copy)
@@ -368,65 +368,82 @@
   /// Perform the reconfiguration process.
   static void reconfigure (void);
 
-  // = The following methods are static in order to enforce Singleton
+  // = Accessors and mutators for process-wide Singletons.
+
+  // The following methods are deprecated since the service configurator 
+  // has been refactored to allow multiple configuration repositories 
+  // (gestalts). The interfaces are retained for the sake of maintaining 
+  // source-code compatibility and assume to only operate on the one *global* 
+  // configuration instance (gestalt). All the static methods have their 
+  // instance methods counterparts. In new code use the corresponding instance 
+  // methods. For example instead of:
+  //
+  //    ACE_Service_Config::insert (&svcdesc);
+  //
+  // use:
+  //
+  //    ACE_Service_Config::instance ()->insert (&svcdesc);
+  //
+  // The methods were static in order to enforce Singleton/Monostate
   // semantics for the Reactor, Service_Repository, Thread_Manager,
   // and Acceptor/Connector Strategy factory.  Other portions of the
-  // system may need to access them at some point or another...
+  // system may have needed to access them at some point or another...
 
-  // = This is not strictly needed, anymore since the service configurator
-  // has been refactored to allow multiple service configuration
-  // instances (called gestalts). The interfaces, however were retained in for
-  // the sake of maintaining source-code compatibility.
 
-
-  // = Accessors and mutators for process-wide Singletons.
-
-  /// Returns a pointer to the list of statically linked services.
-  ///
-  /// @deprecated - Same as instance(), but still useful in legacy code,
-  /// (notably, one that can not be easily modified) which uses the following
-  /// idiom for registering static services:
-  ///
-  ///    ACE_Service_Config::static_svcs ()->insert (...);
+  // Returns a pointer to the list of statically linked services.
+  //
+  // @deprecated - Same as instance(), but still useful in legacy code,
+  // (notably, one that can not be easily modified) which uses the following
+  // idiom for registering static services:
+  //
+  //    ACE_Service_Config::static_svcs ()->insert (...);
+  //
+  // Use the code in the example above instead.
   static ACE_Service_Gestalt *static_svcs (void);
 
-  /// Insert a static service descriptor for processing on open_i(). The
-  /// corresponding ACE_STATIC_SVC_* macros were chaged to use this method
-  /// instead of obtaining a ptr to a container. See the note on static_svcs().
-  /// Added to prevent exposing the internal storage representation of the
-  /// services repository and provide a better way of debugging service
-  /// loading and registration problems.
+  // @deprecated
+  // Insert a static service descriptor for subsequent processing 
+  // on open_i(). This is a shorthand for using the 
+  // ACE_Service_Gestalt::insert().
   static int insert (ACE_Static_Svc_Descriptor *svc);
 
-  // = Utility methods.
-  /// Dynamically link the shared object file and retrieve a pointer to
-  /// the designated shared object in this file.
+  // @deprecated
+  // Dynamically link the shared object file, retrieve a pointer to
+  // the designated shared object in this file and call the init() method.
+  // This is a shorthand for ACE_Service_Gestalt::initialize(), which
+  // should not be used either for it perpetuates the same issues - see the 
+  // comment in that method for detais.
   static int initialize (const ACE_Service_Type *,
                          const ACE_TCHAR *parameters);
 
-  /// Initialize and activate a statically @a svc_name service.
+  
+  // @deprecated
+  // Initialize and activate a statically @a svc_name service.
   static int initialize (const ACE_TCHAR *svc_name,
                          const ACE_TCHAR *parameters);
 
-  /// Resume a @a svc_name that was previously suspended or has not yet
-  /// been resumed (e.g., a static service).
+  
+  // @deprecated
+  // Resume a @a svc_name that was previously suspended or has not yet
+  // been resumed (e.g., a static service).
   static int resume (const ACE_TCHAR svc_name[]);
 
-  /**
-   * Suspend @a svc_name.  Note that this will not unlink the service
-   * from the daemon if it was dynamically linked, it will mark it as
-   * being suspended in the Service Repository and call the <suspend>
-   * member function on the appropriate ACE_Service_Object.  A
-   * service can be resumed later on by calling the <RESUME> member
-   * function...
-   */
+  // @deprecated
+  // Suspend @a svc_name.  Note that this will not unlink the service
+  // from the daemon if it was dynamically linked, it will mark it as
+  // being suspended in the Service Repository and call the <suspend>
+  // member function on the appropriate ACE_Service_Object.  A
+  // service can be resumed later on by calling the <RESUME> member
+  // function...
   static int suspend (const ACE_TCHAR svc_name[]);
 
-  /// Totally remove @a svc_name from the daemon by removing it
-  /// from the ACE_Reactor, and unlinking it if necessary.
+  // @deprecated
+  // Totally remove @a svc_name from the daemon by removing it
+  // from the ACE_Reactor, and unlinking it if necessary.
   static int remove (const ACE_TCHAR svc_name[]);
 
 #if defined (ACE_HAS_WINCE) && defined (ACE_USES_WCHAR)
+  // @deprecated
   // We must provide these function to bridge the Svc_Conf parser
   // with ACE.
   static int initialize (const ACE_Service_Type *, ACE_ANTI_TCHAR []);
@@ -445,34 +462,34 @@
   /// Declare the dynamic allocation hooks.
   ACE_ALLOC_HOOK_DECLARE;
 
+  // @deprecated
   /// Process a file containing a list of service configuration
   /// directives.
   static int process_file (const ACE_TCHAR file[]);
 
+  // @deprecated
   /// Process one service configuration @a directive, which is passed as
   /// a string.  Returns the number of errors that occurred.
   static int process_directive (const ACE_TCHAR directive[]);
 
-  /// Process one static service definition.
-  /**
-   * Load a new static service into the ACE_Service_Repository.
-   *
-   * @param ssd Service descriptor, see the document of
-   *        ACE_Static_Svc_Descriptor for more details.
-   *
-   * @param force_replace If set the new service descriptor replaces
-   *        any previous instance in the ACE_Service_Repository.
-   *
-   * @return Returns -1 if the service cannot be 'loaded'.
-   */
+  // @deprecated
+  // Process one static service definition.
+  // Load a new static service into the ACE_Service_Repository.
+  //
+  // @param ssd Service descriptor, see the document of
+  //        ACE_Static_Svc_Descriptor for more details.
+  //
+  // @param force_replace If set the new service descriptor replaces
+  //        any previous instance in the ACE_Service_Repository.
+  //
+  // @return Returns -1 if the service cannot be 'loaded'.
   static int process_directive (const ACE_Static_Svc_Descriptor &ssd,
                                 int force_replace = 0);
 
-  /**
-   * Process (or re-process) service configuration requests that are
-   * provided in the svc.conf file(s).  Returns the number of errors
-   * that occurred.
-   */
+  // @deprecated
+  // Process (or re-process) service configuration requests that are
+  // provided in the svc.conf file(s).  Returns the number of errors
+  // that occurred.
   static int process_directives (void);
 
   /// Handles signals to trigger reconfigurations.
@@ -496,6 +513,7 @@
    *        for a list of files and here a list of services.
    */
   static int parse_args (int, ACE_TCHAR *argv[]);
+
 #if (ACE_USES_CLASSIC_SVC_CONF == 0)
   static ACE_Service_Type *create_service_type  (const ACE_TCHAR *n,
                                                  ACE_Service_Type_Impl *o,
@@ -512,22 +530,25 @@
 
 protected:
 
-  /// Process service configuration requests that were provided on the
-  /// command-line.  Returns the number of errors that occurred.
+  // @deprecated
+  // Process service configuration requests that were provided on the
+  // command-line.  Returns the number of errors that occurred.
   static int process_commandline_directives (void);
 
 #if (ACE_USES_CLASSIC_SVC_CONF == 1)
-  /// This is the implementation function that process_directives()
-  /// and process_directive() both call.  Returns the number of errors
-  /// that occurred.
+  // @deprecated
+  // This is the implementation function that process_directives()
+  // and process_directive() both call.  Returns the number of errors
+  // that occurred.
   static int process_directives_i (ACE_Svc_Conf_Param *param);
 #endif /* ACE_USES_CLASSIC_SVC_CONF == 1 */
 
   /// Become a daemon.
   static int start_daemon (void);
 
-  /// Add the default statically-linked services to the
-  /// ACE_Service_Repository.
+  // @deprecated
+  // Add the default statically-linked services to the
+  // ACE_Service_Repository.
   static int load_static_svcs (void);
 
 
Index: ace/Svc_Conf.y
===================================================================
--- ace/Svc_Conf.y	(revision 78243)
+++ ace/Svc_Conf.y	(working copy)
@@ -170,30 +170,6 @@
 module
   : dynamic
     {
-      ACE_Static_Node *svc_type = $<static_node_>1;
-
-      if (svc_type != 0)
-        {
-          ACE_Static_Node *module = $<static_node_>-1;
-
-          ACE_ARGV args (svc_type->parameters ());
-          ACE_Module_Type *mt = ace_get_module (module->record (ACE_SVC_CONF_PARAM->config),
-                                                svc_type->record (ACE_SVC_CONF_PARAM->config),
-                                                ACE_SVC_CONF_PARAM->yyerrno);
-          ACE_Stream_Type *st =
-            dynamic_cast<ACE_Stream_Type *> (const_cast<ACE_Service_Type_Impl *> (module->record (ACE_SVC_CONF_PARAM->config)->type ()));
-
-          if (!st
-              || !mt
-              || mt->init (args.argc (), args.argv ()) == -1
-              || st->push (mt) == -1)
-            {
-              ACE_ERROR ((LM_ERROR,
-                          ACE_LIB_TEXT ("dynamic initialization failed for Module %s\n"),
-                          svc_type->name ()));
-              ACE_SVC_CONF_PARAM->yyerrno++;
-            }
-        }
     }
   | static
     {
Index: ace/Svc_Conf_y.cpp
===================================================================
--- ace/Svc_Conf_y.cpp	(revision 78243)
+++ ace/Svc_Conf_y.cpp	(working copy)
@@ -1242,30 +1242,6 @@
   case 26:
 
     {
-      ACE_Static_Node *svc_type = (ace_yyvsp[0].static_node_);
-
-      if (svc_type != 0)
-        {
-          ACE_Static_Node *module = (ace_yyvsp[-2].static_node_);
-
-          ACE_ARGV args (svc_type->parameters ());
-          ACE_Module_Type *mt = ace_get_module (module->record (ACE_SVC_CONF_PARAM->config),
-                                                svc_type->record (ACE_SVC_CONF_PARAM->config),
-                                                ACE_SVC_CONF_PARAM->yyerrno);
-          ACE_Stream_Type *st =
-            dynamic_cast<ACE_Stream_Type *> (const_cast<ACE_Service_Type_Impl *> (module->record (ACE_SVC_CONF_PARAM->config)->type ()));
-
-          if (!st
-              || !mt
-              || mt->init (args.argc (), args.argv ()) == -1
-              || st->push (mt) == -1)
-            {
-              ACE_ERROR ((LM_ERROR,
-                          ACE_LIB_TEXT ("dynamic initialization failed for Module %s\n"),
-                          svc_type->name ()));
-              ACE_SVC_CONF_PARAM->yyerrno++;
-            }
-        }
     ;}
     break;
 
@@ -1696,7 +1672,10 @@
      : static_cast<ACE_Stream_Type const *> (sr->type ()));
   
   ACE_Module_Type const * const mt =
-    static_cast <ACE_Module_Type const *> (sv->type ());
+      (sv == 0
+     ? 0
+     :
+    static_cast <ACE_Module_Type const *> (sv->type ()));
 
   ACE_TCHAR const * const module_type_name =
     (mt ? mt->name () : ACE_LIB_TEXT ("(nil)"));
Index: ace/Svc_Conf_Tokens.h
===================================================================
--- ace/Svc_Conf_Tokens.h	(revision 78243)
+++ ace/Svc_Conf_Tokens.h	(working copy)
@@ -1,11 +1,11 @@
-// $Id$
+// $Id:$
 #ifndef BISON_SVC_CONF_TAB_H
 # define BISON_SVC_CONF_TAB_H
 # define ACE_YYSTYPE_IS_DECLARED 1
-/* A Bison parser, made by GNU Bison 2.0.  */
+/* A Bison parser, made by GNU Bison 2.1.  */
 
 /* Skeleton parser for Yacc-like parsing with Bison,
-   Copyright (C) 1984, 1989, 1990, 2000, 2001, 2002, 2003, 2004 Free Software Foundation, Inc.
+   Copyright (C) 1984, 1989, 1990, 2000, 2001, 2002, 2003, 2004, 2005 Free Software Foundation, Inc.
 
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
@@ -19,8 +19,8 @@
 
    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
-   Foundation, Inc., 59 Temple Place - Suite 330,
-   Boston, MA 02111-1307, USA.  */
+   Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.  */
 
 /* As a special exception, when this file is copied by Bison into a
    Bison output file, you may use that output file without restriction.
@@ -49,6 +49,7 @@
      ACE_STRING = 271
    };
 #endif
+/* Tokens.  */
 #define ACE_DYNAMIC 258
 #define ACE_STATIC 259
 #define ACE_SUSPEND 260
Index: ace/svcconf.mpb
===================================================================
--- ace/svcconf.mpb	(revision 78243)
+++ ace/svcconf.mpb	(working copy)
@@ -43,6 +43,7 @@
     "       -e 's/ NULL/ 0/g' \\"
     "       -e 's/ace_yyerror[ ]*(\"/ace_yyerror (ACE_SVC_CONF_PARAM->yyerrno, ACE_SVC_CONF_PARAM->yylineno, \"/g' \\"
     "       -e 's/ace_yyerror[ ]*(ace_yymsg/ace_yyerror (ACE_SVC_CONF_PARAM->yyerrno, ACE_SVC_CONF_PARAM->yylineno, ace_yymsg/g' \\"
+    "       -e 's/ace_yyerror[ ]*(ACE_YY_/ace_yyerror (ACE_SVC_CONF_PARAM->yyerrno, ACE_SVC_CONF_PARAM->yylineno, ACE_YY_/g' \\"
     "       -e 's/ace_yyerror (int ace_yyerrno, int ace_yylineno, const ACE_TCHAR/ace_yyerror (int ace_yyerrno, int ace_yylineno, const char/' \\"
     "       -e 's@#include <stdio\.h>@@' \\"
     "       -e 's/Svc_Conf\\.tab\\.c/Svc_Conf_y.cpp/g' < Svc_Conf.tab.c > /tmp/$@"
