Index: ace/Log_Msg.h
===================================================================
--- ace/Log_Msg.h	(revision 83617)
+++ ace/Log_Msg.h	(working copy)
@@ -24,6 +24,8 @@
 #include "ace/Default_Constants.h"
 #include "ace/Log_Priority.h"
 #include "ace/os_include/os_limits.h"
+#include "ace/Atomic_Op.h"
+#include "ace/Synch_Traits.h"
 
 // The ACE_ASSERT macro used to be defined here, include ace/Assert.h
 // for backwards compatibility.
@@ -583,6 +585,8 @@
   ACE_ALLOC_HOOK_DECLARE;
 
 private:
+  void cleanup_ostream ();
+
   /// Status of operation (-1 means failure, >= 0 means success).
   int status_;
 
@@ -607,6 +611,9 @@
   /// The ostream where logging messages can be written.
   ACE_OSTREAM_TYPE *ostream_;
 
+  typedef ACE_Atomic_Op<ACE_SYNCH_MUTEX, unsigned long> Atomic_ULong;
+  Atomic_ULong *ostream_refcount_;
+
   /// The callback object.
   ACE_Log_Msg_Callback *msg_callback_;
 
@@ -619,9 +626,6 @@
   /// Are we allowing tracing in this thread?
   bool tracing_enabled_;
 
-  /// Are we deleting this ostream?
-  bool delete_ostream_;
-
   /**
    * If we're running in the context of an ACE_Thread_Manager this
    * will point to the thread descriptor adapter which holds the
Index: ace/Log_Msg.cpp
===================================================================
--- ace/Log_Msg.cpp	(revision 83617)
+++ ace/Log_Msg.cpp	(working copy)
@@ -650,11 +650,11 @@
     msg_ (0),
     restart_ (1),  // Restart by default...
     ostream_ (0),
+    ostream_refcount_ (0),
     msg_callback_ (0),
     trace_depth_ (0),
     trace_active_ (false),
     tracing_enabled_ (true), // On by default?
-    delete_ostream_(false),
     thr_desc_ (0),
     priority_mask_ (default_priority_mask_),
     timestamp_ (0)
@@ -743,22 +743,28 @@
         }
     }
 
-  //
-  // do we need to close and clean up?
-  //
-  if (this->delete_ostream_)
+  this->cleanup_ostream ();
+
+  delete[] this->msg_;
+}
+
+void
+ACE_Log_Msg::cleanup_ostream ()
+{
+  if (this->ostream_refcount_)
+    {
+      if (--*this->ostream_refcount_ == 0)
+        {
+          delete this->ostream_refcount_;
 #if defined (ACE_LACKS_IOSTREAM_TOTALLY)
-    {
-      ACE_OS::fclose (this->ostream_);
-    }
+          ACE_OS::fclose (this->ostream_);
 #else
-    {
-      delete ostream_;
-      ostream_ = 0;
+          delete this->ostream_;
+          this->ostream_ = 0;
+#endif
+        }
+      this->ostream_refcount_ = 0;
     }
-#endif
-
-  delete[] this->msg_;
 }
 
 // Open the sender-side of the message queue.
@@ -2412,18 +2418,30 @@
 ACE_Log_Msg::msg_ostream (ACE_OSTREAM_TYPE *m, bool delete_ostream)
 {
   if (this->ostream_ == m)
-    return;
+    {
+      // Same stream, allow user to change the delete_ostream "flag"
+      if (delete_ostream && !this->ostream_refcount_)
+        {
+          ACE_NEW (this->ostream_refcount_, Atomic_ULong (1));
+        }
+      else if (!delete_ostream && this->ostream_refcount_)
+        {
+          if (--*this->ostream_refcount_ == 0)
+            {
+              delete this->ostream_refcount_;
+            }
+          this->ostream_refcount_ = 0;
+        }
+      return;
+    }
 
-  if (this->delete_ostream_)
+  this->cleanup_ostream ();
+
+  if (delete_ostream)
     {
-#if defined (ACE_LACKS_IOSTREAM_TOTALLY)
-      ACE_OS::fclose (this->ostream_);
-#else
-      delete this->ostream_;
-#endif
+      ACE_NEW (this->ostream_refcount_, Atomic_ULong (1));
     }
 
-  this->delete_ostream_ = delete_ostream;
   this->ostream_ = m;
 }
 
@@ -2477,6 +2495,15 @@
     {
       ACE_Log_Msg *inherit_log = ACE_LOG_MSG;
       attributes.ostream_ = inherit_log->msg_ostream ();
+      if (attributes.ostream_ && inherit_log->ostream_refcount_)
+        {
+          ++*inherit_log->ostream_refcount_;
+          attributes.ostream_refcount_ = inherit_log->ostream_refcount_;
+        }
+      else
+        {
+          attributes.ostream_refcount_ = 0;
+        }
       attributes.priority_mask_ = inherit_log->priority_mask ();
       attributes.tracing_enabled_ = inherit_log->tracing_enabled ();
       attributes.restart_ = inherit_log->restart ();
@@ -2501,7 +2528,10 @@
 
   if (attributes.ostream_)
     {
-      new_log->msg_ostream (attributes.ostream_);
+      new_log->ostream_ = attributes.ostream_;
+      new_log->ostream_refcount_ =
+        static_cast<Atomic_ULong *> (attributes.ostream_refcount_);
+
       new_log->priority_mask (attributes.priority_mask_);
 
       if (attributes.tracing_enabled_)
Index: ace/OS_Log_Msg_Attributes.h
===================================================================
--- ace/OS_Log_Msg_Attributes.h	(revision 83617)
+++ ace/OS_Log_Msg_Attributes.h	(working copy)
@@ -50,6 +50,7 @@
 
   /// Ostream where the new TSS Log_Msg will use.
   ACE_OSTREAM_TYPE *ostream_;
+  void *ostream_refcount_;
 
   /// Priority_mask to be used in new TSS Log_Msg.
   unsigned long priority_mask_;
Index: tests/tests.mpc
===================================================================
--- tests/tests.mpc	(revision 83617)
+++ tests/tests.mpc	(working copy)
@@ -1601,3 +1601,10 @@
     Process_Env_Test.cpp
   }
 }
+
+project(Log_Thread_Inheritance_Test) : acetest {
+  exename = Log_Thread_Inheritance_Test
+  Source_Files {
+    Log_Thread_Inheritance_Test.cpp
+  }
+}
Index: tests/Log_Thread_Inheritance_Test.cpp
===================================================================
--- tests/Log_Thread_Inheritance_Test.cpp	(revision 0)
+++ tests/Log_Thread_Inheritance_Test.cpp	(revision 0)
@@ -0,0 +1,64 @@
+// $Id$
+
+#include "test_config.h"
+#include "ace/Log_Msg.h"
+#include "ace/Task.h"
+#include <fstream>
+
+ACE_RCSID (tests, Log_Thread_Inheritance_Test, "$Id$");
+
+#if !defined (ACE_HAS_THREADS) || defined (ACE_LACKS_IOSTREAM_TOTALLY)
+int run_main (int, ACE_TCHAR *[])
+{
+  return 0;
+}
+#else
+
+struct MyThread : ACE_Task_Base
+{
+  explicit MyThread (bool openfile = false)
+    : openfile_ (openfile) {}
+
+  bool openfile_;
+  static MyThread childthread_;
+
+  int svc ()
+  {
+    if (openfile_)
+      {
+        ACE_LOG_MSG->msg_ostream (
+          new std::ofstream (ACE_LOG_DIRECTORY
+                             ACE_TEXT ("Log_Thread_Inheritance_Ostream")
+                             ACE_LOG_FILE_EXT_NAME), true);
+        ACE_LOG_MSG->set_flags (ACE_Log_Msg::OSTREAM);
+        ACE_LOG_MSG->clr_flags (ACE_Log_Msg::STDERR | ACE_Log_Msg::LOGGER);
+        childthread_.activate ();
+      }
+    ACE_DEBUG ((LM_DEBUG, "(%P|%t) - in svc() openfile_ is %s\n",
+                (openfile_ ? "true" : "false")));
+    if (!openfile_)
+      {
+        for (int i = 0; i < 100; ++i)
+          {
+            ACE_DEBUG ((LM_DEBUG, "(%P|%t) - loop %d\n", i));
+          }
+      }
+    return 0;
+  }
+};
+
+MyThread MyThread::childthread_;
+
+int run_main (int, ACE_TCHAR *[])
+{
+  ACE_START_TEST (ACE_TEXT ("Log_Thread_Inheritance_Test"));
+  ACE_DEBUG ((LM_DEBUG, "(%P|%t) - in run_main()\n"));
+  MyThread mt (true);
+  mt.activate ();
+  mt.wait ();
+  MyThread::childthread_.wait ();
+  ACE_END_TEST;
+  return 0;
+}
+
+#endif

Property changes on: tests\Log_Thread_Inheritance_Test.cpp
___________________________________________________________________
Added: svn:keywords
   + Author Date Id Revision
Added: svn:eol-style
   + native

