Index: ace/Base_Thread_Adapter.cpp
===================================================================
--- ace/Base_Thread_Adapter.cpp	(revision 84338)
+++ ace/Base_Thread_Adapter.cpp	(working copy)
@@ -39,6 +39,7 @@
   , entry_point_ (entry_point)
   , thr_desc_ (td)
   , ctx_ (ACE_Service_Config::current())
+  , log_msg_attributes_ ()
 {
   ACE_OS_TRACE ("ACE_Base_Thread_Adapter::ACE_Base_Thread_Adapter");
 
Index: tests/Log_Thread_Inheritance_Test.cpp
===================================================================
--- tests/Log_Thread_Inheritance_Test.cpp	(revision 84338)
+++ tests/Log_Thread_Inheritance_Test.cpp	(working copy)
@@ -5,6 +5,10 @@
 #include "ace/Task.h"
 #include <fstream>
 
+#ifdef ACE_HAS_PTHREADS
+# include <pthread.h>
+#endif /* ACE_HAS_PTHREADS */
+
 ACE_RCSID (tests, Log_Thread_Inheritance_Test, "$Id$")
 
 #if !defined (ACE_HAS_THREADS) || defined (ACE_LACKS_IOSTREAM_TOTALLY)
@@ -14,6 +18,47 @@
 }
 #else
 
+#ifdef ACE_HAS_PTHREADS
+struct Inheritor : ACE_Task_Base
+{
+  int svc ()
+  {
+    ACE_DEBUG ((LM_DEBUG, "(%P|%t) - this test might crash ACE if it does not "
+                          "have the fix for the second bug in #3480.\n"));
+    return 0; 
+  }
+};
+
+void* spawn_ace_task (void*)
+{
+  Inheritor inheritor;
+  
+  inheritor.activate ();
+  inheritor.wait ();
+
+  return 0;
+}
+
+bool test_inherited_attributes ()
+{
+  // This test verifies ACE_OS_Log_Msg_Attributes correctly initializes
+  // when an ACE thread is created from a non-ACE thread (i.e. pthreads)
+  // and is then used for logging.
+  // 
+  // This test will cause occasional SEGVs on failure.
+  // stallions 2009/02/05
+  pthread_t parent; 
+
+  if (pthread_create (&parent, 0, spawn_ace_task, 0) != 0)
+    {
+      return false;
+    }
+  pthread_join (parent, 0);
+
+  return true;
+}
+#endif /* ACE_HAS_PTHREADS */
+
 struct MyThread : ACE_Task_Base
 {
 
@@ -72,6 +117,9 @@
   mt.wait ();
   MyThread::childthread_.wait ();
   ACE_LOG_MSG->msg_ostream (initial_stream, 0);
+#ifdef ACE_HAS_PTHREADS 
+  if (!test_inherited_attributes ()) return -1;
+#endif /* ACE_HAS_PTHREADS */
   ACE_DEBUG ((LM_DEBUG, "(%P|%t) - Test passed.\n"));
   ACE_END_TEST;
   return 0;
Index: ChangeLog
===================================================================
--- ChangeLog	(revision 84338)
+++ ChangeLog	(working copy)
@@ -1,3 +1,12 @@
+Thu Feb  5 21:14:15 UTC 2009  Steven Stallion  <stallions@ociweb.com>
+
+        * ace/Base_Thread_Adapter.cpp:
+        * tests/Log_Thread_Inheritance_Test.cpp:
+
+          Fixed a SEGV caused by ACE_Log_Msg where ACE threads inheriting
+          from non-ACE threads copied un-initialized ACE_OS_Log_Msg_Attributes
+          members. Fixes Bugzilla #3480.
+
 Thu Feb  5 15:11:13 UTC 2009  Adam Mitz  <mitza@ociweb.com>
 
         * include/makeinclude/rules.local.GNU:
