Index: Module.h
===================================================================
--- Module.h	(revision 76434)
+++ Module.h	(working copy)
@@ -40,8 +40,8 @@
 public:
   enum
   {
-    /// Indicates that <close> should not delete any Tasks.
-    M_DELETE_NONE = 0,
+    /// Indicates that the flags have not been set
+    M_FLAGS_NOT_SET = 0,
 
     /// Indicates that <close> should delete the writer Task.
     M_DELETE_READER = 1,
@@ -56,7 +56,10 @@
      * The <M_DELETE_READER> and <M_DELETE_WRITER> flags may be or'ed
      * together.
      */
-    M_DELETE = 3
+    M_DELETE = 3,
+
+    /// Indicates that <close> should not delete any Tasks.
+    M_DELETE_NONE = 4,
   };
 };
 
Index: Module.cpp
===================================================================
--- Module.cpp	(revision 76434)
+++ Module.cpp	(working copy)
@@ -158,7 +158,7 @@
 
 template <ACE_SYNCH_DECL>
 ACE_Module<ACE_SYNCH_USE>::ACE_Module (void)
-  : flags_ (0)
+  : flags_ (M_FLAGS_NOT_SET)
 {
   ACE_TRACE ("ACE_Module<ACE_SYNCH_USE>::ACE_Module");
   this->name (ACE_LIB_TEXT ("<unknown>"));
@@ -183,7 +183,7 @@
                                        ACE_Task<ACE_SYNCH_USE> *reader_q,
                                        void *args,
                                        int flags /* = M_DELETE */)
-  : flags_ (0)
+  : flags_ (M_FLAGS_NOT_SET)
 {
   ACE_TRACE ("ACE_Module<ACE_SYNCH_USE>::ACE_Module");
 
@@ -205,7 +205,7 @@
 
   // Only pay attention to the flags parameter if we haven't already
   // set the task delete policies.
-  if (this->flags_ == 0)
+  if (this->flags_ == M_FLAGS_NOT_SET)
     ACE_SET_BITS (flags_, flags);
 
   if (this->close_i (0, flags_) == -1)
