Index: ace/OS_NS_Thread.cpp
===================================================================
--- ace/OS_NS_Thread.cpp	(revision 91473)
+++ ace/OS_NS_Thread.cpp	(revision 91472)
@@ -707,7 +707,7 @@
 
   // scope the guard
   {
-    ACE_GUARD_REACTION(ACE_Thread_Mutex, guard, *mutex_, ACE_UNEXPECTED(;));
+    ACE_Guard<ACE_Thread_Mutex> guard (*mutex_);
     if (ptr_ != 0)
       {
         if (ACE_BIT_ENABLED (flags_, FLAG_DELETING))
@@ -1345,15 +1345,11 @@
 
 #   if defined (ACE_HAS_SIGNAL_OBJECT_AND_WAIT)
   if (external_mutex->type_ == USYNC_PROCESS)
-    {
-      // This call will automatically release the mutex and wait on the semaphore.
-      ACE_WIN32CALL (ACE_ADAPT_RETVAL (::SignalObjectAndWait (external_mutex->proc_mutex_,
-                                                              cv->sema_, INFINITE, FALSE),
-                                      result),
-                    int, -1, result);
-      if (result == -1)
-        return result;
-    }
+    // This call will automatically release the mutex and wait on the semaphore.
+    ACE_WIN32CALL (ACE_ADAPT_RETVAL (::SignalObjectAndWait (external_mutex->proc_mutex_,
+                                                            cv->sema_, INFINITE, FALSE),
+                                     result),
+                   int, -1, result);
   else
 #   endif /* ACE_HAS_SIGNAL_OBJECT_AND_WAIT */
     {
@@ -1403,8 +1399,7 @@
         // We must always regain the <external_mutex>, even when
         // errors occur because that's the guarantee that we give to
         // our callers.
-        if (ACE_OS::mutex_lock (external_mutex) != 0)
-          return -1;
+        ACE_OS::mutex_lock (external_mutex);
 
       return result;
       /* NOTREACHED */
@@ -1563,13 +1558,10 @@
                                          result),
                        int, -1, result);
       else
-        {
-          // We must always regain the <external_Mutex>, even when
-          // errors occur because that's the guarantee that we give to
-          // our callers.
-          if (ACE_OS::mutex_lock (external_mutex) != 0)
-            return -1;
-        }
+        // We must always regain the <external_Mutex>, even when
+        // errors occur because that's the guarantee that we give to
+        // our callers.
+        ACE_OS::mutex_lock (external_mutex);
 
       return result;
       /* NOTREACHED */
@@ -1579,20 +1571,16 @@
   // if" statement since the caller may have timed out and hence the
   // result would have been -1 above.
   if (last_waiter)
-    {
-      // Release the signaler/broadcaster if we're the last waiter.
+    // Release the signaler/broadcaster if we're the last waiter.
 #     if defined (ACE_WIN32)
-      if (ACE_OS::event_signal (&cv->waiters_done_) != 0)
+    ACE_OS::event_signal (&cv->waiters_done_);
 #     else
-      if (ACE_OS::sema_post (&cv->waiters_done_) != 0)
+    ACE_OS::sema_post (&cv->waiters_done_);
 #     endif /* ACE_WIN32 */
-        return -1;
-    }
 
   // We must always regain the <external_mutex>, even when errors
   // occur because that's the guarantee that we give to our callers.
-  if (ACE_OS::mutex_lock (external_mutex) != 0)
-    return -1;
+  ACE_OS::mutex_lock (external_mutex);
 
   return result;
 #   endif /* ACE_HAS_WTHREADS || ACE_HAS_VXWORKS */
@@ -1716,24 +1704,15 @@
     }
 
   if (last_waiter)
-    {
-      // Release the signaler/broadcaster if we're the last waiter.
-      if (ACE_OS::event_signal (&cv->waiters_done_) != 0)
-        return -1;
-    }
-    
+    // Release the signaler/broadcaster if we're the last waiter.
+    ACE_OS::event_signal (&cv->waiters_done_);
+
   // We must always regain the <external_mutex>, even when errors
   // occur because that's the guarantee that we give to our callers.
   if (ACE_OS::thread_mutex_lock (external_mutex) != 0)
     result = -1;
 
-  if (error != 0)
-    {
-      /* This assignment must only be done if error != 0,
-       *   since writing 0 to errno violates the POSIX specification.
-       */
-      errno = error;
-    }
+  errno = error;
   return result;
 #   endif
 #   else
@@ -1806,11 +1785,8 @@
         }
     }
   else if (last_waiter)
-    {
-      // Release the signaler/broadcaster if we're the last waiter.
-      if (ACE_OS::event_signal (&cv->waiters_done_) != 0)
-        return -1;
-    }
+    // Release the signaler/broadcaster if we're the last waiter.
+    ACE_OS::event_signal (&cv->waiters_done_);
 
   // We must always regain the <external_mutex>, even when errors
   // occur because that's the guarantee that we give to our callers.
@@ -1818,13 +1794,7 @@
     result = -1;
 
   // Reset errno in case mutex_lock() also fails...
-  if (error != 0)
-  {
-    /* This assignment must only be done if error != 0,
-    *   since writing 0 to errno violates the POSIX specification.
-    */
-    errno = error;
-  }
+  errno = error;
   return result;
 #endif
 #   else
@@ -2885,11 +2855,9 @@
         (!defined (ACE_LACKS_MUTEXATTR_PSHARED) || !defined (ACE_LACKS_CONDATTR_PSHARED))) || \
      (!defined (ACE_USES_FIFO_SEM) && \
         (!defined (ACE_HAS_POSIX_SEM) || !defined (ACE_HAS_POSIX_SEM_TIMEOUT) || defined (ACE_LACKS_NAMED_POSIX_SEM)))
-    if (ACE_OS::mutex_unlock (&event->eventdata_->lock_) != 0)
-      return -1;
+    ACE_OS::mutex_unlock (&event->eventdata_->lock_);
 # else
-    if (ACE_OS::sema_post (&event->lock_) != 0)
-      return -1;
+    ACE_OS::sema_post (&event->lock_);
 # endif
     if (result == -1)
       // Reset errno in case mutex_unlock() also fails...
@@ -2931,11 +2899,9 @@
         (!defined (ACE_LACKS_MUTEXATTR_PSHARED) || !defined (ACE_LACKS_CONDATTR_PSHARED))) || \
      (!defined (ACE_USES_FIFO_SEM) && \
         (!defined (ACE_HAS_POSIX_SEM) || !defined (ACE_HAS_POSIX_SEM_TIMEOUT) || defined (ACE_LACKS_NAMED_POSIX_SEM)))
-    if (ACE_OS::mutex_unlock (&event->eventdata_->lock_) != 0)
-      return -1;
+    ACE_OS::mutex_unlock (&event->eventdata_->lock_);
 # else
-    if (ACE_OS::sema_post (&event->lock_) != 0)
-      return -1;
+    ACE_OS::sema_post (&event->lock_);
 # endif
   }
   else
@@ -3017,11 +2983,9 @@
         (!defined (ACE_LACKS_MUTEXATTR_PSHARED) || !defined (ACE_LACKS_CONDATTR_PSHARED))) || \
      (!defined (ACE_USES_FIFO_SEM) && \
         (!defined (ACE_HAS_POSIX_SEM) || !defined (ACE_HAS_POSIX_SEM_TIMEOUT) || defined (ACE_LACKS_NAMED_POSIX_SEM)))
-    if (ACE_OS::mutex_unlock (&event->eventdata_->lock_) != 0)
-      return -1;
+    ACE_OS::mutex_unlock (&event->eventdata_->lock_);
 # else
-    if (ACE_OS::sema_post (&event->lock_) != 0)
-      return -1;
+    ACE_OS::sema_post (&event->lock_);
 # endif
 
     if (result == -1)
@@ -3217,11 +3181,9 @@
         (!defined (ACE_LACKS_MUTEXATTR_PSHARED) || !defined (ACE_LACKS_CONDATTR_PSHARED))) || \
      (!defined (ACE_USES_FIFO_SEM) && \
         (!defined (ACE_HAS_POSIX_SEM) || !defined (ACE_HAS_POSIX_SEM_TIMEOUT) || defined (ACE_LACKS_NAMED_POSIX_SEM)))
-      if (ACE_OS::mutex_unlock (&event->eventdata_->lock_) != 0)
-        return -1;
+      ACE_OS::mutex_unlock (&event->eventdata_->lock_);
 # else
-      if (ACE_OS::sema_post (&event->lock_) != 0)
-        return -1;
+      ACE_OS::sema_post (&event->lock_);
 # endif
 
       if (result == -1)
@@ -3361,11 +3323,9 @@
         (!defined (ACE_LACKS_MUTEXATTR_PSHARED) || !defined (ACE_LACKS_CONDATTR_PSHARED))) || \
      (!defined (ACE_USES_FIFO_SEM) && \
         (!defined (ACE_HAS_POSIX_SEM) || !defined (ACE_HAS_POSIX_SEM_TIMEOUT) || defined (ACE_LACKS_NAMED_POSIX_SEM)))
-      if (ACE_OS::mutex_unlock (&event->eventdata_->lock_) != 0)
-        return -1;
+      ACE_OS::mutex_unlock (&event->eventdata_->lock_);
 # else
-      if (ACE_OS::sema_post (&event->lock_) != 0)
-        return -1;
+      ACE_OS::sema_post (&event->lock_);
 # endif
 
       if (result == -1)
@@ -3516,14 +3476,10 @@
     {
       // Save/restore errno.
       ACE_Errno_Guard error (errno);
-
-      /* We're about to return -1 anyway, so
-       * no need to check return values of these clean-up calls:
-       */
-      (void)ACE_OS::mutex_destroy (&rw->lock_);
-      (void)ACE_OS::cond_destroy (&rw->waiting_readers_);
-      (void)ACE_OS::cond_destroy (&rw->waiting_writers_);
-      (void)ACE_OS::cond_destroy (&rw->waiting_important_writer_);
+      ACE_OS::mutex_destroy (&rw->lock_);
+      ACE_OS::cond_destroy (&rw->waiting_readers_);
+      ACE_OS::cond_destroy (&rw->waiting_writers_);
+      ACE_OS::cond_destroy (&rw->waiting_important_writer_);
     }
   return result;
 # else
Index: ChangeLog
===================================================================
--- ChangeLog	(revision 91473)
+++ ChangeLog	(revision 91472)
@@ -1,9 +1,3 @@
-Thu Aug 26 10:20:41 UTC 2010  Martin Corino  <mcorino@remedy.nl>
-
-        * ace/OS_NS_Thread.cpp:
-          Changes implementing the patches from Bugzilla #3653 (with
-          changes and some rejects; *no* asserts allowed in ACE).
-
 Wed Aug 25 19:58:19 UTC 2010  Steve Huston  <shuston@riverace.com>
 
         * ace/Dev_Poll_Reactor.{h cpp}: Regarding change number 2 at
