Index: Connector.cpp
===================================================================
--- Connector.cpp	(revision 86155)
+++ Connector.cpp	(working copy)
@@ -30,6 +30,9 @@
 
   this->reference_counting_policy ().value
     (ACE_Event_Handler::Reference_Counting_Policy::ENABLED);
+
+  if (this->svc_handler_ != 0)
+    this->svc_handler_->add_reference ();
 }
 
 template <class SVC_HANDLER> SVC_HANDLER *
@@ -129,6 +132,9 @@
     svc_handler->handle_close (svc_handler->get_handle (),
                                ACE_Event_Handler::TIMER_MASK);
 
+  if (svc_handler != 0)
+    svc_handler->remove_reference ();
+
   return retval;
 }
 
@@ -145,8 +151,12 @@
 
   // Close Svc_Handler.
   if (svc_handler != 0)
-    svc_handler->close (NORMAL_CLOSE_OPERATION);
+    {
+      svc_handler->close (NORMAL_CLOSE_OPERATION);
 
+      svc_handler->remove_reference ();
+    }
+
   return retval;
 }
 
@@ -162,8 +172,12 @@
   int const retval = this->close (svc_handler) ? 0 : -1;
 
   if (svc_handler != 0)
-    connector.initialize_svc_handler (handle, svc_handler);
+    {
+      connector.initialize_svc_handler (handle, svc_handler);
 
+      svc_handler->remove_reference ();
+    }
+
   return retval;
 }
 
