Index: HTIOP/HTIOP_Connector.cpp
===================================================================
--- HTIOP/HTIOP_Connector.cpp	(revision 86580)
+++ HTIOP/HTIOP_Connector.cpp	(working copy)
@@ -217,6 +217,9 @@
   this->connect_creation_strategy_->make_svc_handler (svc_handler);
   // we now have a connection handler that has an unconnected stream
 
+  // Make sure that we always do a remove_reference
+  ACE_Event_Handler_var svc_handler_auto_ptr (svc_handler);
+
   svc_handler->peer().session(session);
   session->handler(svc_handler);
   ACE::HTBP::Channel *outbound = session->outbound();
@@ -233,10 +236,6 @@
   // strategy, which does not apply here. Therefore the whole bit of logic
   // of dealing with a failed wait but an unclosed svc_handler is skipped.
 
-
-  // Regardless of success or failure, remove the extra #REFCOUNT#.
-  svc_handler->remove_reference ();
-
   if (closed) // would be result == -1 in IIOP_Connector
     {
       if (TAO_debug_level)
@@ -299,6 +298,7 @@
       return 0;
     }
 
+  svc_handler_auto_ptr.release ();
   return transport;
 }
 
