Index: SSLIOP/SSLIOP_Connector.cpp
===================================================================
--- SSLIOP/SSLIOP_Connector.cpp	(revision 86580)
+++ SSLIOP/SSLIOP_Connector.cpp	(working copy)
@@ -479,9 +479,9 @@
         this->retrieve_credentials (resolver->stub (),
                                     svc_handler->peer ().ssl ());
 
-      safe_handler.release ();
-
       ssl_endpoint->set_sec_attrs (qop, trust, credentials.in());
+
+      safe_handler.release ();
     }
 
   // Check the Cache first for connections
@@ -493,19 +493,6 @@
         == TAO::Transport_Cache_Manager::CACHE_FOUND_AVAILABLE)
     {
       // ...eliminate svc_handle memory leak...
-      // The make_svc_handler() method creates the service handler and
-      // bumps the #REFCOUNT# up one extra.  The extra reference count
-      // in TAO_Connect_Creation_Strategy::make_svc_handler() is
-      // needed in the case when connection completion is pending and
-      // we are going to wait on a variable in the handler to changes,
-      // signifying success or failure.  Note, that this increment
-      // cannot be done once the connect() returns since this might be
-      // too late if another thread pick up the completion and
-      // potentially deletes the handler before we get a chance to
-      // increment the reference count.
-      if (svc_handler)
-          svc_handler->remove_reference();
-
       ACE_Event_Handler_var
         safe_handler (svc_handler);
 
@@ -529,6 +516,8 @@
 
             }
         }
+
+      safe_handler.release ();
     }
   else
     {
@@ -546,17 +535,6 @@
       // ACE_Strategy_Connector (the "base_connector_").  This is
       // thread-safe and reentrant, hence no synchronization is
       // necessary.
-      //
-      // The make_svc_handler() method creates the service handler and
-      // bumps the #REFCOUNT# up one extra.  The extra reference count
-      // in TAO_Connect_Creation_Strategy::make_svc_handler() is
-      // needed in the case when connection completion is pending and
-      // we are going to wait on a variable in the handler to changes,
-      // signifying success or failure.  Note, that this increment
-      // cannot be done once the connect() returns since this might be
-      // too late if another thread pick up the completion and
-      // potentially deletes the handler before we get a chance to
-      // increment the reference count.
       if (svc_handler == 0 &&
           this->base_connector_.creation_strategy ()->make_svc_handler (
                svc_handler) != 0)
@@ -634,20 +612,6 @@
                                               remote_address,
                                               synch_options);
 
-      // base_connector_.connect() will increment the handler's
-      // #REFCOUNT# once more. This is not required as we already hold
-      // a reference to the handler, so we discard this second
-      // reference.
-      svc_handler->remove_reference ();
-
-      // There are three possibilities from calling connect(): (a)
-      // connection succeeds immediately - in this case, the
-      // #REFCOUNT# on the handler is two; (b) connection completion
-      // is pending - in this case, the #REFCOUNT# on the handler is
-      // also two; (c) connection fails immediately - in this case,
-      // the #REFCOUNT# on the handler is one since close() gets
-      // called on the handler.
-
       // Make sure that we always do a remove_reference
       ACE_Event_Handler_var svc_handler_auto_ptr (svc_handler);
 
@@ -772,6 +736,8 @@
 
           return 0;
         }
+
+      svc_handler_auto_ptr.release ();
     }
 
   return transport;
Index: SSLIOP/IIOP_SSL_Connector.cpp
===================================================================
--- SSLIOP/IIOP_SSL_Connector.cpp	(revision 86580)
+++ SSLIOP/IIOP_SSL_Connector.cpp	(working copy)
@@ -147,24 +147,6 @@
   int result =
     this->base_connector_.connect (svc_handler, remote_address, synch_options);
 
-  // The connect() method creates the service handler and bumps the
-  // #REFCOUNT# up one extra.  There are three possibilities from
-  // calling connect(): (a) connection succeeds immediately - in this
-  // case, the #REFCOUNT# on the handler is two; (b) connection
-  // completion is pending - in this case, the #REFCOUNT# on the
-  // handler is also two; (c) connection fails immediately - in this
-  // case, the #REFCOUNT# on the handler is one since close() gets
-  // called on the handler.
-  //
-  // The extra reference count in
-  // TAO_Connect_Creation_Strategy::make_svc_handler() is needed in
-  // the case when connection completion is pending and we are going
-  // to wait on a variable in the handler to changes, signifying
-  // success or failure.  Note, that this increment cannot be done
-  // once the connect() returns since this might be too late if
-  // another thread pick up the completion and potentially deletes the
-  // handler before we get a chance to increment the reference count.
-
   // Make sure that we always do a remove_reference
   ACE_Event_Handler_var svc_handler_auto_ptr (svc_handler);
 
@@ -213,9 +195,9 @@
       return 0;
     }
 
-  if (transport->connection_handler ()->keep_waiting ())
+  if (svc_handler->keep_waiting ())
     {
-      svc_handler->add_reference ();
+      svc_handler->connection_pending ();
     }
 
   // At this point, the connection has be successfully connected.
@@ -271,6 +253,7 @@
       return 0;
     }
 
+  svc_handler_auto_ptr.release ();
   return transport;
 }
 
