Index: Malloc_Base.h
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/Malloc_Base.h,v
retrieving revision 4.18
retrieving revision 4.19
diff -u -r4.18 -r4.19
--- Malloc_Base.h	19 Apr 2006 11:51:53 -0000	4.18
+++ Malloc_Base.h	14 Jul 2006 21:27:23 -0000	4.19
@@ -4,7 +4,7 @@
 /**
  *  @file    Malloc_Base.h
  *
- *  $Id: Malloc_Base.h,v 4.18 2006/04/19 11:51:53 jwillemsen Exp $
+ *  $Id: Malloc_Base.h,v 4.19 2006/07/14 21:27:23 ossama Exp $
  *
  *  @author Doug Schmidt and Irfan Pyarali
  */
@@ -39,6 +39,10 @@
 class ACE_Export ACE_Allocator
 {
 public:
+
+  /// Unsigned integer type used for specifying memory block lengths.
+  typedef size_t size_type;
+
   // = Memory Management
 
   /// Get pointer to a default ACE_Allocator.
@@ -58,15 +62,15 @@
   virtual ~ACE_Allocator (void);
 
   /// Allocate <nbytes>, but don't give them any initial value.
-  virtual void *malloc (size_t nbytes) = 0;
+  virtual void *malloc (size_type nbytes) = 0;
 
   /// Allocate <nbytes>, giving them <initial_value>.
-  virtual void *calloc (size_t nbytes, char initial_value = '\0') = 0;
+  virtual void *calloc (size_type nbytes, char initial_value = '\0') = 0;
 
   /// Allocate <n_elem> each of size <elem_size>, giving them
   /// <initial_value>.
-  virtual void *calloc (size_t n_elem,
-                        size_t elem_size,
+  virtual void *calloc (size_type n_elem,
+                        size_type elem_size,
                         char initial_value = '\0') = 0;
 
   /// Free <ptr> (must have been allocated by <ACE_Allocator::malloc>).
@@ -127,7 +131,7 @@
 
   /// Sync @a len bytes of the memory region to the backing store
   /// starting at @a addr.
-  virtual int sync (void *addr, size_t len, int flags = MS_SYNC) = 0;
+  virtual int sync (void *addr, size_type len, int flags = MS_SYNC) = 0;
 
   /**
    * Change the protection of the pages of the mapped region to <prot>
@@ -138,7 +142,7 @@
 
   /// Change the protection of the pages of the mapped region to <prot>
   /// starting at <addr> up to <len> bytes.
-  virtual int protect (void *addr, size_t len, int prot = PROT_RDWR) = 0;
+  virtual int protect (void *addr, size_type len, int prot = PROT_RDWR) = 0;
 
 #if defined (ACE_HAS_MALLOC_STATS)
   /// Dump statistics of how malloc is behaving.
Index: String_Base_Const.h
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/String_Base_Const.h,v
retrieving revision 4.5
retrieving revision 4.6
diff -u -r4.5 -r4.6
--- String_Base_Const.h	28 Oct 2005 16:14:56 -0000	4.5
+++ String_Base_Const.h	14 Jul 2006 21:27:23 -0000	4.6
@@ -4,7 +4,7 @@
 /**
  *  @file    String_Base_Const.h
  *
- *  $Id: String_Base_Const.h,v 4.5 2005/10/28 16:14:56 ossama Exp $
+ *  $Id: String_Base_Const.h,v 4.6 2006/07/14 21:27:23 ossama Exp $
  *
  *  @author Nanbor Wang <nanbor@cs.wustl.edu>
  */
@@ -16,7 +16,7 @@
 #include /**/ "ace/pre.h"
 
 #include "ace/ACE_export.h"
-#include "ace/Basic_Types.h"
+#include "ace/Malloc_Base.h"
 
 #if !defined (ACE_LACKS_PRAGMA_ONCE)
 # pragma once
@@ -29,12 +29,18 @@
  * @class ACE_String_Base_Const
  *
  * @brief This class defines a constant for ACE_String_Base to
- * circumvent a bug in SunCC 6.0.
+ *        circumvent a bug in SunCC 6.0.
  */
 class ACE_Export ACE_String_Base_Const
 {
 public:
-  static ssize_t const npos;
+
+  typedef ACE_Allocator::size_type size_type;
+
+  /// Constant that denotes case where no such character position
+  /// exists.
+  static size_type const npos;
+
 };
 
 ACE_END_VERSIONED_NAMESPACE_DECL
Index: String_Base_Const.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/String_Base_Const.cpp,v
retrieving revision 4.3
retrieving revision 4.4
diff -u -r4.3 -r4.4
--- String_Base_Const.cpp	28 Oct 2005 16:14:56 -0000	4.3
+++ String_Base_Const.cpp	14 Jul 2006 21:27:23 -0000	4.4
@@ -1,9 +1,11 @@
-// $Id: String_Base_Const.cpp,v 4.3 2005/10/28 16:14:56 ossama Exp $
+// $Id: String_Base_Const.cpp,v 4.4 2006/07/14 21:27:23 ossama Exp $
 
 #include "ace/String_Base_Const.h"
+#include "ace/Numeric_Limits.h"
 
 ACE_BEGIN_VERSIONED_NAMESPACE_DECL
 
-ssize_t const ACE_String_Base_Const::npos = -1;
+ACE_String_Base_Const::size_type const ACE_String_Base_Const::npos =
+  ACE_Numeric_Limits<ACE_String_Base_Const::size_type>::max ();;
 
 ACE_END_VERSIONED_NAMESPACE_DECL
Index: String_Base.h
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/String_Base.h,v
retrieving revision 4.30
retrieving revision 4.32
diff -u -r4.30 -r4.32
--- String_Base.h	8 Jun 2006 17:27:13 -0000	4.30
+++ String_Base.h	14 Jul 2006 22:41:33 -0000	4.32
@@ -4,7 +4,7 @@
 /**
  *  @file    String_Base.h
  *
- *  $Id: String_Base.h,v 4.30 2006/06/08 17:27:13 schmidt Exp $
+ *  $Id: String_Base.h,v 4.32 2006/07/14 22:41:33 ossama Exp $
  *
  *  @author Douglas C. Schmidt (schmidt@cs.wustl.edu)
  *  @author Nanbor Wang <nanbor@cs.wustl.edu>
@@ -16,18 +16,16 @@
 
 #include /**/ "ace/pre.h"
 
-#include "ace/String_Base_Const.h"
+#include "ace/Global_Macros.h"
 
 #if !defined (ACE_LACKS_PRAGMA_ONCE)
 # pragma once
 #endif /* ACE_LACKS_PRAGMA_ONCE */
 
-#include "ace/Global_Macros.h"
+#include "ace/String_Base_Const.h"
 
 ACE_BEGIN_VERSIONED_NAMESPACE_DECL
 
-class ACE_Allocator;
-
 /**
  * @class ACE_String_Base
  *
@@ -45,10 +43,22 @@
  * CAUTION: in cases when ACE_String_Base is constructed from a
  * provided buffer with the release parameter set to 0,
  * ACE_String_Base is not guaranteed to be '\0' terminated.
+ *
+ * \li Do not use a "@c -1" magic number to refer to the "no position"
+ *     condition.  This was never the right thing to do.  The "@c npos"
+ *     constant should be used in such cases.
+ * \li Do not assign or pass string positions to or from signed types.
+ *     Use the "@c size_type" @c typedef found in all ACE string
+ *     classes.  This typedef is analogous to the "@c size_type"
+ *     @c typedef found in the standard C++ string class as well as
+ *     many STL class templates.  If you find yourself casting you're
+ *     probably doing something wrong.
  */
-template <class CHAR> class ACE_String_Base : public ACE_String_Base_Const
+template <class CHAR>
+class ACE_String_Base : public ACE_String_Base_Const
 {
 public:
+
    /**
     *  Default constructor.
     *
@@ -94,7 +104,7 @@
    * @return ACE_String_Base containing const CHAR *s
    */
   ACE_String_Base (const CHAR *s,
-                   size_t len,
+                   size_type len,
                    ACE_Allocator *the_allocator = 0,
                    int release = 1);
 
@@ -128,7 +138,7 @@
    *  @param the_allocator ACE_Allocator associated with string
    *  @return Empty ACE_String_Base with room for len CHARs
    */
-  ACE_String_Base (size_t len,
+  ACE_String_Base (size_type len,
                    CHAR c = 0,
                    ACE_Allocator *the_allocator = 0);
 
@@ -144,7 +154,7 @@
    * @param slot Index of the desired character
    * @return The character at index @a slot
    */
-  const CHAR & operator[] (size_t slot) const;
+  const CHAR & operator[] (size_type slot) const;
 
   /**
    * Return the <slot'th> character by reference in the string
@@ -153,7 +163,7 @@
    * @param slot Index of the desired character
    * @return The character at index @a slot
    */
-  CHAR & operator[] (size_t slot);
+  CHAR & operator[] (size_type slot);
 
   /**
    *  Assignment operator (does copy memory).
@@ -216,7 +226,7 @@
    *  @param release Allocator responsible(1)/not reponsible(0) for
    *    freeing memory.
    */
-  void set (const CHAR * s, size_t len, int release);
+  void set (const CHAR * s, size_type len, int release);
 
   /**
    * Clear this string. Memory is _not_ freed if <release> is 0.
@@ -252,16 +262,16 @@
   void fast_clear (void);
 
   /**
-   * Return a substring given an offset and length, if length == -1
-   * use rest of str.  Return empty substring if offset or
-   * offset/length are invalid.
+   * Return a substring given an offset and length.
+   * If length == @c npos use rest of str.  Return empty substring if
+   * offset or offset/length are invalid.
    *
    * @param offset Index of first desired character of the substring.
    * @param length How many characters to return starting at the offset.
    * @return The string containing the desired substring
    */
-  ACE_String_Base < CHAR > substring (size_t offset,
-                                      ssize_t length = -1) const;
+  ACE_String_Base < CHAR > substring (size_type offset,
+                                      size_type length = npos) const;
 
   /**
    *  Same as <substring>.
@@ -270,7 +280,8 @@
    * @param length How many characters to return starting at the offset.
    * @return The string containing the desired substring
    */
-  ACE_String_Base < CHAR > substr (size_t offset, ssize_t length = -1) const;
+  ACE_String_Base < CHAR > substr (size_type offset,
+                                   size_type length = npos) const;
 
   /**
    *  Concat operator (copies memory).
@@ -307,7 +318,7 @@
    *  @return The combined string (input append to the end of the old). New
    *    string is zero terminated.
    */
-  ACE_String_Base < CHAR >& append (const CHAR* s, size_t slen);
+  ACE_String_Base < CHAR >& append (const CHAR* s, size_type slen);
 
   /**
    *  Returns a hash value for this string.
@@ -321,7 +332,7 @@
    *
    *  @return Length of stored string
    */
-  size_t length (void) const;
+  size_type length (void) const;
 
   /**
    * Return @c true if the length of the string is zero, else @c false.
@@ -365,53 +376,57 @@
 
   /**
    *  Comparison operator that will match substrings.  Returns the
-   *  slot of the first location that matches, else -1.
+   *  slot of the first location that matches, else @c npos.
    *
    *  @param s Input ACE_String_Base string
    *  @return Integer index value of the first location of string @a s or
-   *    -1 (not found).
+   *          @c npos (not found).
    */
-  ssize_t strstr (const ACE_String_Base<CHAR> &s) const;
+  size_type strstr (const ACE_String_Base<CHAR> &s) const;
 
   /**
    *  Find <str> starting at pos.  Returns the slot of the first
-   *  location that matches (will be >= pos), else npos.
+   *  location that matches (will be >= pos), else @c npos.
    *
    *  @param str Input ACE_String_Base string to search for in stored string.
    *  @param pos Starting index position to start searching for string @a str.
-   *  @return Index value of the first location of string @a str else npos.
+   *  @return Index value of the first location of string @a str else
+   *          @c npos.
    */
-  ssize_t find (const ACE_String_Base<CHAR> &str, size_t pos = 0) const;
+  size_type find (const ACE_String_Base<CHAR> &str, size_type pos = 0) const;
 
   /**
    *  Find @a s starting at pos.  Returns the slot of the first
-   *  location that matches (will be >= pos), else npos.
+   *  location that matches (will be >= pos), else @c npos.
    *
    *  @param s non-zero input string to search for in stored string.
    *  @param pos Starting index position to start searching for string @a str.
-   *  @return Index value of the first location of string @a str else npos.
+   *  @return Index value of the first location of string @a str else
+   *          @c npos.
    */
-  ssize_t find (const CHAR *s, size_t pos = 0) const;
+  size_type find (const CHAR *s, size_type pos = 0) const;
 
   /**
    *  Find @a c starting at pos.  Returns the slot of the first
-   *  location that matches (will be >= pos), else npos.
+   *  location that matches (will be >= pos), else @c npos.
    *
    *  @param c Input character to search for in stored string.
    *  @param pos Starting index position to start searching for string @a str.
-   *  @return Index value of the first location of string @a str else npos.
+   *  @return Index value of the first location of string @a str else
+   *          @c npos.
    */
-  ssize_t find (CHAR c, size_t pos = 0) const;
+  size_type find (CHAR c, size_type pos = 0) const;
 
   /**
    *  Find @a c starting at pos (counting from the end).  Returns the
-   *  slot of the first location that matches, else npos.
+   *  slot of the first location that matches, else @c npos.
    *
    *  @param c Input character to search for in stored string.
    *  @param pos Starting index position to start searching for string @a str.
-   *  @return Index value of the first location of string @a str else npos.
+   *  @return Index value of the first location of string @a str else
+   *          @c npos.
    */
-  ssize_t rfind (CHAR c, ssize_t pos = npos) const;
+  size_type rfind (CHAR c, size_type pos = npos) const;
 
   /**
    *  Equality comparison operator (must match entire string).
@@ -491,8 +506,14 @@
    * @param len The number of CHARs to reserve
    * @param c The CHAR to use when filling the string.
    */
-  void resize (size_t len, CHAR c = 0);
+  void resize (size_type len, CHAR c = 0);
 
+  /// Swap the contents of this @c ACE_String_Base with @a str.
+  /**
+   * @note This is non-throwing operation.
+   */
+  void swap (ACE_String_Base<CHAR> & str);
+  
   /**
    *  Declare the dynamic allocation hooks.
    */
@@ -507,13 +528,13 @@
   /**
    *  Length of the ACE_String_Base data (not counting the trailing '\0').
    */
-  size_t len_;
+  size_type len_;
 
   /**
    *  Length of the ACE_String_Base data buffer.  Keeping track of the
    *  length allows to avoid unnecessary dynamic allocations.
    */
-  size_t buf_len_;
+  size_type buf_len_;
 
   /**
    *  Pointer to data.
Index: String_Base.inl
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/String_Base.inl,v
retrieving revision 4.9
retrieving revision 4.10
diff -u -r4.9 -r4.10
--- String_Base.inl	4 Jun 2006 14:03:12 -0000	4.9
+++ String_Base.inl	14 Jul 2006 21:27:23 -0000	4.10
@@ -1,6 +1,6 @@
 // -*- C++ -*-
 //
-// $Id: String_Base.inl,v 4.9 2006/06/04 14:03:12 schmidt Exp $
+// $Id: String_Base.inl,v 4.10 2006/07/14 21:27:23 ossama Exp $
 
 #include "ace/Malloc_Base.h"
 #include "ace/Min_Max.h"
@@ -26,7 +26,7 @@
   return *this;
 }
 
-template <class CHAR> ACE_INLINE size_t
+template <class CHAR> ACE_INLINE typename ACE_String_Base<CHAR>::size_type
 ACE_String_Base<CHAR>::length (void) const
 {
   ACE_TRACE ("ACE_String_Base<CHAR>::length");
@@ -46,8 +46,8 @@
 }
 
 template <class CHAR> ACE_INLINE ACE_String_Base<CHAR>
-ACE_String_Base<CHAR>::substr (size_t offset,
-                               ssize_t length) const
+ACE_String_Base<CHAR>::substr (size_type offset,
+                               size_type length) const
 {
   ACE_TRACE ("ACE_String_Base<CHAR>::substr");
   return this->substring (offset, length);
@@ -56,7 +56,7 @@
 // Return the <slot'th> character in the string.
 
 template <class CHAR> ACE_INLINE const CHAR &
-ACE_String_Base<CHAR>::operator[] (size_t slot) const
+ACE_String_Base<CHAR>::operator[] (size_type slot) const
 {
   ACE_TRACE ("ACE_String_Base<CHAR>::operator[]");
   return this->rep_[slot];
@@ -65,7 +65,7 @@
 // Return the <slot'th> character in the string by reference.
 
 template <class CHAR> ACE_INLINE CHAR &
-ACE_String_Base<CHAR>::operator[] (size_t slot)
+ACE_String_Base<CHAR>::operator[] (size_type slot)
 {
   ACE_TRACE ("ACE_String_Base<CHAR>::operator[]");
   return this->rep_[slot];
@@ -117,14 +117,15 @@
   return !(*this == s);
 }
 
-template <class CHAR> ACE_INLINE ssize_t
-ACE_String_Base<CHAR>::find (const ACE_String_Base<CHAR>&str, size_t pos) const
+template <class CHAR> ACE_INLINE typename ACE_String_Base<CHAR>::size_type
+ACE_String_Base<CHAR>::find (const ACE_String_Base<CHAR>&str,
+                             size_type pos) const
 {
   ACE_TRACE ("ACE_String_Base<CHAR>::find");
   return this->find (str.rep_, pos);
 }
 
-template <class CHAR> ACE_INLINE ssize_t
+template <class CHAR> ACE_INLINE typename ACE_String_Base<CHAR>::size_type
 ACE_String_Base<CHAR>::strstr (const ACE_String_Base<CHAR> &s) const
 {
   ACE_TRACE ("ACE_String_Base<CHAR>::strstr");
Index: String_Base.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/String_Base.cpp,v
retrieving revision 4.22
retrieving revision 4.23
diff -u -r4.22 -r4.23
--- String_Base.cpp	4 Jun 2006 00:50:30 -0000	4.22
+++ String_Base.cpp	14 Jul 2006 21:27:23 -0000	4.23
@@ -1,4 +1,4 @@
-// $Id: String_Base.cpp,v 4.22 2006/06/04 00:50:30 schmidt Exp $
+// $Id: String_Base.cpp,v 4.23 2006/07/14 21:27:23 ossama Exp $
 
 #ifndef ACE_STRING_BASE_CPP
 #define ACE_STRING_BASE_CPP
@@ -9,6 +9,8 @@
 #include "ace/Auto_Ptr.h"
 #include "ace/OS_NS_string.h"
 
+#include <algorithm>  // For std::swap<>
+
 #if !defined (__ACE_INLINE__)
 #include "ace/String_Base.inl"
 #endif /* __ACE_INLINE__ */
@@ -18,7 +20,7 @@
 ACE_ALLOC_HOOK_DEFINE(ACE_String_Base)
 
 template <class CHAR>
-  CHAR ACE_String_Base<CHAR>::NULL_String_ = 0;
+CHAR ACE_String_Base<CHAR>::NULL_String_ = 0;
 
 // Default constructor.
 
@@ -67,7 +69,7 @@
 
 template <class CHAR>
 ACE_String_Base<CHAR>::ACE_String_Base (const CHAR *s,
-                                        size_t len,
+                                        size_type len,
                                         ACE_Allocator *the_allocator,
                                         int release)
   : allocator_ (the_allocator ? the_allocator : ACE_Allocator::instance ()),
@@ -97,7 +99,7 @@
 }
 
 template <class CHAR>
-ACE_String_Base<CHAR>::ACE_String_Base (size_t len, CHAR c, ACE_Allocator *the_allocator)
+ACE_String_Base<CHAR>::ACE_String_Base (size_type len, CHAR c, ACE_Allocator *the_allocator)
   : allocator_ (the_allocator ? the_allocator : ACE_Allocator::instance ()),
     len_ (0),
     buf_len_ (0),
@@ -120,10 +122,10 @@
 
 // this method might benefit from a little restructuring.
 template <class CHAR> void
-ACE_String_Base<CHAR>::set (const CHAR *s, size_t len, int release)
+ACE_String_Base<CHAR>::set (const CHAR *s, size_type len, int release)
 {
   // Case 1. Going from memory to more memory
-  size_t new_buf_len = len + 1;
+  size_type new_buf_len = len + 1;
   if (s != 0 && len != 0 && release && this->buf_len_ < new_buf_len)
     {
       CHAR *temp;
@@ -177,10 +179,10 @@
 
 // Return substring.
 template <class CHAR> ACE_String_Base<CHAR>
-ACE_String_Base<CHAR>::substring (size_t offset, ssize_t length) const
+ACE_String_Base<CHAR>::substring (size_type offset, size_type length) const
 {
   ACE_String_Base<CHAR> nill;
-  size_t count = length;
+  size_type count = length;
 
   // case 1. empty string
   if (this->len_ == 0)
@@ -193,17 +195,17 @@
   else if (length == 0)
     return nill;
   // Get all remaining bytes.
-  else if (length == -1 || count > (this->len_ - offset))
+  else if (length == npos || count > (this->len_ - offset))
     count = this->len_ - offset;
 
   return ACE_String_Base<CHAR> (&this->rep_[offset], count, this->allocator_);
 }
 
 template <class CHAR> ACE_String_Base<CHAR> &
-ACE_String_Base<CHAR>::append (const CHAR* s, size_t slen)
+ACE_String_Base<CHAR>::append (const CHAR* s, size_type slen)
 {
-  ACE_TRACE ("ACE_String_Base<CHAR>::append(const CHAR*, size_t)");
-  if (slen > 0)
+  ACE_TRACE ("ACE_String_Base<CHAR>::append(const CHAR*, size_type)");
+  if (slen > 0 && slen != npos)
   {
     // case 1. No memory allocation needed.
     if (this->buf_len_ >= this->len_ + slen + 1)
@@ -213,7 +215,7 @@
     }
     else // case 2. Memory reallocation is needed
     {
-      const size_t new_buf_len =
+      const size_type new_buf_len =
         ace_max(this->len_ + slen + 1, this->buf_len_ + this->buf_len_ / 2);
 
       CHAR *t = 0;
@@ -251,7 +253,7 @@
 }
 
 template <class CHAR> void
-ACE_String_Base<CHAR>::resize (size_t len, CHAR c)
+ACE_String_Base<CHAR>::resize (size_type len, CHAR c)
 {
   ACE_TRACE ("ACE_String_Base<CHAR>::resize");
 
@@ -306,7 +308,7 @@
 {
   ACE_TRACE ("ACE_String_Base<CHAR>::operator=");
 
-  // Check for identify.
+  // Check for self-assignment.
   if (this != &s)
     {
       this->set (s.rep_, s.len_, 1);
@@ -366,7 +368,7 @@
     return 0;
 
   // Pick smaller of the two lengths and perform the comparison.
-  size_t smaller_length = ace_min (this->len_, s.len_);
+  size_type smaller_length = ace_min (this->len_, s.len_);
 
   int result = ACE_OS::memcmp (this->rep_,
                                s.rep_,
@@ -398,8 +400,8 @@
                          len * sizeof (CHAR)) == 0;
 }
 
-template <class CHAR> ssize_t
-ACE_String_Base<CHAR>::find (const CHAR *s, size_t pos) const
+template <class CHAR> typename ACE_String_Base<CHAR>::size_type
+ACE_String_Base<CHAR>::find (const CHAR *s, size_type pos) const
 {
   CHAR *substr = this->rep_ + pos;
   size_t len = ACE_OS::strlen (s);
@@ -410,8 +412,8 @@
     return pointer - this->rep_;
 }
 
-template <class CHAR> ssize_t
-ACE_String_Base<CHAR>::find (CHAR c, size_t pos) const
+template <class CHAR> typename ACE_String_Base<CHAR>::size_type
+ACE_String_Base<CHAR>::find (CHAR c, size_type pos) const
 {
   CHAR *substr = this->rep_ + pos;
   CHAR *pointer = ACE_OS::strnchr (substr, c, this->len_ - pos);
@@ -421,23 +423,38 @@
     return pointer - this->rep_;
 }
 
-template <class CHAR> ssize_t
-ACE_String_Base<CHAR>::rfind (CHAR c, ssize_t pos) const
-{
-  if (pos == npos || pos > static_cast<ssize_t> (this->len_))
-    pos = static_cast<ssize_t> (this->len_);
-
-  for (ssize_t i = pos - 1; i >= 0; i--)
+template <class CHAR> typename ACE_String_Base<CHAR>::size_type
+ACE_String_Base<CHAR>::rfind (CHAR c,
+                              size_type pos) const
+{
+  if (pos == npos || pos > this->len_)
+    pos = this->len_;
+
+  // Do not change to prefix operator!  Proper operation of this loop
+  // depends on postfix decrement behavior.
+  for (size_type i = pos; i-- != 0; )
     if (this->rep_[i] == c)
       return i;
 
   return ACE_String_Base<CHAR>::npos;
 }
 
+template <class CHAR> void
+ACE_String_Base<CHAR>::swap (ACE_String_Base<CHAR> & str)
+{
+  std::swap (this->allocator_ , str.allocator_);
+  std::swap (this->len_       , str.len_);
+  std::swap (this->buf_len_   , str.buf_len_);
+  std::swap (this->rep_       , str.rep_);
+  std::swap (this->release_   , str.release_);
+}
+
+// ----------------------------------------------
+
 template <class CHAR> ACE_String_Base<CHAR>
 operator+ (const ACE_String_Base<CHAR> &s, const ACE_String_Base<CHAR> &t)
 {
-  ACE_String_Base<CHAR> temp (s.length() + t.length());
+  ACE_String_Base<CHAR> temp (s.length () + t.length ());
   temp += s;
   temp += t;
   return temp;
@@ -449,9 +466,9 @@
   size_t slen = 0;
   if (s != 0)
     slen = ACE_OS::strlen (s);
-  ACE_String_Base<CHAR> temp (slen + t.length());
+  ACE_String_Base<CHAR> temp (slen + t.length ());
   if (slen > 0)
-    temp.append(s, slen);
+    temp.append (s, slen);
   temp += t;
   return temp;
 }
@@ -462,34 +479,35 @@
   size_t tlen = 0;
   if (t != 0)
     tlen = ACE_OS::strlen (t);
-  ACE_String_Base<CHAR> temp (s.length() + tlen);
+  ACE_String_Base<CHAR> temp (s.length () + tlen);
   temp += s;
   if (tlen > 0)
-    temp.append(t, tlen);
+    temp.append (t, tlen);
   return temp;
 }
 
-template <class CHAR>
-ACE_String_Base<CHAR> operator + (const ACE_String_Base<CHAR> &t,
-                                  const CHAR c)
+template <class CHAR> ACE_String_Base<CHAR>
+operator + (const ACE_String_Base<CHAR> &t,
+            const CHAR c)
 {
-  ACE_String_Base<CHAR> temp (t.length() + 1);
+  ACE_String_Base<CHAR> temp (t.length () + 1);
   temp += t;
   temp += c;
   return temp;
 }
 
-template <class CHAR>
-ACE_String_Base<CHAR> operator + (const CHAR c,
-                                  const ACE_String_Base<CHAR> &t)
+template <class CHAR> ACE_String_Base<CHAR>
+operator + (const CHAR c,
+            const ACE_String_Base<CHAR> &t)
 {
-  ACE_String_Base<CHAR> temp (t.length() + 1);
+  ACE_String_Base<CHAR> temp (t.length () + 1);
   temp += c;
   temp += t;
   return temp;
 }
 
-template <class CHAR> ACE_String_Base<CHAR> &
+template <class CHAR>
+ACE_String_Base<CHAR> &
 ACE_String_Base<CHAR>::operator+= (const CHAR* s)
 {
   size_t slen = 0;
@@ -498,19 +516,21 @@
   return this->append (s, slen);
 }
 
-template <class CHAR> ACE_String_Base<CHAR> &
+template <class CHAR>
+ACE_String_Base<CHAR> &
 ACE_String_Base<CHAR>::operator+= (const ACE_String_Base<CHAR> &s)
 {
   ACE_TRACE ("ACE_String_Base<CHAR>::operator+=(const ACE_String_Base<CHAR> &)");
-  return this->append(s.rep_, s.len_);
+  return this->append (s.rep_, s.len_);
 }
 
-template <class CHAR> ACE_String_Base<CHAR> &
+template <class CHAR>
+ACE_String_Base<CHAR> &
 ACE_String_Base<CHAR>::operator+= (const CHAR c)
 {
   ACE_TRACE ("ACE_String_Base<CHAR>::operator+=(const CHAR)");
-  const size_t slen = 1;
-  return this->append(&c, slen);
+  const size_type slen = 1;
+  return this->append (&c, slen);
 }
 
 ACE_END_VERSIONED_NAMESPACE_DECL
Index: SString.h
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/SString.h,v
retrieving revision 4.86
retrieving revision 4.87
diff -u -r4.86 -r4.87
--- SString.h	25 Jan 2006 19:56:26 -0000	4.86
+++ SString.h	14 Jul 2006 21:27:23 -0000	4.87
@@ -4,7 +4,7 @@
 /**
  *  @file    SString.h
  *
- *  $Id: SString.h,v 4.86 2006/01/25 19:56:26 jwillemsen Exp $
+ *  $Id: SString.h,v 4.87 2006/07/14 21:27:23 ossama Exp $
  *
  *  @author Douglas C. Schmidt (schmidt@cs.wustl.edu)
  */
@@ -52,6 +52,9 @@
 class ACE_Export ACE_NS_WString : public ACE_WString
 {
 public:
+
+  using ACE_WString::size_type;
+
   /// Default constructor.
   ACE_NS_WString (ACE_Allocator *alloc = 0);
 
@@ -67,19 +70,19 @@
   /// Constructor that takes in a ushort16 string (mainly used by the
   /// ACE Name_Space classes)
   ACE_NS_WString (const ACE_USHORT16 *s,
-                  size_t len,
+                  size_type len,
                   ACE_Allocator *alloc = 0);
 #endif /* ACE_WSTRING_HAS_USHORT_SUPPORT */
 
   /// Constructor that copies @a len ACE_WSTRING_TYPE's of @a s into dynamically
   /// allocated memory (will NUL terminate the result).
   ACE_NS_WString (const ACE_WSTRING_TYPE *s,
-                  size_t len,
+                  size_type len,
                   ACE_Allocator *alloc = 0);
 
   /// Constructor that dynamically allocates memory for @a len + 1
   /// ACE_WSTRING_TYPE characters. The newly created memory is set memset to 0.
-  ACE_NS_WString (size_t len, ACE_Allocator *alloc = 0);
+  ACE_NS_WString (size_type len, ACE_Allocator *alloc = 0);
 
   /// Copy constructor.
   ACE_NS_WString (const ACE_NS_WString &s);
@@ -124,8 +127,11 @@
 class ACE_Export ACE_SString
 {
 public:
+
+  typedef ACE_Allocator::size_type size_type;
+
   /// No position constant
-  static const int npos;
+  static const size_type npos;
 
   /// Default constructor.
   ACE_SString (ACE_Allocator *alloc = 0);
@@ -135,7 +141,7 @@
 
   /// Constructor that copies @a len chars of  @s  into dynamically
   /// allocated memory (will NUL terminate the result).
-  ACE_SString (const char *s, size_t len, ACE_Allocator *alloc = 0);
+  ACE_SString (const char *s, size_type len, ACE_Allocator *alloc = 0);
 
   /// Copy constructor.
   ACE_SString (const ACE_SString &);
@@ -148,30 +154,30 @@
 
   /// Return the <slot'th> character in the string (doesn't perform
   /// bounds checking).
-  char operator [] (size_t slot) const;
+  char operator [] (size_type slot) const;
 
   /// Return the <slot'th> character by reference in the string
   /// (doesn't perform bounds checking).
-  char &operator [] (size_t slot);
+  char &operator [] (size_type slot);
 
   /// Assignment operator (does copy memory).
   ACE_SString &operator = (const ACE_SString &);
 
   /**
-   * Return a substring given an offset and length, if length == -1
+   * Return a substring given an offset and length, if length == npos
    * use rest of str return empty substring if offset or offset/length
    * are invalid
    */
-  ACE_SString substring (size_t offset, ssize_t length = -1) const;
+  ACE_SString substring (size_type offset, size_type length = npos) const;
 
   /// Same as substring
-  ACE_SString substr (size_t offset, ssize_t length = -1) const;
+  ACE_SString substr (size_type offset, size_type length = npos) const;
 
   /// Returns a hash value for this string.
   u_long hash (void) const;
 
   /// Return the length of the string.
-  size_t length (void) const;
+  size_type length (void) const;
 
   /// Set the underlying pointer.  Since this does not copy memory or
   /// delete existing memory use with extreme caution!!!
@@ -187,24 +193,24 @@
   const char *c_str (void) const;
 
   /// Comparison operator that will match substrings.  Returns the
-  /// slot of the first location that matches, else -1.
-  int strstr (const ACE_SString &s) const;
+  /// slot of the first location that matches, else @c npos.
+  size_type strstr (const ACE_SString &s) const;
 
   /// Find <str> starting at pos.  Returns the slot of the first
   /// location that matches (will be >= pos), else npos.
-  int find (const ACE_SString &str, int pos = 0) const;
+  size_type find (const ACE_SString &str, size_type pos = 0) const;
 
   /// Find <s> starting at pos.  Returns the slot of the first
   /// location that matches (will be >= pos), else npos.
-  int find (const char *s, int pos = 0) const;
+  size_type find (const char *s, size_type pos = 0) const;
 
   /// Find <c> starting at pos.  Returns the slot of the first
   /// location that matches (will be >= pos), else npos.
-  int find (char c, int pos = 0) const;
+  size_type find (char c, size_type pos = 0) const;
 
   /// Find <c> starting at pos (counting from the end).  Returns the
   /// slot of the first location that matches, else npos.
-  int rfind (char c, int pos = npos) const;
+  size_type rfind (char c, size_type pos = npos) const;
 
   /// Equality comparison operator (must match entire string).
   bool operator == (const ACE_SString &s) const;
@@ -232,7 +238,7 @@
   ACE_Allocator *allocator_;
 
   /// Length of the ACE_SString (not counting the trailing '\0').
-  size_t len_;
+  size_type len_;
 
   /// Pointer to data.
   char *rep_;
@@ -474,7 +480,7 @@
   ~ACE_Auto_String_Free (void);
 
   char* operator* () const;
-  char operator[] (int i) const;
+  char operator[] (size_t i) const;
   char* get (void) const;
   char* release (void);
   void reset (char* p = 0);
Index: SString.inl
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/SString.inl,v
retrieving revision 4.3
retrieving revision 4.4
diff -u -r4.3 -r4.4
--- SString.inl	22 Dec 2005 01:17:51 -0000	4.3
+++ SString.inl	14 Jul 2006 21:27:23 -0000	4.4
@@ -1,6 +1,6 @@
 // -*- C++ -*-
 //
-// $Id: SString.inl,v 4.3 2005/12/22 01:17:51 shuston Exp $
+// $Id: SString.inl,v 4.4 2006/07/14 21:27:23 ossama Exp $
 
 // Include ACE.h only if it hasn't already been included, e.g., if
 // ACE_TEMPLATES_REQUIRE_SOURCE, ACE.h won't have been pulled in by
@@ -21,7 +21,7 @@
 
 ACE_INLINE
 ACE_NS_WString::ACE_NS_WString (const ACE_WSTRING_TYPE *s,
-                                size_t len,
+                                size_type len,
                                 ACE_Allocator *alloc)
   : ACE_WString (s, len, alloc)
 {
@@ -35,7 +35,7 @@
 }
 
 ACE_INLINE
-ACE_NS_WString::ACE_NS_WString (size_t len, ACE_Allocator *alloc)
+ACE_NS_WString::ACE_NS_WString (size_type len, ACE_Allocator *alloc)
   : ACE_WString (len, 0, alloc)
 {
 }
@@ -68,8 +68,8 @@
 }
 
 ACE_INLINE ACE_SString
-ACE_SString::substr (size_t offset,
-                     ssize_t length) const
+ACE_SString::substr (size_type offset,
+                     size_type length) const
 {
   return this->substring (offset, length);
 }
@@ -77,7 +77,7 @@
 // Return the <slot'th> character in the string.
 
 ACE_INLINE char
-ACE_SString::operator[] (size_t slot) const
+ACE_SString::operator[] (size_type slot) const
 {
   ACE_TRACE ("ACE_SString::operator[]");
   return this->rep_[slot];
@@ -86,7 +86,7 @@
 // Return the <slot'th> character in the string by reference.
 
 ACE_INLINE char &
-ACE_SString::operator[] (size_t slot)
+ACE_SString::operator[] (size_type slot)
 {
   ACE_TRACE ("ACE_SString::operator[]");
   return this->rep_[slot];
@@ -169,8 +169,8 @@
   return ACE_OS::strcmp (this->rep_, s.rep_);
 }
 
-ACE_INLINE int
-ACE_SString::find (const char *s, int pos) const
+ACE_INLINE ACE_SString::size_type
+ACE_SString::find (const char *s, size_type pos) const
 {
   char *substr = this->rep_ + pos;
   char *pointer = ACE_OS::strstr (substr, s);
@@ -180,8 +180,8 @@
     return pointer - this->rep_;
 }
 
-ACE_INLINE int
-ACE_SString::find (char c, int pos) const
+ACE_INLINE ACE_SString::size_type
+ACE_SString::find (char c, size_type pos) const
 {
   char *substr = this->rep_ + pos;
   char *pointer = ACE_OS::strchr (substr, c);
@@ -191,7 +191,7 @@
     return pointer - this->rep_;
 }
 
-ACE_INLINE int
+ACE_INLINE ACE_SString::size_type
 ACE_SString::strstr (const ACE_SString &s) const
 {
   ACE_TRACE ("ACE_SString::strstr");
@@ -199,19 +199,21 @@
   return this->find (s.rep_);
 }
 
-ACE_INLINE int
-ACE_SString::find (const ACE_SString &str, int pos) const
+ACE_INLINE ACE_SString::size_type
+ACE_SString::find (const ACE_SString &str, size_type pos) const
 {
   return this->find (str.rep_, pos);
 }
 
-ACE_INLINE int
-ACE_SString::rfind (char c, int pos) const
+ACE_INLINE ACE_SString::size_type
+ACE_SString::rfind (char c, size_type pos) const
 {
   if (pos == ACE_SString::npos)
-    pos = static_cast<int> (this->len_);
+    pos = this->len_;
 
-  for (int i = pos - 1; i >= 0; i--)
+  // Do not change to prefix operator!  Proper operation of this loop
+  // depends on postfix decrement behavior.
+  for (size_type i = pos; i-- != 0; )
     if (this->rep_[i] == c)
       return i;
 
@@ -224,7 +226,7 @@
   return ACE::hash_pjw (this->rep_);
 }
 
-ACE_INLINE size_t
+ACE_INLINE ACE_SString::size_type
 ACE_SString::length (void) const
 {
   ACE_TRACE ("ACE_SString::length");
@@ -247,8 +249,7 @@
 ACE_INLINE void
 ACE_Auto_String_Free::reset (char* p)
 {
-  if (this->p_ != 0)
-    ACE_OS::free (this->p_);
+  ACE_OS::free (this->p_);
   this->p_ = p;
 }
 
@@ -276,7 +277,7 @@
 }
 
 ACE_INLINE char
-ACE_Auto_String_Free::operator[] (int i) const
+ACE_Auto_String_Free::operator[] (size_t i) const
 {
   return this->p_[i];
 }
Index: SString.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/ace/SString.cpp,v
retrieving revision 4.94
retrieving revision 4.95
diff -u -r4.94 -r4.95
--- SString.cpp	19 Apr 2006 19:13:09 -0000	4.94
+++ SString.cpp	14 Jul 2006 21:27:23 -0000	4.95
@@ -1,4 +1,4 @@
-// $Id: SString.cpp,v 4.94 2006/04/19 19:13:09 jwillemsen Exp $
+// $Id: SString.cpp,v 4.95 2006/07/14 21:27:23 ossama Exp $
 
 #include "ace/Malloc_T.h"
 #include "ace/OS_Memory.h"
@@ -18,6 +18,10 @@
 #include "ace/SString.inl"
 #endif /* __ACE_INLINE__ */
 
+#undef max
+#include <limits>
+
+
 ACE_RCSID (ace,
            SString,
            "SString.cpp,v 4.61 2001/03/04 00:55:30 brunsch Exp")
@@ -65,7 +69,7 @@
 ACE_NS_WString::char_rep (void) const
 {
   ACE_TRACE ("ACE_NS_WString::char_rep");
-  if (this->len_ <= 0)
+  if (this->len_ == 0)
     return 0;
   else
     {
@@ -75,7 +79,7 @@
                       char[this->len_ + 1],
                       0);
 
-      for (size_t i = 0; i < this->len_; i++)
+      for (size_type i = 0; i < this->len_; ++i)
         // Note that this cast may lose data if wide chars are
         // actually used!
         t[i] = char (this->rep_[i]);
@@ -99,7 +103,7 @@
                       ACE_USHORT16[this->len_ + 1],
                       0);
 
-      for (size_t i = 0; i < this->len_; i++)
+      for (size_type i = 0; i < this->len_; ++i)
         // Note that this cast may lose data if wide chars are
         // actually used!
         t[i] = (ACE_USHORT16)this->rep_[i];
@@ -126,13 +130,13 @@
                  this->allocator_->malloc ((this->buf_len_ + 1) *
                                            sizeof (ACE_WSTRING_TYPE)));
   this->release_ = 1;
-  for (size_t i = 0; i <= this->buf_len_; i++)
+  for (size_type i = 0; i <= this->buf_len_; ++i)
     this->rep_[i] = s[i];
 }
 
 #if defined (ACE_WSTRING_HAS_USHORT_SUPPORT)
 ACE_NS_WString::ACE_NS_WString (const ACE_USHORT16 *s,
-                                size_t len,
+                                size_type len,
                                 ACE_Allocator *alloc)
   : ACE_WString (alloc)
 {
@@ -149,14 +153,15 @@
                  this->allocator_->malloc ((this->buf_len_) *
                                            sizeof (ACE_WSTRING_TYPE)));
   this->release_ = 1;
-  for (size_t i = 0; i < this->buf_len_; i++)
+  for (size_type i = 0; i < this->buf_len_; ++i)
     this->rep_[i] = s[i];
 }
 #endif /* ACE_WSTRING_HAS_USHORT_SUPPORT */
 
 // *****************************************************************
 
-const int ACE_SString::npos = -1;
+ACE_SString::size_type const ACE_SString::npos =
+  std::numeric_limits<ACE_SString::size_type>::max ();
 
 ACE_ALLOC_HOOK_DEFINE(ACE_SString)
 
@@ -262,7 +267,7 @@
 // Constructor that actually copies memory.
 
 ACE_SString::ACE_SString (const char *s,
-                          size_t len,
+                          size_type len,
                           ACE_Allocator *alloc)
   : allocator_ (alloc)
 {
@@ -311,8 +316,8 @@
 
 // Return substring.
 ACE_SString
-ACE_SString::substring (size_t offset,
-                        ssize_t length) const
+ACE_SString::substring (size_type offset,
+                        size_type length) const
 {
   ACE_SString nill;
   size_t count = length;
@@ -326,7 +331,7 @@
     return nill;
 
   // get all remaining bytes
-  if (length == -1 || count > (this->len_ - offset))
+  if (length == npos || count > (this->len_ - offset))
     count = len_ - offset;
 
   return ACE_SString (&rep_[offset], count, this->allocator_);
