Index: IFR_Service/IFR_Service.h
===================================================================
--- IFR_Service/IFR_Service.h	(revision 80118)
+++ IFR_Service/IFR_Service.h	(working copy)
@@ -50,6 +50,9 @@
   /// Run the IFR service.
   int run (void);
 
+  /// Shutdown the Service.
+  void shutdown (void);
+
 protected:
 
   /// Reference to our ORB.
Index: IFR_Service/IFR_Server.cpp
===================================================================
--- IFR_Service/IFR_Server.cpp	(revision 80118)
+++ IFR_Service/IFR_Server.cpp	(working copy)
@@ -3,16 +3,44 @@
 
 #include "IFR_Service.h"
 #include "tao/Environment.h"
+#include "orbsvcs/Shutdown_Utilities.h"
 
 ACE_RCSID (IFR_Service,
            IFR_Server,
            "$Id$")
 
+class IFR_Service_Shutdown_Functor : public Shutdown_Functor
+{
+public:
+  IFR_Service_Shutdown_Functor (IFR_Service& ifr);
+
+  void operator() (int which_signal);
+private:
+  IFR_Service& ifr_;
+};
+
+IFR_Service_Shutdown_Functor::IFR_Service_Shutdown_Functor (IFR_Service &ifr)
+  : ifr_(ifr)
+{
+}
+
+void
+IFR_Service_Shutdown_Functor::operator() (int which_signal)
+{
+  if (TAO_debug_level > 0)
+    ACE_DEBUG ((LM_DEBUG,
+                "shutting down on signal %d\n", which_signal));
+  (void) this->ifr_.shutdown ();
+}
+
 int
 main (int argc, char *argv[])
 {
   IFR_Service server;
 
+  IFR_Service_Shutdown_Functor killer (server);
+  Service_Shutdown kill_contractor (killer);
+
   try
     {
       int status = server.init (argc, argv);
Index: IFR_Service/IFR_Service.cpp
===================================================================
--- IFR_Service/IFR_Service.cpp	(revision 80118)
+++ IFR_Service/IFR_Service.cpp	(working copy)
@@ -75,3 +75,8 @@
   return 0;
 }
 
+void
+IFR_Service::shutdown (void)
+{
+  this->orb_->shutdown ();
+}
