diff -r -u TAO/orbsvcs/IFR_Service.orig/ifr_adding_visitor.cpp TAO/orbsvcs/IFR_Service/ifr_adding_visitor.cpp
--- TAO/orbsvcs/IFR_Service.orig/ifr_adding_visitor.cpp	2008-07-22 10:37:43.458310500 -0400
+++ TAO/orbsvcs/IFR_Service/ifr_adding_visitor.cpp	2008-07-22 11:04:35.697696900 -0400
@@ -2449,6 +2449,8 @@
       return CORBA::pk_void;
     case AST_PredefinedType::PT_object:
       return CORBA::pk_objref;
+    case AST_PredefinedType::PT_value:
+      return CORBA::pk_value_base;
     case AST_PredefinedType::PT_pseudo:
       {
         const char *local_name = node->local_name ()->get_string ();
