--- TTY_IO.cpp	2008-03-04 17:51:23.000000000 +0300
+++ ../../TTY_IO.cpp	2008-09-10 17:31:56.718750000 +0400
@@ -400,17 +400,24 @@
       return -1; // Wrong cmd.
     }
 #elif defined (ACE_WIN32)
+  DCB dcb;
+  dcb.DCBlength = sizeof dcb;
+  if (!::GetCommState (this->get_handle (), &dcb))
+  {
+    ACE_OS::set_errno_to_last_error ();
+    return -1;
+  }
+
+  COMMTIMEOUTS timeouts;
+  if (!::GetCommTimeouts (this->get_handle(), &timeouts))
+  {
+    ACE_OS::set_errno_to_last_error ();
+    return -1;
+  }
+
   switch (cmd)
     {
     case SETPARAMS:
-      DCB dcb;
-      dcb.DCBlength = sizeof dcb;
-      if (!::GetCommState (this->get_handle (), &dcb))
-        {
-          ACE_OS::set_errno_to_last_error ();
-          return -1;
-        }
-
       dcb.BaudRate = arg->baudrate;
 
       switch (arg->databits)
@@ -523,13 +530,6 @@
           return -1;
         }
 
-      COMMTIMEOUTS timeouts;
-      if (!::GetCommTimeouts (this->get_handle(), &timeouts))
-        {
-          ACE_OS::set_errno_to_last_error ();
-          return -1;
-        }
-
       if (arg->readtimeoutmsec < 0)
         {
           // Settings for infinite timeout.
@@ -561,7 +561,133 @@
        return 0;
 
     case GETPARAMS:
-      ACE_NOTSUP_RETURN (-1); // Not yet implemented.
+      arg->baudrate = dcb.BaudRate;
+
+      switch (dcb.ByteSize)
+        {
+        case 4:
+        case 5:
+        case 6:
+        case 7:
+        case 8:
+          arg->databits = dcb.ByteSize;
+          break;
+        default:
+          return -1;
+        }
+
+      // StopBits==1 doesn't supported by ACE.
+      // So if StopBits if equal to 1 we will return -1.
+      switch (dcb.StopBits)
+        {
+        case 0:
+          arg->stopbits = 1;
+          break;
+        case 2:
+          arg->stopbits = 2;
+          break;
+        default:
+          return -1;
+        }
+
+      if (!dcb.fParity)
+        {
+          arg->parityenb = false;
+          arg->paritymode = ACE_TTY_IO_NONE;
+        }
+      else
+        {
+          arg->parityenb = true;
+          switch (dcb.Parity)
+            {
+            case ODDPARITY:
+              arg->paritymode = ACE_TTY_IO_ODD;
+              break;
+            case EVENPARITY:
+              arg->paritymode = ACE_TTY_IO_EVEN;
+              break;
+            case NOPARITY:
+              arg->paritymode = ACE_TTY_IO_NONE;
+              break;
+            case MARKPARITY:
+              arg->paritymode = ACE_TTY_IO_MARK;
+              break;
+            case SPACEPARITY:
+              arg->paritymode = ACE_TTY_IO_SPACE;
+              break;
+            default:
+              return -1;
+            }
+        }
+
+      // Enable/disable RTS protocol.
+      switch (dcb.fRtsControl)
+        {
+        case RTS_CONTROL_ENABLE:
+          arg->rtsenb = 1;
+          break;
+        case RTS_CONTROL_HANDSHAKE:
+          arg->rtsenb = 2;
+          break;
+        case RTS_CONTROL_TOGGLE:
+          arg->rtsenb = 3;
+          break;
+        case RTS_CONTROL_DISABLE:
+          arg->rtsenb = 0;
+        default:
+          return -1;
+        }
+
+      // Enable/disable CTS protocol.
+      if (dcb.fOutxCtsFlow)
+        arg->ctsenb = true;
+      else
+        arg->ctsenb = false;
+
+      // Enable/disable DSR protocol.
+      if (dcb.fOutxDsrFlow)
+        arg->dsrenb = true;
+      else
+        arg->dsrenb = false;
+
+      // Disable/enable DTR protocol
+      // Attention: DTR_CONTROL_HANDSHAKE doesn't supported.
+      switch (dcb.fDtrControl)
+        {
+        case DTR_CONTROL_DISABLE:
+          arg->dtrdisable = true;
+          break;
+        case DTR_CONTROL_ENABLE:
+          arg->dtrdisable = false;
+          break;
+        default:
+          return -1;
+        }
+
+      // Enable/disable software flow control on input
+      if (dcb.fInX)
+        arg->xinenb = true;
+      else
+        arg->xinenb = false;
+
+      // Enable/disable software flow control on output
+      if (dcb.fOutX)
+        arg->xoutenb = true;
+      else
+        arg->xoutenb = false;
+
+      arg->xonlim = static_cast<int>(dcb.XonLim);
+      arg->xofflim = static_cast<int>(dcb.XoffLim);
+
+      if (timeouts.ReadIntervalTimeout == 0 &&
+        timeouts.ReadTotalTimeoutMultiplier == 0 &&
+        timeouts.ReadTotalTimeoutConstant == 0)
+          arg->readtimeoutmsec = -1;
+      else
+        arg->readtimeoutmsec = timeouts.ReadTotalTimeoutConstant;
+
+      return 0;
+
     default:
       return -1; // Wrong cmd.
 
