diff -up ACE_wrappers/ace/Select_Reactor_Base.cpp.orig ACE_wrappers/ace/Select_Reactor_Base.cpp
--- ACE_wrappers/ace/Select_Reactor_Base.cpp.orig	2008-04-02 13:29:29.000000000 -0400
+++ ACE_wrappers/ace/Select_Reactor_Base.cpp	2008-04-29 16:27:30.000000000 -0400
@@ -82,6 +82,14 @@ ACE_Select_Reactor_Handler_Repository::h
   return false;
 }
 
+bool
+ACE_Select_Reactor_Handler_Repository::handle_is_bound (ACE_HANDLE handle)
+{
+  ACE_TRACE ("ACE_Select_Reactor_Handler_Repository::handle_is_bound");
+
+  return find(handle) != 0 ? true : false;
+}
+
 int
 ACE_Select_Reactor_Handler_Repository::open (size_t size)
 {
@@ -961,7 +969,7 @@ ACE_Select_Reactor_Impl::bit_ops (ACE_HA
                                   int ops)
 {
   ACE_TRACE ("ACE_Select_Reactor_Impl::bit_ops");
-  if (this->handler_rep_.handle_in_range (handle) == 0)
+  if (this->handler_rep_.handle_is_bound (handle) == 0)
     return -1;
 
 #if !defined (ACE_WIN32)
diff -up ACE_wrappers/ace/Select_Reactor_Base.inl.orig ACE_wrappers/ace/Select_Reactor_Base.inl
--- ACE_wrappers/ace/Select_Reactor_Base.inl.orig	2008-04-29 16:24:51.000000000 -0400
+++ ACE_wrappers/ace/Select_Reactor_Base.inl	2008-04-29 16:26:38.000000000 -0400
@@ -34,7 +34,7 @@ ACE_Select_Reactor_Handler_Repository::u
   // To resolve bug 2653, unbind must be called even when find_eh returns
   // event_handlers_.end().
 
-  return !this->handle_in_range (handle) ? -1
+  return !this->invalid_handle (handle) ? -1
           : this->unbind (handle,
                           this->find_eh (handle),
                           mask);
@@ -47,7 +47,7 @@ ACE_Select_Reactor_Handler_Repository::f
 
   ACE_Event_Handler * eh = 0;
 
-  if (this->handle_in_range (handle))
+  if (!this->invalid_handle (handle))
     {
       map_type::iterator const pos = this->find_eh (handle);
 
diff -up ACE_wrappers/ace/Select_Reactor_Base.h.orig ACE_wrappers/ace/Select_Reactor_Base.h
--- ACE_wrappers/ace/Select_Reactor_Base.h.orig	2008-04-29 16:19:01.000000000 -0400
+++ ACE_wrappers/ace/Select_Reactor_Base.h	2008-04-29 16:20:14.000000000 -0400
@@ -363,6 +363,9 @@ public:
   // @c max_handlep1_).
   bool handle_in_range (ACE_HANDLE handle);
 
+  // Check the @c handle to make sure it's a bound @c ACE_HANDLE
+  bool handle_is_bound (ACE_HANDLE handle);
+
   // = Accessors.
   /// Returns the current table size.
   size_type size (void) const;
