--- Malloc_T.cpp	2003-12-19 10:29:11.000000000 -0500
+++ /cygdrive/c/welchj/c/apps/ACE_wrappers/ace/Malloc_T.cpp	2004-02-17 12:44:56.503093500 -0500
@@ -1,5 +1,5 @@
 // Malloc_T.cpp
-// Malloc_T.cpp,v 4.95 2003/12/19 15:29:11 schmidt Exp
+// Malloc_T.cpp,v 4.93 2003/11/01 11:15:13 dhinton Exp
 
 #ifndef ACE_MALLOC_T_C
 #define ACE_MALLOC_T_C
@@ -17,7 +17,7 @@
 #include "ace/ACE.h"
 #include "ace/OS_NS_string.h"
 
-ACE_RCSID(ace, Malloc_T, "Malloc_T.cpp,v 4.95 2003/12/19 15:29:11 schmidt Exp")
+ACE_RCSID(ace, Malloc_T, "Malloc_T.cpp,v 4.93 2003/11/01 11:15:13 dhinton Exp")
 
 template <class T, class ACE_LOCK>
 ACE_Cached_Allocator<T, ACE_LOCK>::ACE_Cached_Allocator (size_t n_chunks)
@@ -289,15 +289,14 @@
     bad_flag_ (0)
 {
   ACE_TRACE ("ACE_Malloc_T<ACE_MEM_POOL_2, ACE_LOCK, ACE_CB>::ACE_Malloc_T");
-  this->lock_ = ACE_Malloc_Lock_Adapter_T<ACE_LOCK> ()(pool_name);
-  if (this->lock_ == 0)
-    return;
-
+  if (pool_name == 0)
+    ACE_NEW (this->lock_, ACE_LOCK (pool_name));  // Want the ctor with char*
+  else
+    ACE_NEW (this->lock_, ACE_LOCK (ACE::basename (pool_name,
+                                                   ACE_DIRECTORY_SEPARATOR_CHAR)));
   this->delete_lock_ = 1;
 
-  this->bad_flag_ = this->open ();
-
-  if (this->bad_flag_ == -1)
+  if ((this->bad_flag_ = this->open ()) == -1)
     ACE_ERROR ((LM_ERROR,
                 ACE_LIB_TEXT ("%p\n"),
                 ACE_LIB_TEXT ("ACE_Malloc_T<ACE_MEM_POOL_2, ACE_LOCK, ACE_CB>::ACE_Malloc_T")));
@@ -311,16 +310,14 @@
     bad_flag_ (0)
 {
   ACE_TRACE ("ACE_Malloc_T<ACE_MEM_POOL_2, ACE_LOCK, ACE_CB>::ACE_Malloc_T");
-  // Use pool_name for lock_name if lock_name not passed.
-  const ACE_TCHAR *name = lock_name ? lock_name : pool_name;
-  this->lock_ = ACE_Malloc_Lock_Adapter_T<ACE_LOCK> ()(name);
-  if (this->lock_ == 0)
-    return;
-
+  if (lock_name != 0)
+    ACE_NEW (this->lock_, ACE_LOCK (lock_name,0,ACE_DEFAULT_FILE_PERMS,options->sa_));
+  else
+    ACE_NEW (this->lock_, ACE_LOCK (ACE::basename (pool_name,
+                                                   ACE_DIRECTORY_SEPARATOR_CHAR),0,ACE_DEFAULT_FILE_PERMS,options->sa_));
   this->delete_lock_ = 1;
 
-  this->bad_flag_ = this->open ();
-  if (this->bad_flag_ == -1)
+  if ((this->bad_flag_ = this->open ()) == -1)
     ACE_ERROR ((LM_ERROR,
                 ACE_LIB_TEXT ("%p\n"),
                 ACE_LIB_TEXT ("ACE_Malloc_T<ACE_MEM_POOL_2, ACE_LOCK, ACE_CB>::ACE_Malloc_T")));
@@ -345,8 +342,7 @@
       return;
     }
 
-  this->bad_flag_ = this->open ();
-  if (this->bad_flag_ == -1)
+  if ((this->bad_flag_ = this->open ()) == -1)
     ACE_ERROR ((LM_ERROR,
                 ACE_LIB_TEXT ("%p\n"),
                 ACE_LIB_TEXT ("ACE_Malloc_T<ACE_MEM_POOL_2, ACE_LOCK, ACE_CB>::ACE_Malloc_T")));
@@ -363,13 +359,9 @@
 {
   ACE_TRACE ("ACE_Malloc_T<ACE_MEM_POOL_2, ACE_LOCK, ACE_CB>::ACE_Malloc_T");
 
-  this->lock_ = ACE_Malloc_Lock_Adapter_T<ACE_LOCK> ()(lock_name);
-  if (this->lock_ == 0)
-    return;
-
+  ACE_NEW (this->lock_, ACE_LOCK (lock_name));
   this->delete_lock_ = 1;
-  this->bad_flag_ = this->open ();
-  if (this->bad_flag_ == -1)
+  if ((this->bad_flag_ = this->open ()) == -1)
     ACE_ERROR ((LM_ERROR,
                 ACE_LIB_TEXT ("%p\n"),
                 ACE_LIB_TEXT ("ACE_Malloc_T<ACE_MEM_POOL_2, ACE_LOCK, ACE_CB>::ACE_Malloc_T")));
@@ -839,23 +831,6 @@
   return this->unbind (name, temp);
 }
 
-/*****************************************************************************/
-
-template <class ACE_LOCK> ACE_LOCK *
-ACE_Malloc_Lock_Adapter_T<ACE_LOCK>::operator () (const ACE_TCHAR *name)
-{
-  ACE_LOCK *p = 0;
-  if (name == 0)
-    ACE_NEW_RETURN (p, ACE_LOCK (name), 0);
-  else
-    ACE_NEW_RETURN (p, ACE_LOCK (ACE::basename (name,
-                                                ACE_DIRECTORY_SEPARATOR_CHAR)),
-                    0);
-  return p;
-}
-
-/*****************************************************************************/
-
 template <ACE_MEM_POOL_1, class ACE_LOCK, class ACE_CB> void
 ACE_Malloc_LIFO_Iterator_T<ACE_MEM_POOL_2, ACE_LOCK, ACE_CB>::dump (void) const
 {
