--- Mutex.h	2003-11-01 06:15:14.000000000 -0500
+++ /cygdrive/c/welchj/c/apps/ACE_wrappers/ace/Mutex.h	2004-02-17 11:53:24.714836000 -0500
@@ -52,7 +52,8 @@
   ACE_Mutex (int type = USYNC_THREAD,
              const ACE_TCHAR *name = 0,
              ACE_mutexattr_t *arg = 0,
-             mode_t mode = ACE_DEFAULT_FILE_PERMS);
+             mode_t mode = ACE_DEFAULT_FILE_PERMS,
+             LPSECURITY_ATTRIBUTES sa = 0);
 
   /// Implicitly destroy the mutex.
   ~ACE_Mutex (void);
