Index: Acceptor_Impl.cpp
===================================================================
--- Acceptor_Impl.cpp	(revision 86997)
+++ Acceptor_Impl.cpp	(working copy)
@@ -80,14 +80,24 @@
 
   if (this->ACE_Concurrency_Strategy<SVC_HANDLER>::activate_svc_handler (sh,
                                                                          arg) == -1)
-    return -1;
+    {
+      // Activation fails, decrease reference.
+      sh->transport ()->remove_reference ();
 
+      // #REFCOUNT# is zero at this point.
+
+      return -1;
+    }
+
   // The service handler has been activated. Now cache the handler.
   if (sh->add_transport_to_cache () == -1)
     {
       // Adding to the cache fails, close the handler.
       sh->close ();
 
+      // close() doesn't really decrease reference.
+      sh->transport ()->remove_reference ();
+
       // #REFCOUNT# is zero at this point.
 
       if (TAO_debug_level > 0)
@@ -154,6 +164,9 @@
       // Close handler.
       sh->close ();
 
+      // close() doesn't really decrease reference.
+      sh->transport ()->remove_reference ();
+
       // #REFCOUNT# is zero at this point.
 
       if (TAO_debug_level > 0)
@@ -194,7 +207,15 @@
 template <class SVC_HANDLER, ACE_PEER_ACCEPTOR_1> int
 TAO_Accept_Strategy<SVC_HANDLER, ACE_PEER_ACCEPTOR_2>::accept_svc_handler (SVC_HANDLER *svc_handler)
 {
-  return ACCEPT_STRATEGY_BASE::accept_svc_handler (svc_handler);
+  int const result = ACCEPT_STRATEGY_BASE::accept_svc_handler (svc_handler);
+  if (result == -1)
+    {
+      svc_handler->transport ()->remove_reference ();
+
+      // #REFCOUNT# is zero at this point.
+    }
+
+  return result;
 }
 
 
