Index: config-g++-common.h
===================================================================
--- config-g++-common.h	(revision 77871)
+++ config-g++-common.h	(working copy)
@@ -79,7 +79,7 @@
 #    define ACE_HAS_CUSTOM_EXPORT_MACROS
 #  endif  /* !ACE_HAS_CUSTOM_EXPORT_MACROS */
 #  define ACE_Proper_Export_Flag __attribute__ ((visibility("default")))
-#  define ACE_Proper_Import_Flag
+#  define ACE_Proper_Import_Flag __attribute__ ((visibility("default")))
 
 #  if (__GNUC__ > 4) || ((__GNUC__ == 4) && (__GNUC_MINOR__ >= 2))
 // Sadly, G++ 4.x silently ignores visibility attributes on
