Index: tao/PortableServer/POA.cpp
===================================================================
RCS file: /home/src/ACE_wrappers/TAO/tao/PortableServer/POA.cpp,v
retrieving revision 1.1.1.4
diff -u -r1.1.1.4 POA.cpp
--- tao/PortableServer/POA.cpp	13 Dec 2002 14:35:15 -0000	1.1.1.4
+++ tao/PortableServer/POA.cpp	31 Dec 2002 19:31:53 -0000
@@ -340,12 +340,6 @@
   // descrease it upon destruction.
   CORBA::add_ref (this->ort_template_);
   this->obj_ref_factory_ = this->ort_template_;
-
-  // Iterate over the registered IOR interceptors so that they may be
-  // given the opportunity to add tagged components to the profiles
-  // for this servant.
-  this->establish_components (ACE_ENV_SINGLE_ARG_PARAMETER);
-  ACE_CHECK;
 }
 
 TAO_POA::~TAO_POA (void)
@@ -563,6 +557,12 @@
                         0);
     }
 
+  // Iterate over the registered IOR interceptors so that they may be
+  // given the opportunity to add tagged components to the profiles
+  // for this servant.
+  poa->establish_components (ACE_ENV_SINGLE_ARG_PARAMETER);
+  ACE_CHECK_RETURN (0);
+
   // Note: Creating a POA using a POA manager that is in the active
   // state can lead to race conditions if the POA supports preexisting
   // objects, because the new POA may receive a request before its
@@ -3581,6 +3581,16 @@
 void
 TAO_POA::establish_components (ACE_ENV_SINGLE_ARG_DECL)
 {
+  // Iterate over the registered IOR interceptors so that they may be
+  // given the opportunity to add tagged components to the profiles
+  // for this servant.
+  TAO_IORInterceptor_List::TYPE &interceptors =
+    this->orb_core_.ior_interceptors ();
+
+  size_t interceptor_count = interceptors.size ();
+  if (interceptor_count == 0)
+    return;
+
   TAO_IORInfo * tao_info;
   ACE_NEW_THROW_EX (tao_info,
                     TAO_IORInfo (this),
@@ -3593,15 +3603,8 @@
 
   PortableInterceptor::IORInfo_var info = tao_info;
 
-  // Iterate over the registered IOR interceptors so that they may be
-  // given the opportunity to add tagged components to the profiles
-  // for this servant.
-  TAO_IORInterceptor_List::TYPE &interceptors =
-    this->orb_core_.ior_interceptors ();
-
-  size_t interceptor_count = interceptors.size ();
-  if (interceptor_count == 0)
-    return;
+  TAO_Object_Adapter::Non_Servant_Upcall non_servant_upcall (*this);
+  ACE_UNUSED_ARG (non_servant_upcall);
 
   for (size_t i = 0; i < interceptor_count; ++i)
     {
@@ -3664,8 +3667,6 @@
     this->orb_core_.ior_interceptors ();
 
   size_t interceptor_count = interceptors.size ();
-  if (interceptor_count == 0)
-    return;
 
   // All the establish_components() interception points have been
   // invoked. Now call the components_established() interception point
