Index: tests/Big_Twoways/Session.cpp
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tests/Big_Twoways/Session.cpp,v
retrieving revision 1.6
diff -u -r1.6 Session.cpp
--- tests/Big_Twoways/Session.cpp	29 Jan 2002 20:21:07 -0000	1.6
+++ tests/Big_Twoways/Session.cpp	27 Jun 2002 13:43:32 -0000
@@ -136,12 +136,15 @@
 
     if (this->active_thread_count_ != this->thread_count_)
       return;
+  }
 
-    this->barrier_.wait ();
+  this->validate_connections (ACE_ENV_SINGLE_ARG_PARAMETER);
+
+  this->barrier_.wait ();
+
+  if (this->running_ != 0)
+    return;
 
-    if (this->running_ != 0)
-      return;
-  }
   /// None of the threads are running, this session is useless at
   /// this point, report the problem and destroy the local objects
   this->terminate (0 ACE_ENV_ARG_PARAMETER);
@@ -220,6 +223,34 @@
     return 1;
 
   return 0;
+}
+
+void
+Session::validate_connections (ACE_ENV_ARG_DECL)
+  ACE_THROW_SPEC (())
+{
+  CORBA::ULong session_count =
+    this->other_sessions_.length ();
+  for (CORBA::ULong j = 0; j != session_count; ++j)
+    {
+      ACE_TRY
+        {
+#if (TAO_HAS_MESSAGING == 1)
+          CORBA::PolicyList_var unused;
+          this->other_sessions_[j]->_validate_connection(unused
+                                                         ACE_ENV_ARG_PARAMETER);
+          ACE_TRY_CHECK;
+#else
+          (void) this->other_sessions_[j]->_is_a("Not_An_IDL_Type"
+                                                 ACE_ENV_ARG_PARAMETER);
+          ACE_TRY_CHECK;
+#endif /* TAO_HAS_MESSAGING == 1 */
+        }
+      ACE_CATCHANY
+        {
+        }
+      ACE_ENDTRY;
+    }
 }
 
 void
Index: tests/Big_Twoways/Session.h
===================================================================
RCS file: /project/cvs-repository/ACE_wrappers-repository/TAO/tests/Big_Twoways/Session.h,v
retrieving revision 1.6
diff -u -r1.6 Session.h
--- tests/Big_Twoways/Session.h	1 May 2002 18:51:39 -0000	1.6
+++ tests/Big_Twoways/Session.h	27 Jun 2002 13:43:32 -0000
@@ -60,6 +60,10 @@
   /// Return 1 if all the work in this session has been completed
   int more_work (void) const;
 
+  /// Validate all the connections
+  void validate_connections (ACE_ENV_ARG_DECL)
+    ACE_THROW_SPEC (());
+
 private:
   /// Synchronize the internal state
   ACE_SYNCH_MUTEX mutex_;
