Index: UIPMC_Acceptor.cpp
===================================================================
--- UIPMC_Acceptor.cpp	(revision 78295)
+++ UIPMC_Acceptor.cpp	(working copy)
@@ -323,6 +323,8 @@
           ACE_CString name = opt.substring (0, slot);
           ACE_CString value = opt.substring (slot + 1);
 
+          begin = end + 1;
+
           if (name.length () == 0)
             ACE_ERROR_RETURN ((LM_ERROR,
                                ACE_TEXT ("TAO (%P|%t) Zero length UIPMC ")
@@ -342,8 +344,6 @@
                                ACE_TEXT ("TAO (%P|%t) Invalid UIPMC option: <%s>\n"),
                                name.c_str ()),
                               -1);
-
-          begin = end + 1;
         }
       else
         break;  // No other options.
