Index: TypeCode_CDR_Extraction.cpp
===================================================================
--- TypeCode_CDR_Extraction.cpp	(revision 84207)
+++ TypeCode_CDR_Extraction.cpp	(working copy)
@@ -109,6 +109,20 @@
     return true;
   }
 
+  class TAO_InputCDRByteOrderGuard
+  {
+    TAO_InputCDR &cdr_;
+    CORBA::Boolean byte_order_;
+  public:
+    TAO_InputCDRByteOrderGuard (TAO_InputCDR & cdr):
+      cdr_ (cdr),
+      byte_order_ (cdr.byte_order ()) {}
+    ~TAO_InputCDRByteOrderGuard ()
+    {
+      cdr_.reset_byte_order (byte_order_);
+    }
+  };
+
   // ---------------------------------------------------------
 
   CORBA::ULong const TYPECODE_INDIRECTION = 0xffffffffU;
@@ -324,6 +338,7 @@
 {
   // The remainder of a tk_objref TypeCode is encoded in a CDR
   // encapsulation.
+  TAO_InputCDRByteOrderGuard boguard (cdr);
   if (!start_cdr_encap_extraction (cdr))
     return false;
 
@@ -410,6 +425,7 @@
 
   // The remainder of a tk_struct/tk_except TypeCode is encoded in
   // a CDR encapsulation.
+  TAO_InputCDRByteOrderGuard boguard (cdr);
   if (!start_cdr_encap_extraction (cdr))
     return false;
 
@@ -502,6 +518,7 @@
   // The remainder of a tk_enum TypeCode is encoded in a CDR
   // encapsulation.
 
+  TAO_InputCDRByteOrderGuard boguard (cdr);
   if (!start_cdr_encap_extraction (cdr))
     return false;
 
@@ -777,7 +794,7 @@
 {
   // The remainder of a tk_enum TypeCode is encoded in a CDR
   // encapsulation.
-
+  TAO_InputCDRByteOrderGuard boguard (cdr);
   if (!start_cdr_encap_extraction (cdr))
     return false;
 
@@ -858,7 +875,7 @@
 
   // The remainder of a tk_sequence TypeCode is encoded in a CDR
   // encapsulation.
-
+  TAO_InputCDRByteOrderGuard boguard (cdr);
   if (!start_cdr_encap_extraction (cdr))
     return false;
 
@@ -898,7 +915,7 @@
 {
   // The remainder of a tk_alias or tk_value_box TypeCode is encoded
   // in a CDR encapsulation.
-
+  TAO_InputCDRByteOrderGuard boguard (cdr);
   if (!start_cdr_encap_extraction (cdr))
     return false;
 
@@ -1018,7 +1035,7 @@
 {
   // The remainder of a tk_value/tk_event TypeCode is encoded in a
   // CDR encapsulation
-
+  TAO_InputCDRByteOrderGuard boguard (cdr);
   if (!start_cdr_encap_extraction (cdr))
     return false;
 
@@ -1289,11 +1306,15 @@
              || kind == CORBA::tk_union
              || kind == CORBA::tk_value
              || kind == CORBA::tk_event
-             || kind == CORBA::tk_alias)
+             || kind == CORBA::tk_alias))
+      {
+        return false;
+      }
 
-        // Currently all recursive TypeCodes have complex parameter
-        // lists, meaning they are encoded as CDR encapsulations.
-        || !start_cdr_encap_extraction (indir_stream))
+      // Currently all recursive TypeCodes have complex parameter
+      // lists, meaning they are encoded as CDR encapsulations.
+      TAO_InputCDRByteOrderGuard boguard (indir_stream);
+      if (!start_cdr_encap_extraction (indir_stream))
       {
         return false;
       }
